{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d81228f-169c-4cc4-a088-8e7154ae144f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scimap as sm\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import scanpy as sc\n",
    "import seaborn as sns; sns.set(color_codes=True)\n",
    "import anndata\n",
    "adata = anndata.read_h5ad('/data/vasileiosionat2/Xenium/Drake_outputs/ccProcessed.h5ad')\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e6fb268-5782-44d6-8f5c-bc2a689b4f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "perio_count = (adata.obs['status.3'] == 'healthy').sum()\n",
    "print(perio_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b70d9e2-5002-4f48-9442-aebdf57d5740",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install openpyxl\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90ae8eb0-457d-425c-b060-bca3f45c18bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "# Save adata.var to an Excel file\n",
    "adata.var.to_excel(\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Xenium_genes.xlsx\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9fc41dd-cda1-41a0-9daf-1382bff93c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata.obs['total_area_mm2'].iloc[1000:1051].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22739c19-03a3-4af5-99a7-b94eafc5160a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686cfc49-e71b-4232-ad0a-3b14c3d861f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_knn50k10_merged'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92cf154a-27d9-42e4-94e7-49c51c3b4d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define the rotation angle in degrees\n",
    "    rotation_angle = 30\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad),  np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Horizontal flip matrix\n",
    "    horizontal_flip_matrix = np.array([\n",
    "        [1, 0],\n",
    "        [0, 1]\n",
    "    ])\n",
    "    \n",
    "    # Combined transformation matrix: Horizontal flip followed by rotation\n",
    "    combined_matrix = rotation_matrix @ horizontal_flip_matrix\n",
    "\n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        # Subset the data for the current sample\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        # Get the coordinates\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply the combined transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = combined_matrix @ coords\n",
    "        \n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate the aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "\n",
    "        # Set the figure size based on the aspect ratio\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate the scatter plot with adjusted coordinates\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set black background\n",
    "        fig.patch.set_facecolor('black')\n",
    "        ax.set_facecolor('black')\n",
    "\n",
    "        scatter = plt.scatter(\n",
    "            x=new_x_coords,\n",
    "            y=new_y_coords,\n",
    "            c=subset_adata.obs[cluster_key].map(cluster_colors),\n",
    "            s=4.9  # Slightly smaller dot size\n",
    "        )\n",
    "\n",
    "        # Remove grids\n",
    "        ax.grid(False)\n",
    "\n",
    "        # Remove ticks and their labels\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "        # Set axis limits\n",
    "        ax.set_xlim(new_x_coords.min(), new_x_coords.max())\n",
    "        ax.set_ylim(new_y_coords.min(), new_y_coords.max())\n",
    "\n",
    "        # Remove figure spines\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200  # μm\n",
    "\n",
    "    # Calculate the starting and ending positions for the scale bar\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "\n",
    "    # Place the scale bar slightly above the bottom of the plot\n",
    "        scale_bar_y = new_y_coords.min() + 0.01 * (new_y_coords.max() - new_y_coords.min())\n",
    "\n",
    "    # Plot the scale bar\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],  # X coordinates of the scale bar\n",
    "            [scale_bar_y, scale_bar_y],  # Y coordinates (constant for a horizontal bar)\n",
    "            color='white',  # Scale bar color\n",
    "            linewidth=8     # Thickness of the scale bar\n",
    "        )\n",
    "\n",
    "\n",
    "        # Set DPI and save the plot with specific filename\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='tiff', bbox_inches='tight', facecolor=fig.get_facecolor())  # Save the plot\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "sample_names = ['s2r2_HV184']  # List of sample names\n",
    "output_path = '/data/vasileiosionat2/Xenium/Figures/Cluster_tifs/Lvl1/{}.tif'\n",
    "\n",
    "lvl1_cluster_colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "lvl1_cluster_key = 'Lvl1'\n",
    "\n",
    "# Generate plots for level 1 clusters\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path.format(\"HV184C\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "023e61f0-22b5-4ab9-8c92-5c30c806dd4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2.5'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7479acfe-86f6-4e5e-af8b-e249652655d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f80cd02e-0f49-462e-ab32-077850acf6fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path, crop_coords=None):\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Crop if coordinates are provided\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio for figure size\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set white background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "\n",
    "        # Draw points on top\n",
    "        colors = subset_adata.obs[cluster_key].map(cluster_colors).fillna(\"#E5E5E5\").values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=4.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "# Example usage:\n",
    "sample_names = ['HV205B']  # List of sample names\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/HV205B_{}.pdf'\n",
    "\n",
    "Epi_cluster_colors = {\n",
    "    'Keratin': \"#097969\",\n",
    "    'Spinous': \"#BB5566FF\",\n",
    "    'Basal': \"#004488FF\",\n",
    "    'Crev': \"#FD7901FF\",\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "\n",
    "Strom_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#BB5566FF\",\n",
    "    'VEC': \"#004488FF\",\n",
    "    'SMC': \"#DDAA33FF\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "\n",
    "Lymphoid_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'NK': '#DDAA33FF',\n",
    "    'T': '#DE3163',\n",
    "    'Plasma': '#191970',\n",
    "    'B': '#0FFF50',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "  \n",
    "Myeloid_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#AA336A',\n",
    "    'mAPC': '#008000',\n",
    "    'Mast': '#E4D00A',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "# Generate plots\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Epi_cluster_colors, output_path.format(\"Epi\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Strom_cluster_colors, output_path.format(\"Strom\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Lymphoid_cluster_colors, output_path.format(\"Lymphoid\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Myeloid_cluster_colors, output_path.format(\"Myeloid\"), crop_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa8423de-11c9-47bb-ae0d-94a90184810a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2200, 3618  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 180, 2074  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Fibrous CT' : 'red',\n",
    "     'Epi-CT': '#00FFFF',\n",
    "     'Plasma-Fib CT' : '#CF9FFF',\n",
    "     'Plasma' : 'magenta' ,\n",
    "     'Spinous' : '#5D3FD3',\n",
    "     'Keratin' : '#0096FF' ,\n",
    "     'Crevicular' : '#800080' ,\n",
    "     'Lymphoid' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_knn50k10_merged'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_knn50k10_merged'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_knn50k10_merged'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_knn50k10_merged'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Niches_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53c3493e-b441-4675-9598-2b5e57abfd3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path, crop_coords=None):\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Crop if coordinates are provided\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio for figure size\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set white background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Draw points on top\n",
    "        colors = subset_adata.obs[cluster_key].map(cluster_colors).fillna(\"#E5E5E5\").values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=4.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "crop_coords = (2832, 4250, 480, 2374)\n",
    "# Example usage:\n",
    "sample_names = ['s1r3']  # List of sample names\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/HV214s_{}.pdf'\n",
    "\n",
    "Epi_cluster_colors = {\n",
    "    'Keratin': \"#097969\",\n",
    "    'Spinous': \"#BB5566FF\",\n",
    "    'Basal': \"#004488FF\",\n",
    "    'Crev': \"#FD7901FF\",\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "\n",
    "Strom_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#BB5566FF\",\n",
    "    'VEC': \"#004488FF\",\n",
    "    'SMC': \"#DDAA33FF\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "\n",
    "Lymphoid_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'NK': '#DDAA33FF',\n",
    "    'T': '#DE3163',\n",
    "    'Plasma': '#191970',\n",
    "    'B': '#0FFF50',\n",
    "    'Neutrophil': '#b7d2e0',\n",
    "    'mAPC': '#b7d2e0',\n",
    "    'Mast': '#b7d2e0',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "  \n",
    "Myeloid_cluster_colors = {\n",
    "    'Keratin': \"#e0cbb7\",\n",
    "    'Spinous': \"#e0cbb7\",\n",
    "    'Basal': \"#e0cbb7\",\n",
    "    'Crev': \"#e0cbb7\",\n",
    "    'Mel': '#E5E5E5E5',\n",
    "    'Fib': \"#d2e0b7\",\n",
    "    'Neur': \"#E5E5E5E5\",\n",
    "    'VEC': \"#e0b7b7\",\n",
    "    'SMC': \"#e0b7b7\",\n",
    "    'NK': '#b7d2e0',\n",
    "    'T': '#b7d2e0',\n",
    "    'Plasma': '#b7d2e0',\n",
    "    'B': '#b7d2e0',\n",
    "    'Neutrophil': '#AA336A',\n",
    "    'mAPC': '#008000',\n",
    "    'Mast': '#E4D00A',\n",
    "    'Other': \"#E5E5E5E5\"\n",
    "}\n",
    "# Generate plots\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Epi_cluster_colors, output_path.format(\"Epi\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Strom_cluster_colors, output_path.format(\"Strom\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Lymphoid_cluster_colors, output_path.format(\"Lymphoid\"), crop_coords)\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2.5', Myeloid_cluster_colors, output_path.format(\"Myeloid\"), crop_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ae9b876-c71d-49f2-85a2-124e18b5ef11",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2832, 4250,  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 480, 2374  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Fibrous CT' : 'red',\n",
    "     'Epi-CT': '#00FFFF',\n",
    "     'Plasma-Fib CT' : '#CF9FFF',\n",
    "     'Plasma' : 'magenta' ,\n",
    "     'Spinous' : '#5D3FD3',\n",
    "     'Keratin' : '#0096FF' ,\n",
    "     'Crevicular' : '#800080' ,\n",
    "     'Lymphoid' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_knn50k10_merged'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_knn50k10_merged'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_knn50k10_merged'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_knn50k10_merged'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Niches_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff62fc69-3633-4247-ae9e-ebee6291cc55",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d424560c-7f04-46c8-a43b-95d214e7137d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import scimap.plotting as sm\n",
    "import matplotlib.ticker as mtick\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "# List of SampleIDs to exclude\n",
    "excluded_sample_ids = ['s2r9_a', 's2r9_b', 'HV140A_C', 'HV140A_D', 's2r4_a', 's2r4_b', 's2r5', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', \n",
    "                       'HV192A_A', 'HV192A_B','HV192A_C']\n",
    "\n",
    "# Create a subset of the AnnData object excluding the specific SampleIDs\n",
    "adata_subset = adata[~adata.obs['sample'].isin(excluded_sample_ids)].copy()\n",
    "\n",
    "# Create a new figure with custom dimensions\n",
    "plt.figure(figsize=(50, 6))  # Adjust width and height as needed\n",
    "\n",
    "# Define the specific order of major_cluster categories\n",
    "anticipated_order = [ 'Spinous', 'Keratin', 'Crevicular',  'Fibrous CT',  'Epi-CT', 'Lymphoid', 'Plasma-Fib CT',  'Plasma']\n",
    "\n",
    "# Reorder the major_cluster column in the copied AnnData object\n",
    "adata_subset.obs['niche_knn50k10_merged'] = adata_subset.obs['niche_knn50k10_merged'].astype('category')\n",
    "adata_subset.obs['niche_knn50k10_merged'] = adata_subset.obs['niche_knn50k10_merged'].cat.reorder_categories(anticipated_order)\n",
    "adata_subset.obs = adata_subset.obs.sort_values('niche_knn50k10_merged')\n",
    "\n",
    "# Define custom colors for each major_cluster category\n",
    "colors = {\n",
    "    'Fibrous CT' : 'red',\n",
    "     'Epi-CT': '#00FFFF',\n",
    "     'Plasma-Fib CT' : '#CF9FFF',\n",
    "     'Plasma' : 'magenta' ,\n",
    "     'Spinous' : '#5D3FD3',\n",
    "     'Keratin' : '#0096FF' ,\n",
    "     'Crevicular' : '#800080' ,\n",
    "     'Lymphoid' : 'yellow'\n",
    "    \n",
    "}\n",
    "\n",
    "# Plot the stacked bar plot with specified colors\n",
    "plot = sm.stacked_barplot(adata_subset, x_axis='status.3', y_axis='niche_knn50k10_merged',\n",
    "                          order_yaxis=anticipated_order, method='percent', plot_tool='matplotlib', color=colors, figsize=(2, 6))\n",
    "\n",
    "# Get the current axes and set its background color to white\n",
    "plt.gca().set_facecolor('white')\n",
    "\n",
    "# Format y-axis ticks as percentages\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(1.0))\n",
    "\n",
    "# Get the x-axis limits\n",
    "x_min, x_max = plt.xlim()\n",
    "\n",
    "# Plot horizontal dotted lines at y = 0.25, 0.5, and 0.75\n",
    "plt.hlines(y=[0, 0.25, 0.5, 0.75, 1], xmin=x_min, xmax=x_max, linestyle='--', colors='grey')\n",
    "\n",
    "# Set y-axis tick labels to match the percentages\n",
    "plt.yticks([0, 0.25, 0.5, 0.75, 1], ['0%', '25%', '50%', '75%', '100%'])\n",
    "\n",
    "# Modify the borders\n",
    "plt.gca().spines['top'].set_visible(False)       # Make top border invisible\n",
    "plt.gca().spines['right'].set_visible(False)     # Make right border invisible\n",
    "plt.gca().spines['bottom'].set_color('black')    # Set bottom border color to black\n",
    "plt.gca().spines['left'].set_color('black')      # Set left border color to black\n",
    "\n",
    "# Adjust y-axis limits to start a little above the lower edge of the outline\n",
    "y_min, y_max = plt.ylim()\n",
    "plt.ylim(y_min - 0.03, y_max)\n",
    "\n",
    "# Remove 'clinical_status' from x-axis title\n",
    "plt.xlabel('')\n",
    "\n",
    "# Increase font size of x-axis tick labels (clinical_status categories)\n",
    "plt.xticks(fontsize=20)\n",
    "plt.yticks(fontsize=20)  # Adjust font size as needed\n",
    "\n",
    "# Rotate x-axis labels diagonally\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "# Custom legend handles as circles, sorted by anticipated_order in reverse\n",
    "sorted_legend_handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=colors[category], markersize=10) \n",
    "                         for category in reversed(anticipated_order)]\n",
    "sorted_legend_labels = reversed(anticipated_order)\n",
    "\n",
    "# Add legend with sorted custom handles and labels outside the plot\n",
    "legend = plt.legend(sorted_legend_handles, sorted_legend_labels, loc='upper left', bbox_to_anchor=(1, 1), markerscale=1, frameon=False, borderpad=2)\n",
    "\n",
    "# Save the plot as TIFF file\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_niche_stacked_barplot.pdf', format='pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "039ca8ce-6f09-4e52-97eb-3c604786ab3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl4'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303cae53-2d6a-4f8b-abc4-dd386eb81a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'adata.obs' is available and contains 'niche_knn30k11' and 'spatial_cluster'\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['niche_knn50k10_merged'] = df['niche_knn50k10_merged'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_knn50k10_merged', 'Lvl4']).size().unstack(fill_value=0)\n",
    "\n",
    "# Transpose the DataFrame to swap axes\n",
    "niche_groups = niche_groups.T\n",
    "\n",
    "# Define custom order for the x-axis (after transposing, so this is the niche clusters now)\n",
    "custom_order = ['Keratin', 'Spinous', 'Epi-CT', 'Crevicular',  'Fibrous CT', 'Plasma-Fib CT',  'Plasma', 'Lymphoid']\n",
    "\n",
    "# Reindex columns based on the custom order\n",
    "niche_groups = niche_groups.reindex(columns=custom_order)\n",
    "\n",
    "# Define subset of clusters for the y-axis\n",
    "y_axis_clusters = ['Ep.K',  'Ep.Sp', 'Ep.PB',   'Ep.B-PB',  'Ep.prol', \n",
    "                   'Ep.B',   'Mel',  'Lang',  'T.IE', 'cDC2', \n",
    "                   'Ep.Cr.B',  'Ep.Cr.PB',  'Ep.Cr.Sp.1',  'Ep.Cr.Sp.2',  'Fib.1',  'Fib.2',  'Fib.3',  'Fib.4',  'Fib.5', 'MyoF',\n",
    "                    'VEC.1', 'VEC.2', 'VEC.3', 'VEC.4', 'VEC.5',   'SMC', 'LEC', 'Neur',  'Mac', 'Mast', 'Mono', 'pDC',  'cDC1',\n",
    "                     'Pl.1', 'Pl.2',  'PB', 'NK', 'Tc', 'Treg', 'mregDC',    'Th_proximity_to_B', 'Th', 'T.IL7Rhi',\n",
    " 'T_proximity_to_Fib', 'B']\n",
    "niche_groups_filtered = niche_groups.loc[y_axis_clusters]\n",
    "\n",
    "# Plot heatmap with enhanced colormap and custom vmax\n",
    "plt.figure(figsize=(9, 16))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(\n",
    "    niche_groups_filtered, \n",
    "    cmap='coolwarm', \n",
    "    cbar_kws={'label': 'Count', 'ticks': [0, 500, 1000, 2000], 'format': '%.0f'}, \n",
    "    linewidths=0.05, \n",
    "    linecolor='black', \n",
    "    vmax=3000\n",
    ")\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=90, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Customize the color bar label\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "cbar.set_ticks([0, 500, 1000, 2000])\n",
    "cbar.ax.set_yticklabels(['0', '500', '1,000', '>2,000'])  # Set custom tick labels\n",
    "\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Lvl4_Clusters_Niches_Xenium.pdf\"\n",
    "#plt.savefig(output_path, format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "970890c9-1898-4836-8913-81bc4e061893",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'adata.obs' is available and contains 'niche_knn30k11' and 'spatial_cluster'\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['niche_knn50k10_merged'] = df['niche_knn50k10_merged'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_knn50k10_merged', 'Lvl2.5']).size().unstack(fill_value=0)\n",
    "\n",
    "# Transpose the DataFrame to swap axes\n",
    "niche_groups = niche_groups.T\n",
    "\n",
    "# Define custom order for the x-axis (after transposing, so this is the niche clusters now)\n",
    "custom_order = ['Crevicular', 'Keratin', 'Spinous',  'Epi-CT',  'Fibrous CT', 'Plasma-Fib CT',  'Plasma', 'Lymphoid']\n",
    "\n",
    "# Reindex columns based on the custom order\n",
    "niche_groups = niche_groups.reindex(columns=custom_order)\n",
    "\n",
    "# Define subset of clusters for the y-axis\n",
    "y_axis_clusters = ['Crev', 'Keratin', 'Spinous', 'Basal', 'Fib', 'VEC', 'SMC', 'mAPC', 'Mast', 'Plasma',  'T', 'B']\n",
    "niche_groups_filtered = niche_groups.loc[y_axis_clusters]\n",
    "\n",
    "# Plot heatmap with enhanced colormap and custom vmax\n",
    "plt.figure(figsize=(9, 16))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap( niche_groups_filtered, \n",
    "    cmap='coolwarm', \n",
    "    cbar_kws={'label': 'Count', 'ticks': [0, 2000, 5000, 10000], 'format': '%.0f'}, \n",
    "    linewidths=0.05, \n",
    "    linecolor='black', \n",
    "    vmax=10000\n",
    ")\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=90, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Customize the color bar label\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "cbar.set_ticks([0, 2000, 5000, 10000])\n",
    "cbar.ax.set_yticklabels(['0', '2000', '5,000', '>10,000'])  # Set custom tick labels\n",
    "\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lvl2-3_Clusters_Niches_Xenium.pdf\"\n",
    "plt.savefig(output_path, format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e7d8be7-c1e6-48c5-8ca7-f2652744254b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np  # You need this for np.log1p\n",
    "\n",
    "# Create DataFrame from adata.obs\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['niche_knn50k10_merged'] = df['niche_knn50k10_merged'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_knn50k10_merged', 'Lvl2.5']).size().unstack(fill_value=0)\n",
    "\n",
    "# Transpose the DataFrame\n",
    "niche_groups = niche_groups.T\n",
    "\n",
    "# Custom x-axis (niche) order\n",
    "custom_order = ['Keratin', 'Spinous', 'Epi-CT', 'TAE',  \n",
    "                'Im-Str', 'FibCT', 'Plasma', 'T-B-APC']\n",
    "niche_groups = niche_groups.reindex(columns=custom_order)\n",
    "\n",
    "# Custom y-axis (cell type) selection\n",
    "y_axis_clusters = ['Keratin', 'Spinous', 'Basal', 'TAE', 'APC', 'Mast', 'Fibroblast', 'VEC', 'SMC', 'Plasma', 'B', 'T']\n",
    "niche_groups_filtered = niche_groups.loc[y_axis_clusters]\n",
    "\n",
    "# Transpose so clusters are columns (x-axis), and niches are rows (y-axis)\n",
    "niche_groups_flipped = niche_groups_filtered.T  # Now niches are rows, clusters are columns\n",
    "\n",
    "# Apply log1p transformation\n",
    "niche_groups_log = niche_groups_flipped.applymap(lambda x: np.log1p(x))\n",
    "\n",
    "# Plot heatmap\n",
    "plt.figure(figsize=(7, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_log,\n",
    "    vmax=np.log1p(20000), \n",
    "    vmin=np.log1p(2000),       # log(0+1)\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'log(Count + 1)'},\n",
    "    linewidths=0.05,\n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Adjust colorbar ticks (log scale)\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "log_ticks = [np.log1p(2000), np.log1p(10000), np.log1p(20000)]\n",
    "cbar.set_ticks(log_ticks)\n",
    "cbar.ax.set_yticklabels(['2000', '10,000', '20,000'])\n",
    "\n",
    "plt.tight_layout()\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_LV3_Clusters_Niches_Log.pdf\"\n",
    "plt.savefig(output_path, format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89a314fe-b5f0-460d-b0eb-452b07a98ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Manually specify which clusters (x-axis) and niches (y-axis) to show, with order\n",
    "selected_clusters = ['Ep.K',  'Ep.Sp', 'Ep.PB',   'Ep.B-PB',  'Ep.prol', \n",
    "                   'Ep.B',   'Mel',  'Lang',  'T.IE', 'cDC2', \n",
    "                   'Ep.Cr.B',  'Ep.Cr.PB',  'Ep.Cr.Sp.1',  'Ep.Cr.Sp.2',  'Fib.1',  'Fib.2',  'Fib.3',  'Fib.4',  'Fib.5', 'MyoF',\n",
    "                    'VEC.1', 'VEC.2', 'VEC.3', 'VEC.4', 'VEC.5',   'SMC', 'LEC', 'Neur',  'Mac', 'Mast', 'Mono', 'pDC',  'cDC1',\n",
    "                     'Pl.1', 'Pl.2',  'PB', 'NK', 'Tc', 'Treg', 'mregDC',    'Th_proximity_to_B', 'Th', 'T.IL7Rhi',\n",
    " 'T_proximity_to_Fib', 'B']\n",
    "selected_niches = ['Keratin', 'Spinous', 'Epi-CT', 'Crevicular',  'Fibrous CT', 'Plasma-Fib CT',  'Plasma', 'Lymphoid']  # Replace with actual niche names\n",
    "\n",
    "# Ensure selected clusters and niches exist in the DataFrame\n",
    "niche_groups_filtered = niche_groups.loc[\n",
    "    niche_groups.index.intersection(selected_clusters),  # Use intersection() to filter selected clusters\n",
    "    niche_groups.columns.intersection(selected_niches)   # Use intersection() to filter selected niches\n",
    "]\n",
    "\n",
    "# Calculate relative frequencies across all niches, then select the relevant ones\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across all niches for each cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Compute relative frequencies\n",
    "\n",
    "# Filter the relative frequencies based on selected clusters and niches\n",
    "niche_groups_relative_filtered = niche_groups_relative.loc[\n",
    "    niche_groups_relative.index.intersection(selected_clusters),\n",
    "    niche_groups_relative.columns.intersection(selected_niches)\n",
    "]\n",
    "\n",
    "# Specify the order of clusters and niches\n",
    "cluster_order = ['Ep.K',  'Ep.Sp', 'Ep.PB',   'Ep.B-PB',  'Ep.prol', \n",
    "                   'Ep.B',   'Mel',  'Lang',  'T.IE', 'cDC2', \n",
    "                   'Ep.Cr.B',  'Ep.Cr.PB',  'Ep.Cr.Sp.1',  'Ep.Cr.Sp.2',  'Fib.1',  'Fib.2',  'Fib.3',  'Fib.4',  'Fib.5', 'MyoF',\n",
    "                    'VEC.1', 'VEC.2', 'VEC.3', 'VEC.4', 'VEC.5',   'SMC', 'LEC', 'Neur',  'Mac', 'Mast', 'Mono', 'pDC',  'cDC1',\n",
    "                     'Pl.1', 'Pl.2',  'PB', 'NK', 'Tc', 'Treg', 'mregDC',    'Th_proximity_to_B', 'Th', 'T.IL7Rhi',\n",
    " 'T_proximity_to_Fib', 'B']  # Adjust the order as needed\n",
    "\n",
    "niche_order = ['Keratin', 'Spinous', 'Epi-CT', 'Crevicular',  'Fibrous CT', 'Plasma-Fib CT',  'Plasma', 'Lymphoid']  # Adjust the order as needed\n",
    "\n",
    "# Plot heatmap with niches on y-axis and clusters on x-axis\n",
    "plt.figure(figsize=(16, 6))  # Adjusted figsize for readability\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered[niche_order].T[cluster_order], cmap='coolwarm',  # Apply niche_order on y-axis and cluster_order on x-axis\n",
    "    cbar_kws={'label': 'Relative Frequency'}, \n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.6,\n",
    "    xticklabels=cluster_order,  # Apply cluster_order on x-axis\n",
    "    yticklabels=niche_order  # Apply niche_order on y-axis\n",
    ")\n",
    "plt.title('Relative Cell Type Distribution in Niches')\n",
    "plt.xlabel('Cell Cluster')\n",
    "plt.ylabel('Niche')\n",
    "plt.xticks(rotation=45, ha='right')  # Improved alignment\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Prevents clipping\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_clusters_niches.svg'\n",
    "fig = plt.gcf()  # Get current figure\n",
    "#plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bdc268-60ac-436c-a231-2a2e71cb5700",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define the rotation angle in degrees\n",
    "    rotation_angle = -10\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad),  np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Horizontal flip matrix\n",
    "    horizontal_flip_matrix = np.array([\n",
    "        [1, 0],\n",
    "        [0, 1]\n",
    "    ])\n",
    "    \n",
    "    # Combined transformation matrix: Horizontal flip followed by rotation\n",
    "    combined_matrix = rotation_matrix @ horizontal_flip_matrix\n",
    "\n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        # Subset the data for the current sample\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        # Get the coordinates\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply the combined transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = combined_matrix @ coords\n",
    "        \n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate the aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "\n",
    "        # Set the figure size based on the aspect ratio\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate the scatter plot with adjusted coordinates\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set black background\n",
    "        fig.patch.set_facecolor('black')\n",
    "        ax.set_facecolor('black')\n",
    "\n",
    "        scatter = plt.scatter(\n",
    "            x=new_x_coords,\n",
    "            y=new_y_coords,\n",
    "            c=subset_adata.obs[cluster_key].map(cluster_colors),\n",
    "            s=4.9  # Slightly smaller dot size\n",
    "        )\n",
    "\n",
    "        # Remove grids\n",
    "        ax.grid(False)\n",
    "\n",
    "        # Remove ticks and their labels\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "        # Set axis limits\n",
    "        ax.set_xlim(new_x_coords.min(), new_x_coords.max())\n",
    "        ax.set_ylim(new_y_coords.min(), new_y_coords.max())\n",
    "\n",
    "        # Remove figure spines\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200  # μm\n",
    "\n",
    "    # Calculate the starting and ending positions for the scale bar\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "\n",
    "    # Place the scale bar slightly above the bottom of the plot\n",
    "        scale_bar_y = new_y_coords.min() + 0.01 * (new_y_coords.max() - new_y_coords.min())\n",
    "\n",
    "    # Plot the scale bar\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],  # X coordinates of the scale bar\n",
    "            [scale_bar_y, scale_bar_y],  # Y coordinates (constant for a horizontal bar)\n",
    "            color='white',  # Scale bar color\n",
    "            linewidth=8     # Thickness of the scale bar\n",
    "        )\n",
    "\n",
    "\n",
    "        # Set DPI and save the plot with specific filename\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='tiff', bbox_inches='tight', facecolor=fig.get_facecolor())  # Save the plot\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "sample_names = ['s2r1_HV207']  # List of sample names\n",
    "output_path = '/data/vasileiosionat2/Xenium/Figures/Cluster_tifs/Lvl1/{}.tif'\n",
    "\n",
    "lvl1_cluster_colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "lvl1_cluster_key = 'Lvl1'\n",
    "\n",
    "# Generate plots for level 1 clusters\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path.format(\"HV207\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b3563f-f120-4d9d-be8e-a9a54adace98",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445c652a-1443-4ae5-a60e-50d75a7ec703",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbfe3191-67f4-4854-b9ff-9e84c028a76a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['pt.id'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e343d76-d19e-4092-9883-cd5c26bf862b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import scimap.plotting as sm\n",
    "import matplotlib.ticker as mtick\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "# List of SampleIDs to exclude\n",
    "excluded_sample_ids = ['s2r9_a', 's2r9_b', 'HV140A_C', 'HV140A_D', 's2r4_a', 's2r4_b', 's2r5', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', \n",
    "                       'HV192A_A', 'HV192A_B','HV192A_C']\n",
    "\n",
    "# Create a subset of the AnnData object excluding the specific SampleIDs\n",
    "adata_subset = adata[~adata.obs['sample'].isin(excluded_sample_ids)].copy()\n",
    "\n",
    "# Create a new figure with custom dimensions\n",
    "plt.figure(figsize=(50, 6))  # Adjust width and height as needed\n",
    "\n",
    "# Define the specific order of major_cluster categories\n",
    "anticipated_order = ['Other', 'Epithelial', 'Immune', 'Fibroblast', 'Vascular']\n",
    "\n",
    "# Reorder the major_cluster column in the copied AnnData object\n",
    "adata_subset.obs['Lvl1'] = adata_subset.obs['Lvl1'].astype('category')\n",
    "adata_subset.obs['Lvl1'] = adata_subset.obs['Lvl1'].cat.reorder_categories(anticipated_order)\n",
    "adata_subset.obs = adata_subset.obs.sort_values('Lvl1')\n",
    "\n",
    "# Define custom colors for each major_cluster category\n",
    "colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "\n",
    "# Plot the stacked bar plot with specified colors\n",
    "plot = sm.stacked_barplot(adata_subset, x_axis='status.3', y_axis='Lvl1',\n",
    "                          order_yaxis=anticipated_order, method='percent', plot_tool='matplotlib', color=colors, figsize=(2, 6))\n",
    "\n",
    "# Get the current axes and set its background color to white\n",
    "plt.gca().set_facecolor('white')\n",
    "\n",
    "# Format y-axis ticks as percentages\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(1.0))\n",
    "\n",
    "# Get the x-axis limits\n",
    "x_min, x_max = plt.xlim()\n",
    "\n",
    "# Plot horizontal dotted lines at y = 0.25, 0.5, and 0.75\n",
    "plt.hlines(y=[0, 0.25, 0.5, 0.75, 1], xmin=x_min, xmax=x_max, linestyle='--', colors='grey')\n",
    "\n",
    "# Set y-axis tick labels to match the percentages\n",
    "plt.yticks([0, 0.25, 0.5, 0.75, 1], ['0%', '25%', '50%', '75%', '100%'])\n",
    "\n",
    "# Modify the borders\n",
    "plt.gca().spines['top'].set_visible(False)       # Make top border invisible\n",
    "plt.gca().spines['right'].set_visible(False)     # Make right border invisible\n",
    "plt.gca().spines['bottom'].set_color('black')    # Set bottom border color to black\n",
    "plt.gca().spines['left'].set_color('black')      # Set left border color to black\n",
    "\n",
    "# Adjust y-axis limits to start a little above the lower edge of the outline\n",
    "y_min, y_max = plt.ylim()\n",
    "plt.ylim(y_min - 0.03, y_max)\n",
    "\n",
    "# Remove 'clinical_status' from x-axis title\n",
    "plt.xlabel('')\n",
    "\n",
    "# Increase font size of x-axis tick labels (clinical_status categories)\n",
    "plt.xticks(fontsize=20)\n",
    "plt.yticks(fontsize=20)  # Adjust font size as needed\n",
    "\n",
    "# Rotate x-axis labels diagonally\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "# Custom legend handles as circles, sorted by anticipated_order in reverse\n",
    "sorted_legend_handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=colors[category], markersize=10) \n",
    "                         for category in reversed(anticipated_order)]\n",
    "sorted_legend_labels = reversed(anticipated_order)\n",
    "\n",
    "# Add legend with sorted custom handles and labels outside the plot\n",
    "legend = plt.legend(sorted_legend_handles, sorted_legend_labels, loc='upper left', bbox_to_anchor=(1, 1), markerscale=1, frameon=False, borderpad=2)\n",
    "\n",
    "# Save the plot as TIFF file\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Xenium_stacked_barplot.svg', format='svg')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a99244-89e2-4135-ab22-9c6ee327b19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Epithelial', 'Fibroblast', 'Immune', 'Vascular']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'KRT19', 'EHF', 'KRT5', \n",
    "    'COL5A2', 'VCAN', 'PCOLCE', \n",
    "    'CYTIP', 'PTPRC', 'CXCR4',\n",
    "    'CD34', 'EGFL7', 'VWF'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl1'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl1'] = adata_subset.obs['Lvl1'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl1',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Xenium_Lvl1_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f3eb9e3-2a2d-44fc-a8e6-2a03803d7e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# Load the compressed CSV file\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Preview the data\n",
    "print(data.head())\n",
    "\n",
    "# Group the data by 'cell_id'\n",
    "grouped = data.groupby('cell_id')\n",
    "\n",
    "# Set the style for a black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Create the plot\n",
    "plt.figure(figsize=(12, 10))\n",
    "\n",
    "# Plot each mask\n",
    "for cell_id, group in grouped:\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    # Plot the mask with grey fill and black contour\n",
    "    plt.fill(x, y, color='darkgrey', edgecolor='white', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Customize the plot\n",
    "plt.title(\"Nuclei Masks (Full Dataset)\")\n",
    "plt.xlabel(\"X Coordinate\")\n",
    "plt.ylabel(\"Y Coordinate\")\n",
    "plt.axis(\"equal\")  # Ensure equal aspect ratio\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ed49f03-2379-4c43-9540-89be078c2fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# Load the compressed CSV file\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Preview the data\n",
    "print(data.head())\n",
    "\n",
    "# Group the data by 'cell_id'\n",
    "grouped = data.groupby('cell_id')\n",
    "\n",
    "# Set the style for a black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Create the plot\n",
    "plt.figure(figsize=(12, 10))\n",
    "\n",
    "# Plot each mask\n",
    "for cell_id, group in grouped:\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    # Plot the mask with grey fill and black contour\n",
    "    plt.fill(x, y, color='darkgrey', edgecolor='white', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Customize the plot\n",
    "plt.title(\"Nuclei Masks (Full Dataset)\")\n",
    "plt.xlabel(\"X Coordinate\")\n",
    "plt.ylabel(\"Y Coordinate\")\n",
    "plt.axis(\"equal\")  # Ensure equal aspect ratio\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e941ae87-82ad-44d2-8775-2d0ea04b68c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the compressed CSV file\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Preview the data\n",
    "print(data.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3db06d15-7700-49fb-8b1e-d13ff9eb8ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the compressed CSV file\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/transcripts.csv.gz\"\n",
    "data1 = pd.read_csv(csv_path)\n",
    "\n",
    "# Preview the data\n",
    "print(data1.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84743659-d96f-4e5d-aeec-00a7242f640c",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmin, xmax = 4500, 5000  # Horizontal bounds\n",
    "ymin, ymax = 2500, 3000  # Vertical bounds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67f69983-9860-4ac6-9b76-b142a4d3c169",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "total_xmin, total_xmax = 200, 3800  # Horizontal bounds\n",
    "total_ymin, total_ymax = 200, 2500  # Vertical bounds\n",
    "\n",
    "total_vertices = data[\n",
    "    (data['vertex_x'] >= total_xmin) & (data['vertex_x'] <= total_xmax) &\n",
    "    (data['vertex_y'] >= total_ymin) & (data['vertex_y'] <= total_ymax)\n",
    "]\n",
    "total_cell_ids = total_vertices['cell_id'].unique()\n",
    "total_data = data[data['cell_id'].isin(total_cell_ids)]\n",
    "grouped = total_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Epi': \"red\",     # Smooth Muscle Cells (SMC) - Red\n",
    "    'CT': \"blue\",\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#E5E5E5\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(total_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['region'].to_dict()\n",
    "# Add a cluster column to the filtered CSV data\n",
    "total_data = total_vertices.copy()\n",
    "total_data['region'] = total_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "total_data = total_data.dropna(subset=['region'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(36, 23))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in total_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['region'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(total_xmin, total_xmax)\n",
    "plt.ylim(total_ymin, total_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (2800, 3075, 1300, 1575),  # Lymphoid ROI\n",
    "    (2300, 2575, 1000, 1275),  # Deep ROI\n",
    "    (2800, 3075, 225, 500),    # Oral ROI\n",
    "    (2925, 3200, 1725, 2000)   # Crevicular ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Total_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7feb9e25-fb6c-4271-a63b-1bc4abb7591f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "total_xmin, total_xmax = 2832, 4250  # Horizontal bounds\n",
    "total_ymin, total_ymax = 480, 2374  # Vertical bounds\n",
    "\n",
    "total_vertices = data[\n",
    "    (data['vertex_x'] >= total_xmin) & (data['vertex_x'] <= total_xmax) &\n",
    "    (data['vertex_y'] >= total_ymin) & (data['vertex_y'] <= total_ymax)\n",
    "]\n",
    "total_cell_ids = total_vertices['cell_id'].unique()\n",
    "total_data = data[data['cell_id'].isin(total_cell_ids)]\n",
    "grouped = total_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Epi': \"red\",     # Smooth Muscle Cells (SMC) - Red\n",
    "    'CT': \"blue\",\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#E5E5E5\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(total_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['region'].to_dict()\n",
    "# Add a cluster column to the filtered CSV data\n",
    "total_data = total_vertices.copy()\n",
    "total_data['region'] = total_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "total_data = total_data.dropna(subset=['region'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(36, 23))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in total_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['region'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(total_xmin, total_xmax)\n",
    "plt.ylim(total_ymin, total_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (3300, 3575, 900, 1175),  # Lymphoid ROI\n",
    "    (2875, 3150, 1400, 1675),  # Deep ROI\n",
    "    (3475, 3750, 1800, 2075),    # Oral ROI\n",
    "    (3800, 4075, 750, 1025)   # Crevicular ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Total_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a09fbe24-6c5e-4c57-b4d3-4f22db2db4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(total_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67809325-ad87-4596-9bf6-d71d4418fba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "crev_xmin, crev_xmax = 2925, 3200  # Horizontal bounds\n",
    "crev_ymin, crev_ymax = 1725, 2000  # Vertical bounds\n",
    "\n",
    "crev_vertices = data[\n",
    "    (data['vertex_x'] >= crev_xmin) & (data['vertex_x'] <= crev_xmax) &\n",
    "    (data['vertex_y'] >= crev_ymin) & (data['vertex_y'] <= crev_ymax)\n",
    "]\n",
    "crev_cell_ids = crev_vertices['cell_id'].unique()\n",
    "crev_data = data[data['cell_id'].isin(crev_cell_ids)]\n",
    "grouped = crev_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Crev': \"#702963\",     \n",
    "    'VEC': \"#FFAC1C\",\n",
    "    'Fib': \"#FFEA00\",     \n",
    "    'mAPC': \"#097969\",    \n",
    "}\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(crev_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "crev_data = crev_vertices.copy()\n",
    "crev_data['Lvl2.5'] = crev_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "crev_data = crev_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in crev_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(crev_xmin, crev_xmax)\n",
    "plt.ylim(crev_ymin, crev_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Crev_Segmentation_Mask_{sample_id}.svg\"\n",
    "# Save the plot as an SVG file with a black background\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4290d730-3086-489c-b2f8-850e373d896b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "crev_xmin, crev_xmax = 3800, 4075  # Horizontal bounds\n",
    "crev_ymin, crev_ymax = 750, 1025  # Vertical bounds\n",
    "\n",
    "crev_vertices = data[\n",
    "    (data['vertex_x'] >= crev_xmin) & (data['vertex_x'] <= crev_xmax) &\n",
    "    (data['vertex_y'] >= crev_ymin) & (data['vertex_y'] <= crev_ymax)\n",
    "]\n",
    "crev_cell_ids = crev_vertices['cell_id'].unique()\n",
    "crev_data = data[data['cell_id'].isin(crev_cell_ids)]\n",
    "grouped = crev_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Crev': \"#702963\",     \n",
    "    'VEC': \"#FFAC1C\",\n",
    "    'Fib': \"#FFEA00\",     \n",
    "    'mAPC': \"#097969\",    \n",
    "}\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(crev_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "crev_data = crev_vertices.copy()\n",
    "crev_data['Lvl2.5'] = crev_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "crev_data = crev_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in crev_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(crev_xmin, crev_xmax)\n",
    "plt.ylim(crev_ymin, crev_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Crev_Segmentation_Mask_{sample_id}.svg\"\n",
    "# Save the plot as an SVG file with a black background\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af3aec53-4944-4b5a-9790-de58ae17cdb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "oral_xmin, oral_xmax = 2800, 3075  # Horizontal bounds\n",
    "oral_ymin, oral_ymax = 225, 500  # Vertical bounds\n",
    "\n",
    "oral_vertices = data[\n",
    "    (data['vertex_x'] >= oral_xmin) & (data['vertex_x'] <= oral_xmax) &\n",
    "    (data['vertex_y'] >= oral_ymin) & (data['vertex_y'] <= oral_ymax)\n",
    "]\n",
    "oral_cell_ids = oral_vertices['cell_id'].unique()\n",
    "oral_data = data[data['cell_id'].isin(oral_cell_ids)]\n",
    "grouped = oral_data.groupby('cell_id')\n",
    "\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Keratin': \"#A7C7E7\",\n",
    "    'Spinous': \"#5D3FD3\",\n",
    "    'Basal': \"#0096FF\",\n",
    "    'mAPC': \"#097969\"\n",
    "}\n",
    "\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(oral_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "oral_data = oral_vertices.copy()\n",
    "oral_data['Lvl2.5'] = oral_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "oral_data = oral_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in oral_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(oral_xmin, oral_xmax)\n",
    "plt.ylim(oral_ymin, oral_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Oral_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15fe608a-1fc9-477d-a214-96ec37a5f1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "oral_xmin, oral_xmax = 3475, 3750  # Horizontal bounds\n",
    "oral_ymin, oral_ymax = 1800, 2075  # Vertical bounds\n",
    "\n",
    "oral_vertices = data[\n",
    "    (data['vertex_x'] >= oral_xmin) & (data['vertex_x'] <= oral_xmax) &\n",
    "    (data['vertex_y'] >= oral_ymin) & (data['vertex_y'] <= oral_ymax)\n",
    "]\n",
    "oral_cell_ids = oral_vertices['cell_id'].unique()\n",
    "oral_data = data[data['cell_id'].isin(oral_cell_ids)]\n",
    "grouped = oral_data.groupby('cell_id')\n",
    "\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "unique_cluster_colors = {\n",
    "    'Keratin': \"#A7C7E7\",\n",
    "    'Spinous': \"#5D3FD3\",\n",
    "    'Basal': \"#0096FF\",\n",
    "    'mAPC': \"#097969\"\n",
    "}\n",
    "\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(oral_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "oral_data = oral_vertices.copy()\n",
    "oral_data['Lvl2.5'] = oral_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "oral_data = oral_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in oral_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(oral_xmin, oral_xmax)\n",
    "plt.ylim(oral_ymin, oral_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Oral_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f840e515-08e9-489e-ae93-144113a6ded7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2300, 2575  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1000, 1275  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Fib': \"#FFEA00\",\n",
    "    'VEC': \"#FFAC1C\",\n",
    "    'mAPC': \"#097969\",\n",
    "    'Plasma' : '#7F00FF',\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl2.5'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Deep_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5d91d89-92a9-4702-8da4-32a93b23f89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2875, 3150  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1400, 1675  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Fib': \"#FFEA00\",\n",
    "    'VEC': \"#FFAC1C\",\n",
    "    'mAPC': \"#097969\",\n",
    "    'Plasma' : '#7F00FF',\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#BEBEBE\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl2.5'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Deep_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b9ca54-6b80-476a-816f-c5e1374f26d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "lymph_xmin, lymph_xmax = 2800, 3075  # Horizontal bounds\n",
    "lymph_ymin, lymph_ymax = 1300, 1575  # Vertical bounds\n",
    "\n",
    "lymph_vertices = data[\n",
    "    (data['vertex_x'] >= lymph_xmin) & (data['vertex_x'] <= lymph_xmax) &\n",
    "    (data['vertex_y'] >= lymph_ymin) & (data['vertex_y'] <= lymph_ymax)\n",
    "]\n",
    "lymph_cell_ids = lymph_vertices['cell_id'].unique()\n",
    "lymph_data = data[data['cell_id'].isin(lymph_cell_ids)]\n",
    "grouped = lymph_data.groupby('cell_id')\n",
    "\n",
    "\n",
    "unique_cluster_colors = {\n",
    "    'Plasma': \"#7F00FF\",   # Plasma Cells - Purple\n",
    "    'T': \"#8B0000\",       # T Cells - Dark Red\n",
    "    'mAPC': \"#097969\",    # Macrophages (Mac) - Dark Orange\n",
    "    'B': '#CF9FFF',\n",
    "    'Fib': '#FFEA00'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#E5E5E5\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV205B\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(lymph_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "lymph_data = lymph_vertices.copy()\n",
    "lymph_data['Lvl2.5'] = lymph_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "lymph_data = lymph_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in lymph_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(lymph_xmin, lymph_xmax)\n",
    "plt.ylim(lymph_ymin, lymph_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lymph_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68159881-4d5d-440f-8862-4c6f7754f4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/nucleus_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "lymph_xmin, lymph_xmax = 3300, 3575  # Horizontal bounds\n",
    "lymph_ymin, lymph_ymax = 900, 1175  # Vertical bounds\n",
    "\n",
    "lymph_vertices = data[\n",
    "    (data['vertex_x'] >= lymph_xmin) & (data['vertex_x'] <= lymph_xmax) &\n",
    "    (data['vertex_y'] >= lymph_ymin) & (data['vertex_y'] <= lymph_ymax)\n",
    "]\n",
    "lymph_cell_ids = lymph_vertices['cell_id'].unique()\n",
    "lymph_data = data[data['cell_id'].isin(lymph_cell_ids)]\n",
    "grouped = lymph_data.groupby('cell_id')\n",
    "\n",
    "\n",
    "unique_cluster_colors = {\n",
    "    'Plasma': \"#7F00FF\",   # Plasma Cells - Purple\n",
    "    'T': \"#8B0000\",       # T Cells - Dark Red\n",
    "    'mAPC': \"#097969\",    # Macrophages (Mac) - Dark Orange\n",
    "    'B': '#CF9FFF',\n",
    "    'Fib': '#FFEA00'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#E5E5E5\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(lymph_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl2.5'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "lymph_data = lymph_vertices.copy()\n",
    "lymph_data['Lvl2.5'] = lymph_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "lymph_data = lymph_data.dropna(subset=['Lvl2.5'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in lymph_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl2.5'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(lymph_xmin, lymph_xmax)\n",
    "plt.ylim(lymph_ymin, lymph_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lymph_Segmentation_Mask_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7358f7e9-fa25-485a-bf7a-3fcf9c49af64",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2.5'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60cc84e8-57f9-47be-99f1-4e91776b08ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cropped.obs.columns.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fc4fa89-265d-4d14-8d3d-09301e74b57c",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_data['region'] = total_data['cell_id'].map(adata_cropped.obs['region'])\n",
    "print(total_data[['cell_id', 'region']].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3da2d23-000c-47e6-960a-1dc5fd4a5767",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 2800, 3075  # Horizontal bounds\n",
    "ymin, ymax = 225, 500   # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"KRT19\", \"KRT5\", \"HLA-DRB5\", \"CNFN\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"KRT19\": \"#5D3FD3\",  \n",
    "    \"HLA-DRB5\": \"#0FFF50\",  \n",
    "    \"KRT5\": \"#0096FF\",\n",
    "    \"CNFN\": \"#40E0D0\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=1)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Oral_transcripts_Mask_HV205B.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "296c6b58-1144-4937-a1af-4ee60e220ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/nucleus_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 3475, 3750  # Horizontal bounds\n",
    "ymin, ymax = 1800, 2075\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"KRT19\", \"KRT5\", \"HLA-DRB5\", \"CNFN\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"KRT19\": \"#5D3FD3\",  \n",
    "    \"HLA-DRB5\": \"#0FFF50\",  \n",
    "    \"KRT5\": \"#0096FF\",\n",
    "    \"CNFN\": \"#40E0D0\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=1)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Oral_transcripts_Mask_HV214As.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9cb5de8-5d0f-473b-9eb0-2a26502c0a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 2300, 2575\n",
    "ymin, ymax =1000, 1275  # Horizontal bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"VCAN\", \"VWF\", \"PRDM1\", \"HLA-DRB5\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"VCAN\": \"#FF5F1F\",  \n",
    "    \"VWF\": \"#FFFF00\",  \n",
    "    \"PRDM1\": \"#00FFFF\",\n",
    "    \"HLA-DRB5\": \"#0FFF50\"\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Deep_transcripts_Mask_HV205B.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0887ee76-a59b-4c85-bcc3-51f97bf7232e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 2875, 3150  # Horizontal bounds\n",
    "ymin, ymax = 1400, 1675  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"VCAN\", \"VWF\", \"PRDM1\", \"HLA-DRB5\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"VCAN\": \"#FF5F1F\",  \n",
    "    \"VWF\": \"#FFFF00\",  \n",
    "    \"PRDM1\": \"#00FFFF\",\n",
    "    \"HLA-DRB5\": \"#0FFF50\"\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Deep_transcripts_Mask_HV214Bs.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803b1137-9028-41bc-b7e3-10ac27000475",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 2925, 3200  # Horizontal bounds\n",
    "ymin, ymax = 1725, 2000  # Horizontal bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"ODAM\", \"VCAN\", \"VWF\", \"HLA-DRB5\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"ODAM\": \"#DA70D6\",  \n",
    "    \"VCAN\": \"#FF5F1F\",  \n",
    "    \"HLA-DRB5\": \"#0FFF50\",  \n",
    "    \"VWF\": \"#FFFF00\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Crev_transcripts_Mask_HV205B.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf2565a-7f30-493e-88c2-b76381f1666f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r3_HV214As/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 3800, 4075  # Horizontal bounds\n",
    "ymin, ymax = 750, 1025  # Horizontal bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"ODAM\", \"VCAN\", \"VWF\", \"HLA-DRB5\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"ODAM\": \"#DA70D6\",  \n",
    "    \"VCAN\": \"#FF5F1F\",  \n",
    "    \"HLA-DRB5\": \"#0FFF50\",  \n",
    "    \"VWF\": \"#FFFF00\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Crev_transcripts_Mask_HV214Bs.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b124652a-64b0-4384-9bdb-79741cd86777",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV205B/nucleus_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Define the boundaries of the cropped area\n",
    "xmin, xmax = 2800, 3075  # Horizontal bounds\n",
    "ymin, ymax = 1300, 1575  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [ \"TRAC\", \"BANK1\", \"HLA-DRB5\", \"PRDM1\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    'BANK1': '#F9F6EE',\n",
    "    \"TRAC\": \"#E0115F\",  \n",
    "    \"HLA-DRB5\": \"#0FFF50\",  \n",
    "    \"PRDM1\": \"#00FFFF\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=2, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lymph_transcripts_Mask_HV205B.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02e2b5c2-29d8-4398-af90-6c2cee4ea5fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2.5'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b85f875-7fea-47f5-a13b-f30ed95f1744",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cropped.obs.columns.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a42e8b8-3b1d-45d6-921e-ded884c59a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_data['region'] = total_data['cell_id'].map(adata_cropped.obs['region'])\n",
    "print(total_data[['cell_id', 'region']].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a221f6-545b-4774-ad6a-fee3b10c3937",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 1: Subset the adata object to only the specific clusters and create a copy\n",
    "desired_clusters = [ 'Plasma', 'T', 'VEC', 'Fib', 'mAPC', 'Mast', 'SMC', 'Basal', 'Spinous', 'Keratin', 'NK', 'Neur', 'Crev', 'B', 'Mel']\n",
    "\n",
    "adata_subset = adata[adata.obs['Lvl2.5'].isin(desired_clusters)].copy()\n",
    "\n",
    "# Ensure 'Lvl2.5' is categorical\n",
    "adata_subset.obs['Lvl2.5'] = adata_subset.obs['Lvl2.5'].astype('category')\n",
    "\n",
    "# Step 2: Check if the data is already log-transformed, otherwise transform it\n",
    "if 'log1p' not in adata_subset.uns_keys():\n",
    "    sc.pp.log1p(adata_subset)  # Apply log transformation if not already done\n",
    "\n",
    "# Manually set the base in log1p if missing\n",
    "if 'log1p' in adata_subset.uns_keys() and 'base' not in adata_subset.uns['log1p']:\n",
    "    adata_subset.uns['log1p']['base'] = 2.718281828459045  # Default to e\n",
    "\n",
    "# Step 3: Perform differential expression analysis for the selected clusters\n",
    "sc.tl.rank_genes_groups(adata_subset, 'Lvl2.5', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for the selected clusters\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata_subset.obs['Lvl2.5'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata_subset.uns['rank_genes_groups']['names'][cluster][:10]\n",
    "\n",
    "# Step 5: Generate the dendrogram to get cluster order\n",
    "sc.tl.dendrogram(adata_subset, groupby='Lvl2.5')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata_subset.uns['dendrogram_Lvl2.5']['categories_ordered']\n",
    "\n",
    "# Step 7: Ensure each cluster has at least 3 unique genes\n",
    "unique_genes = set()\n",
    "genes_for_plot = []\n",
    "\n",
    "for cluster in cluster_order:\n",
    "    if cluster in top_genes_per_cluster:\n",
    "        cluster_genes = top_genes_per_cluster[cluster]\n",
    "        genes_for_this_cluster = []\n",
    "        for gene in cluster_genes:\n",
    "            if len(genes_for_this_cluster) >= 5:\n",
    "                break\n",
    "            if gene not in unique_genes:\n",
    "                unique_genes.add(gene)\n",
    "                genes_for_this_cluster.append(gene)\n",
    "        genes_for_plot.extend(genes_for_this_cluster)\n",
    "\n",
    "# Step 8: Plot the matrixplot with the unique genes\n",
    "sc.pl.dotplot(adata_subset, var_names=genes_for_plot, groupby='Lvl2.5', \n",
    "                 dendrogram=True, use_raw=False, cmap=\"vlag\", \n",
    "                 standard_scale='var', swap_axes=True)\n",
    "print(top_genes_per_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99424342-5ffc-49d5-8791-d9f9911e1282",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures fonts are embedded as text, not outlines\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Keratin', 'Spinous', 'Basal', 'Crev', 'Plasma', 'T', 'B', 'mAPC', 'VEC', 'Mast',  'Fib', 'SMC']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'CNFN', 'EHF', 'AQP3', \n",
    "    'KRT19', 'SDC1', 'KRT5', 'COL17A1', \n",
    "    'ODAM',    'PRDM1', 'MZB1', 'TRAC', 'CD3E',\n",
    "    'BANK1', 'HLA-DRB5', 'MS4A6A', 'CD14',\n",
    "    'CD34', 'VWF', 'CPA3', 'MS4A2',  'VCAN', 'COL16A1', 'MYH11', 'ACTA2'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl2.5'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl2.5'] = adata_subset.obs['Lvl2.5'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl2.5',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8, # Set a max value for the scale\n",
    "    figsize=(6, 10)\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lvl2.5_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23a70c58-ed41-4ce2-a80b-ade11cda7a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_knn50k10_merged'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e8c9fe-5ee6-4230-abc5-bb17d6380dce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the replacement dictionary\n",
    "replacement_dict = {\n",
    "    'Crevicular':'TAE',\n",
    "    'Plasma-Fib CT' : 'Im-Str',\n",
    "    'Fibrous CT': 'FibCT',\n",
    "    'Lymphoid' : 'T-B-APC',\n",
    "}\n",
    "\n",
    "# Replace values in the 'niche_knn30k11' column\n",
    "adata.obs['niche_knn50k10_merged'] = adata.obs['niche_knn50k10_merged'].replace(replacement_dict)\n",
    "\n",
    "# Print unique values after replacement\n",
    "print(adata.obs['niche_knn50k10_merged'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95062e5e-49c3-4e53-b852-9d9efe21c239",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2.5'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a768d0-8da1-4cf3-bcee-fe3c06fc5805",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the replacement dictionary\n",
    "replacement_dict = {\n",
    "    'Fib':'Fibroblast',\n",
    "    'mAPC' : 'APC',\n",
    "    'Crev': 'TAE',\n",
    "}\n",
    "\n",
    "# Replace values in the 'niche_knn30k11' column\n",
    "adata.obs['Lvl2.5'] = adata.obs['Lvl2.5'].replace(replacement_dict)\n",
    "\n",
    "# Print unique values after replacement\n",
    "print(adata.obs['Lvl2.5'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d426854-37c6-4614-bb1d-4dff528e0e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures fonts are embedded as text, not outlines\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Keratin', 'Spinous', 'Basal', 'TAE', 'APC', 'Mast', 'Fibroblast', 'VEC', 'SMC', 'Plasma', 'B', 'T']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'CNFN', 'EHF', 'AQP3', \n",
    "    'KRT19', 'SDC1', 'KRT5', 'COL17A1', \n",
    "    'ODAM',   'HLA-DRB5', 'MS4A6A', 'CD14', 'CPA3', 'MS4A2',  'VCAN', 'COL16A1',\n",
    "    'CD34', 'VWF', 'MYH11', 'ACTA2', 'PRDM1', 'MZB1',\n",
    "    'BANK1', 'TRAC', 'CD3E'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl2.5'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl2.5'] = adata_subset.obs['Lvl2.5'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl2.5',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8, # Set a max value for the scale\n",
    "    figsize=(6, 7)\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lvl2.5_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "231e67e5-61fb-4caf-9d44-6fcbcb19a345",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c893659-ea4a-491b-a573-03f1143e4bd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import ranksums\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# Extract indices for Crevicular and Oral groups\n",
    "crevicular_indices = adata.obs['Lvl2'] == 'Crevicular Epi'\n",
    "oral_indices = adata.obs['Lvl2'] == 'Oral Epi'\n",
    "\n",
    "# Extract expression data for Crevicular and Oral groups\n",
    "crev_expr = adata[crevicular_indices, :].X.toarray()\n",
    "oral_expr = adata[oral_indices, :].X.toarray()\n",
    "\n",
    "# Initialize lists to store results\n",
    "gene_names = []  # List for gene names\n",
    "log2fc = []      # List for log2 fold changes\n",
    "p_values = []     # List for p-values\n",
    "\n",
    "# Assuming the gene names are stored in the column headers of crev_expr and oral_expr\n",
    "# You can adjust this depending on how your gene names are stored in your data\n",
    "gene_names = adata.var_names.tolist()\n",
    "\n",
    "# Perform Wilcoxon Rank-Sum test for each gene\n",
    "for gene in range(crev_expr.shape[1]):  # Iterate over genes (columns)\n",
    "    crev_gene_expression = crev_expr[:, gene]  # Expression data for gene in Crevicular Epi\n",
    "    oral_gene_expression = oral_expr[:, gene]  # Expression data for gene in Oral Epi\n",
    "\n",
    "    # Wilcoxon Rank-Sum Test (Mann-Whitney U Test)\n",
    "    stat, p_val = ranksums(crev_gene_expression, oral_gene_expression)\n",
    "\n",
    "    # Log2 Fold Change (mean of Crevicular Epi / Oral Epi)\n",
    "    mean_crev = np.mean(crev_gene_expression)\n",
    "    mean_oral = np.mean(oral_gene_expression)\n",
    "    log2_fold_change = np.log2(mean_crev / mean_oral)  # Crevicular Epi vs Oral Epi fold change\n",
    "    \n",
    "    # Store results\n",
    "    log2fc.append(log2_fold_change)\n",
    "    p_values.append(p_val)\n",
    "\n",
    "# Step 3: Adjust p-values for multiple testing using Benjamini-Hochberg correction\n",
    "_, p_values_adj, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "# Step 4: Create a DataFrame with the results\n",
    "dge_results = pd.DataFrame({\n",
    "    'Gene Name': gene_names,\n",
    "    'Log2 Fold Change (Crev vs Oral)': log2fc,\n",
    "    'P-Value': p_values,\n",
    "    'Adjusted P-Value (FDR)': p_values_adj\n",
    "})\n",
    "\n",
    "# Step 5: Display the top significant genes (optional, based on FDR)\n",
    "# You can filter based on your preferred significance threshold (e.g., FDR < 0.05)\n",
    "significant_genes = dge_results[dge_results['Adjusted P-Value (FDR)'] < 0.01]\n",
    "\n",
    "# Display results\n",
    "print(dge_results.head())  # Show the first few results\n",
    "print(f\"Number of significant genes (FDR < 0.01): {len(significant_genes)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bc2ecaa-092d-4a25-9a91-14644c472262",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming dge_results is the dataframe containing statistics results (with 'pvalue' column)\n",
    "pval_0 = dge_results[dge_results['Adjusted P-Value (FDR)'] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dde6040a-1ba2-416b-ad34-e348900c47fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a jittered list of very small p-values around 1e-300\n",
    "pval_jitter = np.round(np.random.normal(loc=295, scale=5, size=pval_0.shape[0]), 0)\n",
    "pval_jitter = 1 * 10 ** -pval_jitter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1d737d-2409-44c6-a0d3-9f91580feb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "dge_results.loc[dge_results['Adjusted P-Value (FDR)'] == 0, 'Adjusted P-Value (FDR)'] = pval_jitter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1568b46-677f-4ee4-9db3-0ae6c358a6e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "dge_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dc822de-178a-443c-9fab-c752b18881f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for rows with fold change > 2 or < -2\n",
    "filtered_genes = most_significant[(most_significant['Log2 Fold Change (Crev vs Oral)'] > 2) | \n",
    "                                  (most_significant['Log2 Fold Change (Crev vs Oral)'] < -2)]\n",
    "\n",
    "print(\"\\nFiltered Most Significant Genes:\")\n",
    "print(filtered_genes[['Gene Name', 'Log2 Fold Change (Crev vs Oral)', 'Adjusted P-Value (FDR)']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "176fd3e7-ed82-4134-8c30-5a7c72f100f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the row corresponding to the 'ODAM' gene\n",
    "odam_row = dge_results.loc[dge_results['Gene Name'] == 'ODAM']\n",
    "print(odam_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60fb67b9-4486-49b2-ab53-eb13fa3a9ec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install adjustText"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2686f62c-5b4f-4b6b-bb25-d0f94e6b7d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from brokenaxes import brokenaxes\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from adjustText import adjust_text\n",
    "\n",
    "# List of genes you want to annotate\n",
    "genes_of_interest = ['ICAM1', 'LAMC2', \"ODAM\", \"FGFBP1\", \"ODAPH\", \"SAA2\", \"CXCL2\", \"SAA1\", \"CXCL1\", \"CSF3\", \"ADAM28\", \"SERPINB2\", \"CXCL14\", \"KLK11\", \"CFHR3\", \"C3\", \"CXCL6\"]\n",
    "\n",
    "# Define your data (replace these with your actual data)\n",
    "log2fc_values = dge_results['Log2 Fold Change (Crev vs Oral)']\n",
    "pvalues = dge_results['Adjusted P-Value (FDR)']\n",
    "gene_names = dge_results['Gene Name']\n",
    "\n",
    "# Transform p-values to -log10 scale and handle 0-values\n",
    "log10_pvalues = -np.log10(np.maximum(pvalues, 1e-300))\n",
    "\n",
    "# Create a figure with broken y-axis, starting at ymin=50\n",
    "fig = plt.figure(figsize=(10, 8))\n",
    "bax = brokenaxes(ylims=((-1, 85), (280, 310)), hspace=0.02)\n",
    "\n",
    "# Define colors and annotations\n",
    "colors = []\n",
    "texts = []\n",
    "for i, (log2fc, log10_pval, gene_name) in enumerate(zip(log2fc_values, log10_pvalues, gene_names)):\n",
    "    # Determine color\n",
    "    if log2fc < -2:\n",
    "        colors.append('blue')\n",
    "    elif log2fc > 2:\n",
    "        colors.append('red')\n",
    "    else:\n",
    "        colors.append('grey')\n",
    "\n",
    "    # Add text annotations only for genes of interest\n",
    "    if gene_name in genes_of_interest:\n",
    "        # Determine which axes segment the point belongs to\n",
    "        for ax in bax.axs:\n",
    "            xlim, ylim = ax.get_xlim(), ax.get_ylim()\n",
    "            if ylim[0] < log10_pval < ylim[1]:\n",
    "                # Add annotation to the correct sub-plot\n",
    "                texts.append(ax.text(log2fc, log10_pval, gene_name, fontsize=9, \n",
    "                                     alpha=0.9, color='black', weight='bold', ha='right'))\n",
    "                break\n",
    "\n",
    "# Scatter plot\n",
    "bax.scatter(log2fc_values, log10_pvalues, c=colors, alpha=0.7)\n",
    "\n",
    "# Adjust text to avoid overlaps\n",
    "adjust_text(texts, only_move={'points': 'xy', 'text': 'xy'}, \n",
    "            arrowprops=dict(arrowstyle=\"->\", color='black'))\n",
    "\n",
    "# Labels and Title\n",
    "bax.set_xlabel('Log2 Fold Change', fontsize=14, color='black')\n",
    "bax.set_ylabel('-Log10(Adjusted P-Value)', fontsize=14)\n",
    "bax.fig.suptitle('Volcano Plot: Log2 Fold Change vs Adjusted P-Value (FDR)', fontsize=16)\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure3/OralvsCrevicular_Volcano.svg', format='svg')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91b9d1ed-a090-42ac-80d4-cabb140b74ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures fonts are embedded as text, not outlines\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Ep.B.1',  'Ep.B.2', 'Ep.B-PB', 'Ep.PB.1', 'Ep.PB.2', 'Ep.PB.3', 'Ep.prol', 'Ep.Sp', 'Ep.K', \n",
    "                        'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "   'KRT5', 'COL17A1', 'DST', 'ITGB4', 'CXCL14', 'C1R', 'IL1R2',\n",
    "      'CAVIN1', 'PDPN', 'BASP1',\n",
    "'SLC26A2', 'MYC', 'CFH', 'PTN',\n",
    "'MKI67', 'CDK1', 'AQP3', 'KRT19', 'FGFBP1', 'S100A16', 'LY6D', 'SDC1', \n",
    "  'CYP4B1', 'MAMDC2', 'SERPINB3', 'NOD2',\n",
    "'S100A16', 'SLPI', 'IL1RN', 'ANXA1',\n",
    "'C15orf48', 'IL36A','CNFN', 'IL36G',\n",
    " 'SERPINB2', 'EHF',  'ERBB2', 'ODAM', 'ODAPH', 'LAMC2', 'CXCL1', 'CXCL2', 'CXCL6', 'SLPI',\n",
    "'IL1A', 'IL36G', 'C15orf48', 'ANXA1', 'DNASE1L3',\n",
    "'COL17A1', 'DST', 'C1R', 'ITGB4', 'SAA1', 'SAA2',\n",
    "'CYP4B1', 'CFH', 'EHF',  'CLEC7A', 'CFHR3'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl5'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl5'] = adata_subset.obs['Lvl5'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl5',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8,                       # Set a max value for the scale\n",
    "    figsize=(20, 4)\n",
    "\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=False)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8,                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl5_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93d2481b-ea46-4ae2-9036-056a3b8569bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.style.use('default')  # Set the background to black\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "'ODAM', 'ODAPH', 'LAMC2', 'CXCL1', 'CXCL2', 'CXCL6', 'SLPI',\n",
    "'IL1A', 'IL36G', 'C15orf48', 'ANXA1', 'DNASE1L3',\n",
    "'COL17A1', 'DST', 'C1R', 'ITGB4', 'SAA1', 'SAA2',\n",
    "'CYP4B1', 'CFH', 'EHF', 'KRT19', 'CLEC7A', 'CFHR3'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "plt.rcParams.update({'font.size': 14, 'text.color': 'black', 'font.weight': 'normal', 'axes.spines.right' : False, 'axes.spines.top': False, 'axes.linewidth': 2})\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Crev_Epi_Lvl4_dotplot.svg'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18a17a9-1a63-4228-b401-8eb0c04919ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Duplicate the Lvl4 column\n",
    "adata.obs['Epithelial_layers'] = adata.obs['Lvl4'].copy()\n",
    "\n",
    "# Step 1: Inspect and ensure unique values\n",
    "unique_categories = adata.obs['Epithelial_layers'].unique()\n",
    "print(\"Original Categories in 'Lvl4':\", unique_categories)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c1e5e1-5c14-461a-b73a-e6123d826f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Duplicate the column\n",
    "adata.obs['Epithelial_layers'] = adata.obs['Lvl4'].copy()\n",
    "\n",
    "# Step 2: Convert the new column to categorical type\n",
    "adata.obs['Epithelial_layers'] = adata.obs['Epithelial_layers'].astype('category')\n",
    "\n",
    "# Step 3: Rename categories using a mapping\n",
    "category_mapping = {\n",
    "    'Ep.Cr.B': 'Ep.Cr.lower',\n",
    "    'Ep.Cr.PB': 'Ep.Cr.lower',\n",
    "    'Ep.Cr.Sp.1': 'Ep.Cr.upper',\n",
    "    'Ep.Cr.Sp.2': 'Ep.Cr.upper',\n",
    "    'Ep.B.1': 'Ep.Oral.lower',\n",
    "    'Ep.B.2': 'Ep.Oral.lower',\n",
    "    'Ep.B.3': 'Ep.Oral.lower',\n",
    "    'Ep.B.4': 'Ep.Oral.lower',\n",
    "    'Ep.B-PB': 'Ep.Oral.lower',\n",
    "    'Ep.PB.1': 'Ep.Oral.lower',\n",
    "    'Ep.PB.2': 'Ep.Oral.lower',\n",
    "    'Ep.PB.3': 'Ep.Oral.lower',\n",
    "    'Ep.PB.4': 'Ep.Oral.lower',\n",
    "    'Ep.prol': 'Ep.Oral.lower',\n",
    "    'Ep.Sp': 'Ep.Oral.upper',\n",
    "    'Ep.K': 'Ep.Oral.upper'\n",
    "\n",
    "}\n",
    "\n",
    "# Map the values manually\n",
    "adata.obs['Epithelial_layers'] = adata.obs['Epithelial_layers'].map(category_mapping).fillna(adata.obs['Epithelial_layers'])\n",
    "\n",
    "# Step 4: Remove unused categories\n",
    "adata.obs['Epithelial_layers'] = adata.obs['Epithelial_layers'].astype('category')\n",
    "\n",
    "# Step 5: Verify the changes\n",
    "print(\"Updated Categories in 'Epithelial_layers':\", adata.obs['Epithelial_layers'].cat.categories)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "342340c6-f9e7-4bdf-9a64-178f06683972",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Epithelial_layers'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b10ce4c1-8f13-488d-bcbe-3a845f9cd7b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs['Lvl4'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a371651-e5dd-447e-92ca-7fd3fa4ee63c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 1: Subset the adata object to only the specific clusters and create a copy\n",
    "desired_clusters = ['Ep.Oral.upper', 'Ep.Oral.lower', 'Ep.Cr.upper', 'Ep.Cr.lower']\n",
    "\n",
    "adata_subset = adata[adata.obs['Epithelial_layers'].isin(desired_clusters)].copy()\n",
    "\n",
    "# Ensure 'Lvl4' is categorical\n",
    "adata_subset.obs['Epithelial_layers'] = adata_subset.obs['Epithelial_layers'].astype('category')\n",
    "\n",
    "# Step 2: Check if the data is already log-transformed, otherwise transform it\n",
    "if 'log1p' not in adata_subset.uns_keys():\n",
    "    sc.pp.log1p(adata_subset)  # Apply log transformation if not already done\n",
    "\n",
    "# Manually set the base in log1p if missing\n",
    "if 'log1p' in adata_subset.uns_keys() and 'base' not in adata_subset.uns['log1p']:\n",
    "    adata_subset.uns['log1p']['base'] = 2.718281828459045  # Default to e\n",
    "\n",
    "# Step 3: Perform differential expression analysis for the selected clusters\n",
    "sc.tl.rank_genes_groups(adata_subset, 'Epithelial_layers', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for the selected clusters\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata_subset.obs['Epithelial_layers'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata_subset.uns['rank_genes_groups']['names'][cluster]\n",
    "\n",
    "# Step 5: Generate the dendrogram to get cluster order\n",
    "sc.tl.dendrogram(adata_subset, groupby='Epithelial_layers')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata_subset.uns['dendrogram_Epithelial_layers']['categories_ordered']\n",
    "\n",
    "# Step 7: Ensure each cluster has at least 3 unique genes\n",
    "unique_genes = set()\n",
    "genes_for_plot = []\n",
    "\n",
    "for cluster in cluster_order:\n",
    "    if cluster in top_genes_per_cluster:\n",
    "        cluster_genes = top_genes_per_cluster[cluster]\n",
    "        genes_for_this_cluster = []\n",
    "        for gene in cluster_genes:\n",
    "            if len(genes_for_this_cluster) >= 10:\n",
    "                break\n",
    "            if gene not in unique_genes:\n",
    "                unique_genes.add(gene)\n",
    "                genes_for_this_cluster.append(gene)\n",
    "        genes_for_plot.extend(genes_for_this_cluster)\n",
    "\n",
    "# Step 8: Plot the matrixplot with the unique genes\n",
    "sc.pl.dotplot(adata_subset, var_names=genes_for_plot, groupby='Epithelial_layers', \n",
    "                 dendrogram=True, use_raw=False, cmap=\"vlag\", \n",
    "                 standard_scale='var', swap_axes=True)\n",
    "print(top_genes_per_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1922218b-7166-40dd-b5c5-e5b10533493e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 1: Subset the adata object to only the specific clusters and create a copy\n",
    "desired_clusters = ['Ep.B.1',  'Ep.B.2', 'Ep.B-PB', 'Ep.PB.1', 'Ep.PB.2', 'Ep.PB.3', 'Ep.prol', 'Ep.Sp', 'Ep.K', \n",
    "                        'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB']\n",
    "\n",
    "adata_subset = adata[adata.obs['Lvl5'].isin(desired_clusters)].copy()\n",
    "\n",
    "# Ensure 'Lvl4' is categorical\n",
    "adata_subset.obs['Lvl5'] = adata_subset.obs['Lvl5'].astype('category')\n",
    "\n",
    "# Step 2: Check if the data is already log-transformed, otherwise transform it\n",
    "if 'log1p' not in adata_subset.uns_keys():\n",
    "    sc.pp.log1p(adata_subset)  # Apply log transformation if not already done\n",
    "\n",
    "# Manually set the base in log1p if missing\n",
    "if 'log1p' in adata_subset.uns_keys() and 'base' not in adata_subset.uns['log1p']:\n",
    "    adata_subset.uns['log1p']['base'] = 2.718281828459045  # Default to e\n",
    "\n",
    "# Step 3: Perform differential expression analysis for the selected clusters\n",
    "sc.tl.rank_genes_groups(adata_subset, 'Lvl5', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for the selected clusters\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata_subset.obs['Lvl5'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata_subset.uns['rank_genes_groups']['names'][cluster]\n",
    "\n",
    "# Step 5: Generate the dendrogram to get cluster order\n",
    "sc.tl.dendrogram(adata_subset, groupby='Lvl5')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata_subset.uns['dendrogram_Lvl5']['categories_ordered']\n",
    "\n",
    "# Step 7: Ensure each cluster has at least 3 unique genes\n",
    "unique_genes = set()\n",
    "genes_for_plot = []\n",
    "\n",
    "for cluster in cluster_order:\n",
    "    if cluster in top_genes_per_cluster:\n",
    "        cluster_genes = top_genes_per_cluster[cluster]\n",
    "        genes_for_this_cluster = []\n",
    "        for gene in cluster_genes:\n",
    "            if len(genes_for_this_cluster) >= 5:\n",
    "                break\n",
    "            if gene not in unique_genes:\n",
    "                unique_genes.add(gene)\n",
    "                genes_for_this_cluster.append(gene)\n",
    "        genes_for_plot.extend(genes_for_this_cluster)\n",
    "\n",
    "# Step 8: Plot the matrixplot with the unique genes\n",
    "sc.pl.dotplot(adata_subset, var_names=genes_for_plot, groupby='Lvl5', \n",
    "                 dendrogram=True, use_raw=False, cmap=\"vlag\", \n",
    "                 standard_scale='var', swap_axes=True)\n",
    "print(top_genes_per_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe4f719-71f9-40db-afd1-0ba583c674c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Ensure 'PatientID_Atlas' is a string type\n",
    "adata.obs['Lvl4'] = adata.obs['Lvl4'].astype(str)\n",
    "\n",
    "# Define the replacement dictionary\n",
    "replacement_dict = {\n",
    "'Ep.PB.1' : 'Ep.PB',\n",
    "'Ep.PB.2' : 'Ep.PB',\n",
    "'Ep.PB.3'\t: 'Ep.PB',\n",
    "'Ep.PB.4' : 'Ep.PB',\n",
    "'Ep.B.1' : 'Ep.B',\n",
    "'Ep.B.2' : 'Ep.B',\n",
    "'Ep.B.3'\t: 'Ep.B',\n",
    "'Ep.B.4' : 'Ep.B',\n",
    "}\n",
    "\n",
    "# Replace only the specified categories in 'T_leiden2'\n",
    "adata.obs['Lvl4'] = adata.obs['Lvl4'].astype(str)  # Ensure it's string\n",
    "adata.obs['Lvl4'] = adata.obs['Lvl4'].replace(replacement_dict)\n",
    "\n",
    "# Check the unique values in 'T_leiden2' to confirm\n",
    "print(adata.obs['Lvl4'].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1f22443-fcc3-40de-87cc-334984a94016",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 1: Subset the adata object to only the specific clusters and create a copy\n",
    "desired_clusters = ['Ep.B', 'Ep.B-PB', 'Ep.PB', 'Ep.prol', 'Ep.Sp', 'Ep.K', \n",
    "                        'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB']\n",
    "\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(desired_clusters)].copy()\n",
    "\n",
    "# Ensure 'Lvl4' is categorical\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype('category')\n",
    "\n",
    "# Step 2: Check if the data is already log-transformed, otherwise transform it\n",
    "if 'log1p' not in adata_subset.uns_keys():\n",
    "    sc.pp.log1p(adata_subset)  # Apply log transformation if not already done\n",
    "\n",
    "# Manually set the base in log1p if missing\n",
    "if 'log1p' in adata_subset.uns_keys() and 'base' not in adata_subset.uns['log1p']:\n",
    "    adata_subset.uns['log1p']['base'] = 2.718281828459045  # Default to e\n",
    "\n",
    "# Step 3: Perform differential expression analysis for the selected clusters\n",
    "sc.tl.rank_genes_groups(adata_subset, 'Lvl4', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for the selected clusters\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata_subset.obs['Lvl4'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata_subset.uns['rank_genes_groups']['names'][cluster]\n",
    "\n",
    "# Step 5: Generate the dendrogram to get cluster order\n",
    "sc.tl.dendrogram(adata_subset, groupby='Lvl4')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata_subset.uns['dendrogram_Lvl4']['categories_ordered']\n",
    "\n",
    "# Step 7: Ensure each cluster has at least 3 unique genes\n",
    "unique_genes = set()\n",
    "genes_for_plot = []\n",
    "\n",
    "for cluster in cluster_order:\n",
    "    if cluster in top_genes_per_cluster:\n",
    "        cluster_genes = top_genes_per_cluster[cluster]\n",
    "        genes_for_this_cluster = []\n",
    "        for gene in cluster_genes:\n",
    "            if len(genes_for_this_cluster) >= 5:\n",
    "                break\n",
    "            if gene not in unique_genes:\n",
    "                unique_genes.add(gene)\n",
    "                genes_for_this_cluster.append(gene)\n",
    "        genes_for_plot.extend(genes_for_this_cluster)\n",
    "\n",
    "# Step 8: Plot the matrixplot with the unique genes\n",
    "sc.pl.dotplot(adata_subset, var_names=genes_for_plot, groupby='Lvl4', \n",
    "                 dendrogram=True, use_raw=False, cmap=\"vlag\", \n",
    "                 standard_scale='var', swap_axes=True)\n",
    "print(top_genes_per_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d1abdf-903e-491d-81ae-e093146deea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.style.use('default')  # Set the background to black\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Ep.Oral.lower', 'Ep.Cr.lower', 'Ep.Cr.upper', 'Ep.Oral.upper']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "'S100A16','KRT5', 'DST', 'MYC', 'EGFR', 'COL17A1',  'CXCL14',\n",
    "'FGFBP1', 'CFHR3', 'C1R', 'SAA1', 'SAA2',\n",
    "'ADAM28', 'CFH', 'ODAM', 'ODAPH', 'CXCL1', 'CXCL6',\n",
    "'CSF3', 'DNASE1L3', 'LAMC2', 'IL1A', 'ANXA1', 'SERPINB2', 'CNFN', 'ELF5', 'IL36G', 'IL36A', 'C15orf48'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Epithelial_layers'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Epithelial_layers'] = adata_subset.obs['Epithelial_layers'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Epithelial_layers',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "plt.rcParams.update({'font.size': 14, 'text.color': 'black', 'font.weight': 'normal', 'axes.spines.right' : False, 'axes.spines.top': False, 'axes.linewidth': 2})\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Or_Crev_layers_dotplot.svg'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d540c9b7-1e2b-47f7-98b6-80ebd3fe269c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc457a62-02f7-452e-90f0-30e51c61e5e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import os\n",
    "\n",
    "\n",
    "rotation_angle = 30  # Define rotation for this sample\n",
    "\n",
    "# Specify the directory where you want to save the PDF\n",
    "save_directory = '/data/vasileiosionat2/Xenium/Figures/'\n",
    "pdf_filename = os.path.join(save_directory, f'all_clusters_4rows.pdf')\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV140A_A']\n",
    "\n",
    "# Extract unique clusters in the 'Lvl4' column matching the specified categories\n",
    "unique_clusters = adata_sample.obs.loc[\n",
    "    adata_sample.obs['Lvl5'].isin([\n",
    "'Ep.B.1',  'Ep.B.2', 'Ep.B-PB', 'Ep.PB.1', 'Ep.PB.2', 'Ep.PB.3', 'Ep.prol', 'Ep.Sp', 'Ep.K', 'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB', \n",
    "    'Fib.3', 'VEC.2', 'VEC.5', 'cDC2', 'Lang', 'MyoF', 'Mac']), \n",
    "    'Lvl4'\n",
    "].unique()\n",
    "\n",
    "# Print the unique clusters\n",
    "print(unique_clusters)\n",
    "\n",
    "\n",
    "\n",
    "# Reorder clusters (if a custom order is provided, replace `sorted(unique_clusters)`)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define the number of rows and columns\n",
    "num_rows = 5\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns based on clusters and rows\n",
    "\n",
    "# Create a PdfPages object to save the plots\n",
    "with PdfPages(pdf_filename) as pdf:\n",
    "    # Create a figure for the sample\n",
    "    fig, axes = plt.subplots(num_rows, num_cols, figsize=(26, 26))\n",
    "    fig.patch.set_facecolor('white')\n",
    "\n",
    "    # Flatten the axes array for easier indexing\n",
    "    axes = axes.flatten()\n",
    "\n",
    "    # Get the coordinates for the sample\n",
    "    x_coords = adata_sample.obs['x_centroid']\n",
    "    y_coords = adata_sample.obs['y_centroid']\n",
    "    \n",
    "    # Apply rotation\n",
    "    angle = np.deg2rad(rotation_angle)\n",
    "    new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "    new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "    # Determine aspect ratio\n",
    "    x_range = new_x_coords.max() - new_x_coords.min()\n",
    "    y_range = new_y_coords.max() - new_y_coords.min()\n",
    "    aspect_ratio = x_range / y_range\n",
    "\n",
    "       # Loop over each unique cluster in the ordered list\n",
    "    for idx, cluster in enumerate(ordered_clusters):\n",
    "        ax = axes[idx]\n",
    "\n",
    "        # Set white background for the subplot\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Remove the outline\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        # Scatter plot for the grey dot203Bs (all other clusters)\n",
    "        ax.scatter(\n",
    "            x=new_x_coords[adata_sample.obs['Lvl5'] != cluster],\n",
    "            y=new_y_coords[adata_sample.obs['Lvl5'] != cluster],\n",
    "            c='#C0C0C0', \n",
    "            s=3  # Adjust dot size\n",
    "        )\n",
    "\n",
    "        # Scatter plot for the red dots (the current cluster)\n",
    "        ax.scatter(\n",
    "            x=new_x_coords[adata_sample.obs['Lvl5'] == cluster],\n",
    "            y=new_y_coords[adata_sample.obs['Lvl5'] == cluster],\n",
    "            c='red', \n",
    "            s=9  # Adjust dot size\n",
    "        )\n",
    "\n",
    "        # Set aspect ratio for each subplot\n",
    "        ax.set_aspect(aspect_ratio)\n",
    "\n",
    "        # Add the cluster name below the plot\n",
    "        ax.text(\n",
    "            0.5, 0.02, f'{cluster}', \n",
    "            horizontalalignment='center', \n",
    "            verticalalignment='center', \n",
    "            transform=ax.transAxes, \n",
    "            color='black', fontsize=20, weight='bold'\n",
    "        )\n",
    "\n",
    "        # Remove grids\n",
    "        ax.grid(False)\n",
    "\n",
    "        # Remove ticks and their labels\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "    # Turn off empty subplots if there are any\n",
    "    for ax in axes[len(ordered_clusters):]:\n",
    "        ax.set_visible(False)\n",
    "\n",
    "    # Define the directory and file name for the TIFF file\n",
    "    tiff_filename = os.path.join(save_directory, f'all_clusters_4rows.tiff')\n",
    "\n",
    "    # Adjust the spacing between subplots\n",
    "    plt.subplots_adjust(\n",
    "        left=0.05,    # Space from left edge\n",
    "        right=0.95,   # Space from right edge\n",
    "        top=0.95,     # Space from top edge\n",
    "        bottom=0.05,  # Space from bottom edge\n",
    "        wspace=0.1,   # Width space between columns\n",
    "        hspace=0.05    # Height space between rows\n",
    "    )\n",
    "\n",
    "    # Add a title to the overall figure\n",
    "    plt.suptitle(f'All Clusters in s2r2_HV184', color='white', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "    # Save the current figure as a TIFF file\n",
    "    #fig.savefig(tiff_filename, dpi=300, format='tiff')\n",
    "\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c59282-fcc8-43a4-a0f4-dc840ddd488e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_cc14'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef910306-f010-4617-a618-a707d0539763",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import os\n",
    "\n",
    "\n",
    "rotation_angle = 30  # Define rotation for this sample\n",
    "\n",
    "# Specify the directory where you want to save the PDF\n",
    "save_directory = '/data/vasileiosionat2/Xenium/Figures/'\n",
    "pdf_filename = os.path.join(save_directory, f'all_clusters_4rows.pdf')\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV140A_A']\n",
    "\n",
    "# Extract unique clusters in the 'Lvl4' column matching the specified categories\n",
    "unique_clusters = adata_sample.obs.loc[\n",
    "    adata_sample.obs['niche_cc14'].isin([\n",
    "'Lymphoid',\n",
    " 'Plasma',\n",
    " 'Subepi CT (crevicular)',\n",
    " 'Lymphoid-CT interface',\n",
    " 'Subepi CT (oral)',\n",
    " 'Plasma-CT interface',\n",
    " 'BV',\n",
    " 'Fibrous CT',\n",
    " 'Myeloid',\n",
    " 'Parabasal',\n",
    " 'Basal',\n",
    " 'Spinous',\n",
    " 'Keratin',\n",
    " 'Crevicular']), \n",
    "    'niche_cc14'\n",
    "].unique()\n",
    "\n",
    "# Print the unique clusters\n",
    "print(unique_clusters)\n",
    "\n",
    "\n",
    "\n",
    "# Reorder clusters (if a custom order is provided, replace `sorted(unique_clusters)`)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define the number of rows and columns\n",
    "num_rows = 8\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns based on clusters and rows\n",
    "\n",
    "# Create a PdfPages object to save the plots\n",
    "with PdfPages(pdf_filename) as pdf:\n",
    "    # Create a figure for the sample\n",
    "    fig, axes = plt.subplots(num_rows, num_cols, figsize=(22, 56))\n",
    "    fig.patch.set_facecolor('white')\n",
    "\n",
    "    # Flatten the axes array for easier indexing\n",
    "    axes = axes.flatten()\n",
    "\n",
    "    # Get the coordinates for the sample\n",
    "    x_coords = adata_sample.obs['x_centroid']\n",
    "    y_coords = adata_sample.obs['y_centroid']\n",
    "    \n",
    "    # Apply rotation\n",
    "    angle = np.deg2rad(rotation_angle)\n",
    "    new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "    new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "    # Determine aspect ratio\n",
    "    x_range = new_x_coords.max() - new_x_coords.min()\n",
    "    y_range = new_y_coords.max() - new_y_coords.min()\n",
    "    aspect_ratio = x_range / y_range\n",
    "\n",
    "       # Loop over each unique cluster in the ordered list\n",
    "    for idx, cluster in enumerate(ordered_clusters):\n",
    "        ax = axes[idx]\n",
    "\n",
    "        # Set white background for the subplot\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Remove the outline\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        # Scatter plot for the grey dot203Bs (all other clusters)\n",
    "        ax.scatter(\n",
    "            x=new_x_coords[adata_sample.obs['niche_cc14'] != cluster],\n",
    "            y=new_y_coords[adata_sample.obs['niche_cc14'] != cluster],\n",
    "            c='#C0C0C0', \n",
    "            s=3  # Adjust dot size\n",
    "        )\n",
    "\n",
    "        # Scatter plot for the red dots (the current cluster)\n",
    "        ax.scatter(\n",
    "            x=new_x_coords[adata_sample.obs['niche_cc14'] == cluster],\n",
    "            y=new_y_coords[adata_sample.obs['niche_cc14'] == cluster],\n",
    "            c='red', \n",
    "            s=9  # Adjust dot size\n",
    "        )\n",
    "\n",
    "        # Set aspect ratio for each subplot\n",
    "        ax.set_aspect(aspect_ratio)\n",
    "\n",
    "        # Add the cluster name below the plot\n",
    "        ax.text(\n",
    "            0.5, 0.02, f'{cluster}', \n",
    "            horizontalalignment='center', \n",
    "            verticalalignment='center', \n",
    "            transform=ax.transAxes, \n",
    "            color='black', fontsize=20, weight='bold'\n",
    "        )\n",
    "\n",
    "        # Remove grids\n",
    "        ax.grid(False)\n",
    "\n",
    "        # Remove ticks and their labels\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "    # Turn off empty subplots if there are any\n",
    "    for ax in axes[len(ordered_clusters):]:\n",
    "        ax.set_visible(False)\n",
    "\n",
    "    # Define the directory and file name for the TIFF file\n",
    "    tiff_filename = os.path.join(save_directory, f'all_clusters_4rows.tiff')\n",
    "\n",
    "    # Adjust the spacing between subplots\n",
    "    plt.subplots_adjust(\n",
    "        left=0.05,    # Space from left edge\n",
    "        right=0.95,   # Space from right edge\n",
    "        top=0.95,     # Space from top edge\n",
    "        bottom=0.05,  # Space from bottom edge\n",
    "        wspace=0.1,   # Width space between columns\n",
    "        hspace=0.05    # Height space between rows\n",
    "    )\n",
    "\n",
    "    # Add a title to the overall figure\n",
    "    plt.suptitle(f'All Clusters in s2r2_HV184', color='white', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "    # Save the current figure as a TIFF file\n",
    "    #fig.savefig(tiff_filename, dpi=300, format='tiff')\n",
    "\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b4e6f8-5de6-4f8b-ae70-67c050d3f963",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming 'adata.obs' is available\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['niche_cc14'] = df['niche_cc14'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_cc14', 'Lvl4']).size().unstack(fill_value=0)\n",
    "\n",
    "# Identify clusters to remove\n",
    "clusters_to_remove = df['Lvl4'].str.contains('Mix') | df['Lvl4'].str.contains('prol') | df['Lvl4'].str.contains('B.4') | df['Lvl4'].str.startswith('Pl.3')  | df['Lvl4'].str.startswith('Pl.4')  | df['Lvl4'].str.startswith('Ep.B.3') | df['Lvl4'].str.startswith('Oth.') | df['Lvl4'].str.startswith('Unclea')\n",
    "\n",
    "# Get the unique clusters to be removed\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'Lvl4'].unique()\n",
    "\n",
    "# Filter out these clusters from the plot\n",
    "niche_groups_filtered = niche_groups.drop(columns=clusters_to_remove, errors='ignore')\n",
    "\n",
    "# Transpose the DataFrame to swap axes\n",
    "niche_groups_filtered = niche_groups_filtered.T\n",
    "\n",
    "# Define custom order for the x-axis (after transposing, so this is the niche clusters now)\n",
    "custom_order = ['Basal', 'Parabasal', 'Spinous', 'Keratin', 'Crevicular', 'BV', \n",
    "                'Fibrous CT', 'Subepi CT (oral)', 'Subepi CT (crevicular)', \n",
    "                'Myeloid', 'Lymphoid', 'Plasma', 'Lymphoid-CT interface', \n",
    "                'Plasma-CT interface']\n",
    "\n",
    "# Reindex columns based on the custom order\n",
    "niche_groups_filtered = niche_groups_filtered.reindex(columns=custom_order)\n",
    "\n",
    "# Plot heatmap with enhanced colormap\n",
    "plt.figure(figsize=(16, 20))\n",
    "sns.heatmap(niche_groups_filtered, cmap='inferno', cbar_kws={'label': 'Count', 'ticks': [0, 5000, 10000, 15000], 'format': '%.0f'}, linewidths=0.05, linecolor='black', vmax=15000)\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche Cluster')  # Updated label after transposing\n",
    "plt.ylabel('Cell Type')  # Updated label after transposing\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Customize the color bar label\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "cbar.set_ticks([0, 5000, 10000, 15000])\n",
    "cbar.ax.set_yticklabels(['0', '5000', '10000', '>15000'])  # Set custom tick labels\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a285fef-37c8-45da-abe4-f169506894e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2bc4f7c-be50-4e11-99c4-14d8d48be3bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'plot_niches' is a DataFrame or similar data structure\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_cc14']).size().unstack(fill_value=0)\n",
    "\n",
    "# Identify clusters to remove\n",
    "clusters_to_remove =  df['Lvl4'].str.startswith('Oth.') | df['Lvl4'].str.startswith('Neuro')\n",
    "\n",
    "# Get the unique clusters to be removed\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'Lvl4'].unique()\n",
    "\n",
    "# Filter out these clusters from the plot\n",
    "niche_groups_filtered = niche_groups[~niche_groups.index.isin(clusters_to_remove)]\n",
    "\n",
    "# Specific Lvl4 clusters to show\n",
    "clusters_to_show = ['Ep.B.1',  'Ep.B.2', 'Ep.B-PB', 'Ep.PB.1', 'Ep.PB.2', 'Ep.PB.3', 'Ep.prol', 'Ep.Sp', 'Ep.K', \n",
    "                    'Ep.Cr.B', 'Ep.Cr.PB', 'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2',  'Fib.3', 'VEC.2', 'VEC.5', 'cDC2', 'Lang', 'MyoF', 'Mac', 'Mac.Neut.Mix']\n",
    "\n",
    "# Filter to include only specified clusters\n",
    "niche_groups_filtered = niche_groups_filtered.loc[clusters_to_show]\n",
    "\n",
    "# Calculate relative frequencies\n",
    "niche_groups_relative = niche_groups_filtered.div(niche_groups_filtered.sum(axis=1), axis=0)\n",
    "\n",
    "# Plot heatmap with enhanced colormap\n",
    "plt.figure(figsize=(15, 21))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(niche_groups_relative, cmap='plasma', cbar_kws={'label': ''}, linewidths=0.05, linecolor='black')\n",
    "plt.title('')\n",
    "plt.xlabel('')\n",
    "plt.ylabel('')\n",
    "plt.xticks(rotation=45, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a3d8047-584c-4820-aac2-ddf8561f8b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'adata.obs' is available and contains 'Lvl4' and 'niche_cc14'\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Step 1: Aggregate counts of niches for each Lvl4 cluster\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_cc14']).size().unstack(fill_value=0)\n",
    "\n",
    "# Step 2: Calculate the total number of cells per cluster (across all niches)\n",
    "total_cells_per_cluster = niche_groups.sum(axis=1)\n",
    "\n",
    "# Step 3: Calculate relative frequencies for each niche using the total cells per cluster\n",
    "niche_groups_relative = niche_groups.div(total_cells_per_cluster, axis=0)\n",
    "\n",
    "# Step 4: Define the niches and clusters in the desired order\n",
    "niches_to_plot = ['Basal', 'Parabasal', 'Spinous', 'Keratin', 'Crevicular', 'Subepi CT (oral)', 'Subepi CT (crevicular)']  # Replace with your desired niches\n",
    "clusters_to_show = ['Ep.B', 'Ep.prol', 'Ep.B-PB', 'Ep.PB', 'Ep.Sp', 'Ep.K', \n",
    "                    'Ep.Cr.B', 'Ep.Cr.PB', 'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Fib.3', 'VEC.2', 'cDC2', 'Lang', 'Mel']  # Replace with your clusters of interest\n",
    "\n",
    "# Step 5: Check if the niches to plot are valid and present in the dataset\n",
    "valid_niches = [niche for niche in niches_to_plot if niche in niche_groups.columns]\n",
    "print(f\"Valid niches to plot: {valid_niches}\")\n",
    "\n",
    "# Step 6: Filter to include only the valid niches for plotting\n",
    "niche_groups_filtered = niche_groups_relative[valid_niches]\n",
    "\n",
    "# Step 7: Filter Lvl4 clusters to only include the specified clusters\n",
    "niche_groups_filtered = niche_groups_filtered[niche_groups_filtered.index.isin(clusters_to_show)]\n",
    "\n",
    "# Step 8: Ensure custom order of niches and clusters\n",
    "niche_groups_filtered = niche_groups_filtered[valid_niches]  # Ensure the custom order of niches\n",
    "niche_groups_filtered = niche_groups_filtered.loc[clusters_to_show]  # Ensure the custom order of clusters\n",
    "\n",
    "# Step 9: Filter clusters if necessary (example, remove clusters that contain specific terms)\n",
    "clusters_to_remove =  df['Lvl4'].str.startswith('Oth.') | df['Lvl4'].str.startswith('Neuro')\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'Lvl4'].unique()\n",
    "\n",
    "# Filter out the unwanted clusters\n",
    "niche_groups_filtered = niche_groups_filtered[~niche_groups_filtered.index.isin(clusters_to_remove)]\n",
    "\n",
    "# Step 10: Plot the heatmap with the filtered data and custom orders\n",
    "plt.figure(figsize=(10, 14))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(niche_groups_filtered, cmap='coolwarm', cbar_kws={'label': ''}, linewidths=0.05, linecolor='black')\n",
    "plt.title('Relative abundance of clusters in Epithelial niches')\n",
    "plt.xlabel('')\n",
    "plt.ylabel('')\n",
    "plt.xticks(rotation=45, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_clusters_Epi.svg\"\n",
    "plt.savefig(output_path, format='svg')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "534636d9-4801-4d1f-b5ca-beb7159e4ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1200, 3600  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 100, 2700  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Basal': 'yellow', \n",
    "   'Parabasal' : 'yellow', \n",
    "    'Spinous': 'yellow' , \n",
    "    'Keratin': 'yellow', \n",
    "    'Crevicular': '#FF00FF', \n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV140A_A\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_cc14'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_cc14'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_cc14'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_cc14'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (2000, 2600, 280, 880),  # Lymphoid ROI\n",
    "    (2100, 2600, 2000, 2500),  # Deep ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72297b62-9ff5-46ca-9b6e-0cc5e4ee2ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2000, 2600  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 280, 880  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Ep.B': '#00FF7F', \n",
    "    'Ep.prol': '#FF69B4',\n",
    "   'Ep.PB' : '#F88379', \n",
    "   'Ep.B-PB' : '#FFFF00', \n",
    "    'Ep.Sp': '#FF00FF' , \n",
    "    'Ep.K': '#D70040', \n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"blue\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV140A_A\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Oral_Epi_Clusters_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904fb91f-3b4a-4933-969f-e2d9486cc653",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2100, 2600  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2000, 2500  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Ep.Cr.B': 'cyan', \n",
    "    'Ep.Cr.PB': '#FFBF00',\n",
    "   'Ep.Cr.Sp.1' : '#E34234', \n",
    "   'Ep.Cr.Sp.2' : '#4CBB17'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"blue\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"HV140A_A\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Crev_Epi_Clusters_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a66c846-9c1b-4552-8bd9-6162e5c095fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "xmin, xmax = 2000, 2600  # Horizontal bounds\n",
    "ymin, ymax = 280, 880  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"IL36A\", \"SERPINB3\", \"CXCL14\", \"MKI67\", \"MYC\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"IL36A\": \"red\",  \n",
    "    \"SERPINB3\": \"#0FFF50\",  \n",
    "    \"CXCL14\": \"yellow\",  \n",
    "    \"MKI67\": \"cyan\",\n",
    "    \"MYC\" : '#DA70D6'\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_oral_transcripts1_HV140A.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "018985a6-4b6f-40ef-97c4-495c7325b9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "xmin, xmax = 2000, 2600  # Horizontal bounds\n",
    "ymin, ymax = 280, 880  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"IL36G\", \"SERPINB2\", \"COL17A1\", \"FGFBP1\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"IL36G\": \"#DA70D6\",  \n",
    "    \"SERPINB2\": \"#FFFF00\",  \n",
    "    \"COL17A1\": \"#0FFF50\",  \n",
    "    \"FGFBP1\": \"#C\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Oral_transcripts2_HV140A.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee04f88-ab9c-4170-98ab-c76e2d643269",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "xmin, xmax = 2100, 2600  # Horizontal bounds\n",
    "ymin, ymax = 2000, 2500  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"ODAM\", \"CXCL1\", \"COL17A1\", \"LAMC2\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"COL17A1\": \"cyan\",  \n",
    "    \"CXCL1\": \"#FF5F1F\",  \n",
    "    \"ODAM\": \"#DA70D6\",  \n",
    "    \"LAMC2\": \"#FFFF00\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Crev_transcripts1_HV140A.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7b0430-9bc5-49eb-b77f-517fd4d826f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/HV140A_A/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "xmin, xmax = 2100, 2600  # Horizontal bounds\n",
    "ymin, ymax = 2000, 2500  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"SAA1\", \"CFHR3\", \"IL1A\", \"DNASE1L3\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"SAA1\": \"#DA70D6\",  \n",
    "    \"CFHR3\": \"red\",  \n",
    "    \"IL1A\": \"#0FFF50\",  \n",
    "    \"DNASE1L3\": \"#FFFF00\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Crev_transcripts2_HV140A.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a997e45-7edd-439e-80de-c5ecc097dbd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# Start plotting with black background\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures fonts are embedded as text, not outlines\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Ep.B', 'Ep.prol', 'Ep.B-PB', 'Ep.PB',  'Ep.Sp', 'Ep.K', 'Ep.Cr.B', 'Ep.Cr.PB',\n",
    "                        'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "   'KRT5', 'COL17A1', 'ITGB4', 'CXCL14', 'MKI67', 'CDK1',  'AQP3', 'KRT19', 'FGFBP1',  'S100A16', 'SLPI',  'MYC', \n",
    " 'SERPINB2', 'SERPINB3', 'SLPI', 'NOD2', 'IL36A','CNFN', 'IL36G',   'SAA1', 'SAA2',\n",
    " 'CFH', 'CFHR3', 'ODAM', 'ODAPH', 'LAMC2', 'CXCL1', 'CXCL2', 'CXCL6', \n",
    "'IL1A', 'DNASE1L3',\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.1,\n",
    "    vmax=1,                       # Set a max value for the scale\n",
    "    figsize=(14, 4) \n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=False)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl4_dotplot_horizontal.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19fa8e00-5455-43ed-b404-4525b5f28e65",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c137c089-2af7-49dd-b2e9-dbb1e14c3c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0affa7-cbeb-4116-997c-2a99ecbef604",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 900, 2900  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2600, 4600  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Lymphoid': 'cyan', \n",
    "    'Plasma': '#FF00FF', \n",
    "    'Crevicular' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_cc14'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_cc14'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_cc14'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_cc14'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (2080, 2380, 3600, 3900),  # Lymphoid ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c50ac1d0-5629-40d5-9540-dfed324f886f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1500, 3000  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2700, 4200  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Lymphoid': 'cyan', \n",
    "    'Plasma': '#FF00FF', \n",
    "    'Crevicular' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_cc14'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_cc14'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_cc14'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_cc14'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "plt.gca().add_artist(scalebar)\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (1800, 2300, 3500, 4000),  # Lymphoid ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_Lymphoid_Niches_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b692f97-266a-4df6-8775-c9277a914474",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1500, 3000  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2700, 4200  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "'B' : 'cyan', 'PB': 'yellow', 'Pl.1': '#FF5F15', 'Pl.2': '#FF5F15', 'Th_proximity_to_B': 'magenta'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "plt.gca().add_artist(scalebar)\n",
    "# Define the ROIs and add squares\n",
    "rois = [\n",
    "    (1800, 2300, 3500, 4000),  # Lymphoid ROI\n",
    "]\n",
    "\n",
    "# Draw the squares with no fill and white outline\n",
    "for xmin, xmax, ymin, ymax in rois:\n",
    "    plt.gca().add_patch(plt.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, \n",
    "                                      edgecolor='white', facecolor='none', linewidth=2))\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_Lymphoid_Clusters_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95f18bbf-57e9-40ef-bf4c-f824f9e64ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1800, 2300  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3500, 4000  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "'B' : 'cyan', 'PB': 'yellow', 'Pl.1': '#FF5F15', 'Pl.2': '#FF5F15', 'Th_proximity_to_B': 'magenta'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 50  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=15,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "plt.gca().add_artist(scalebar)\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_Lymphoid_clusters__small_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7255a718-6c11-4ff5-8fdf-853944be4216",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['final_label_citeSeq'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206de01f-db8e-4ba1-a0c0-6309f9e0d2dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1800, 2300  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3500, 4000  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "'B Naive' : '#0FFF50', 'B Memory 2' : 'cyan', 'B Memory 1' : 'cyan',\n",
    " 'Plasmablast': 'yellow', 'Plasma 1': '#FF5F15', 'Plasma 2': '#FF5F15',\n",
    "  'CD4 TFreg' : 'magenta', 'CD4 IL17A+ FOXP3+': 'pink',  \n",
    "    'CD4 T naive': 'pink', 'CD4 TCM': 'pink', 'CD4 TFH': 'red',\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = merged[merged.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Remove '-Xenium' from the index\n",
    "adata_sample.obs.index = adata_sample.obs.index.str.replace(r'-Xenium$', '', regex=True)\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['final_label_citeSeq'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['final_label_citeSeq'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['final_label_citeSeq'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['final_label_citeSeq'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 50  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=15,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_CiteSeq_small{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38d7998-03d5-4636-aa48-79d404b47253",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1800, 2300  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3500, 4000  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "'CD4 TFreg' : 'orange', 'mregDC' : 'cyan', 'CD4 IL17A+ FOXP3+': 'purple',  'CD4 T naive': 'purple','CD4 Treg': 'red', 'CD4 TCM': 'purple', 'CD4 T naive': 'purple', 'CD8 T': 'yellow', 'CD8 T IKZF2+': 'yellow', 'CD8 T naive': 'yellow',\n",
    "                       'CD8 T TNF+': 'yellow', 'CD8 TRM': 'yellow' , 'CD4 TFH': '#39e75f'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = merged[merged.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Remove '-Xenium' from the index\n",
    "adata_sample.obs.index = adata_sample.obs.index.str.replace(r'-Xenium$', '', regex=True)\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['final_label_citeSeq'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['final_label_citeSeq'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['final_label_citeSeq'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['final_label_citeSeq'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=10,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl4{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7a7154e-cdd5-4409-9484-8ac758f83075",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1500, 3000  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2700, 4200  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'Lymphoid': 'cyan', \n",
    "    'Plasma': '#FF00FF', \n",
    "    'Crevicular' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#4682B4\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_cc14'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_cc14'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_cc14'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_cc14'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 200  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca5fa0eb-64d5-46cf-b0da-d94fe2b3a0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2080, 2380  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3600, 3900  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'PB': 'magenta', \n",
    "    'Pl.1': 'purple', \n",
    "    'Pl.2' : 'purple',\n",
    "    'Th' : 'green',\n",
    "    \"Th_proximity_to_B\" : 'green',\n",
    "    'B': 'red', \n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=10,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl4{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac358952-4052-4c93-95d9-f84531c9e771",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099b1d42-a5e0-4a0f-b1ad-eca194b67d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = sc.read_h5ad('/data/vasileiosionat2/Xenium/Integration_2025_only_stromal/Modified_citeseq_files/Xenium_citeSeq_total_harmony.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "571b1882-724e-40d6-a278-a3276bf59075",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580208a9-f9fc-4e5c-920f-af9851cf50bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 2080, 2380  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3600, 3900  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'CD4 T naive': 'green', \n",
    "    'CD4 TFH': 'yellow', \n",
    "    'CD4 TCM' : 'red',\n",
    "    'CD4 TFreg' : 'cyan',\n",
    "    'CD4 IL17A+ FOXP3+' : 'magenta',     \n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = merged[merged.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Remove '-Xenium' from the index\n",
    "adata_sample.obs.index = adata_sample.obs.index.str.replace(r'-Xenium$', '', regex=True)\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['final_label_citeSeq'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['final_label_citeSeq'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['final_label_citeSeq'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['final_label_citeSeq'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=10,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl4{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "230cff7b-d9d8-4af8-a2cc-8e624eeea6a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the occurrences of each cell type\n",
    "cell_type_counts = adata_cropped.obs['final_label_citeSeq'].value_counts()\n",
    "print(cell_type_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfd05ba-7d48-45b3-92a6-caec2876c8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r3/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1350, 1650  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 3550, 3850  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "   'PB': '#FF00FF', \n",
    "    'Pl.1': '#AA336A', \n",
    "    'Pl.2' : '#AA336A',\n",
    "    'Th' : '#00A36C',\n",
    "    \"Th_proximity_to_B\" : '#00A36C',\n",
    "    'Fib.2' : 'yellow',\n",
    "    'VEC.1' : 'blue',\n",
    "    'B': '#FF4433', \n",
    "    'mregDC':'#00FFFF',    \n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r3\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=10,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Total_Epi_Lvl4{sample_id}.svg\"\n",
    "#plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e66a4d8-6b1c-4b3a-9394-d040a56c039b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c0234ba-82a5-43d8-892e-8cd37f5db6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b984be3-1d00-4f49-95e4-42b85a2f1bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4200, 5000  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2400, 3500  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r2_HV184\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl1'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl1'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl1'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl1'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Major_clusters_Lvl1{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db8ef9f-22cb-45b4-81d6-51869b68c681",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4300, 4900  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 2300, 3250  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "     'B' : '#7DF9FF',\n",
    "     'Th' : '#0818A8',\n",
    "     'Th_proximity_to_B' : '#0096FF',\n",
    "     'mregDC' : '#40E0D0',\n",
    "     'Pl.1' : '#800020',\n",
    "     'Pl.2' : '#F88379' ,\n",
    "     'PB' : '#FF0000' ,\n",
    "     'Fib.2' : '#DFFF00',\n",
    "     'Mac' : '#8B8000' ,\n",
    "     'MyoF' : '#FFC000'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#454143\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r2_HV184\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Lvl4_Xenium_TLS_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efef5413-b3b0-4c12-a717-3e9f22fa8210",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r2_HV184/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "xmin, xmax = 4200, 5000  # Horizontal bounds\n",
    "ymin, ymax = 2400, 3500  # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"KRT19\", \"PTPRC\", \"VCAN\", \"VWF\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"KRT19\": \"#FFEA00\",  \n",
    "    \"PTPRC\": \"#008000\",  \n",
    "    \"VCAN\": \"#0000FF\",  \n",
    "    \"VWF\": \"#990F0FFF\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Xenium_major_transcripts_HV184.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f961318a-70af-4377-8a5e-3ddeed4b0149",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s1r2_HV187A/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 1500, 2600  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 450, 1250  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s1r2\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl1'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl1'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl1'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl1'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Major_clusters_Lvl1{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2725815-3aac-47f0-89e3-ccff95017b1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4000, 4800  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1850, 2950  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Vascular': '#990F0FFF',\n",
    "    'Other': '#e5e5e5'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r1_HV207\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl1'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl1'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl1'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl1'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Major_clusters_Lvl1{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3ce0786-e312-42c6-8163-7793a226d555",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4000, 4800  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1850, 2950  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "    'Fibrous CT' : 'red',\n",
    "     'Epi-CT': '#00FFFF',\n",
    "     'Plasma-Fib CT' : '#CF9FFF',\n",
    "     'Plasma' : 'magenta' ,\n",
    "     'Spinous' : '#5D3FD3',\n",
    "     'Keratin' : '#0096FF' ,\n",
    "     'Crevicular' : '#800080' ,\n",
    "     'Lymphoid' : 'yellow'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#e1e1e1\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r1_HV207\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_knn50k10_merged'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_knn50k10_merged'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_knn50k10_merged'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_knn50k10_merged'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Niches_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6fceca2-b12e-475c-9121-471dbdde334d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4100, 4700  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1600, 2550  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "     'Plasma-Fib CT' : '#7F00FF',\n",
    "     'Plasma' : 'red' ,\n",
    "     'Crevicular' : '#D3D3D3' ,\n",
    "     'Lymphoid' : 'cyan'\n",
    "}\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#454143\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r1_HV207\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['niche_knn50k10_merged'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['niche_knn50k10_merged'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['niche_knn50k10_merged'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['niche_knn50k10_merged'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_Niches_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1b5b247-ff3f-4123-a548-d9306e40d6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl4'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d185c385-06b9-4482-b38b-81e0a5bb5deb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Load the data\n",
    "csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/cell_boundaries.csv.gz\"\n",
    "data = pd.read_csv(csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "deep_xmin, deep_xmax = 4100, 4700  # Horizontal bounds\n",
    "deep_ymin, deep_ymax = 1600, 2550  # Vertical bounds\n",
    "\n",
    "deep_vertices = data[\n",
    "    (data['vertex_x'] >= deep_xmin) & (data['vertex_x'] <= deep_xmax) &\n",
    "    (data['vertex_y'] >= deep_ymin) & (data['vertex_y'] <= deep_ymax)\n",
    "]\n",
    "deep_cell_ids = deep_vertices['cell_id'].unique()\n",
    "deep_data = data[data['cell_id'].isin(deep_cell_ids)]\n",
    "grouped = deep_data.groupby('cell_id')\n",
    "\n",
    "# Define the color mapping for the specific clusters you want to color uniquely\n",
    "# Define the specific cluster colors\n",
    "unique_cluster_colors = {\n",
    "     'B' : '#7DF9FF',\n",
    "     'Th' : '#0818A8',\n",
    "     'Th_proximity_to_B' : '#0096FF',\n",
    "     'mregDC' : '#40E0D0',\n",
    "     'Pl.1' : '#800020',\n",
    "     'Pl.2' : '#F88379' ,\n",
    "     'PB' : '#FF0000' ,\n",
    "     'Fib.2' : '#DFFF00',\n",
    "     'Mac' : '#8B8000' ,\n",
    "     'MyoF' : '#FFC000'\n",
    "}\n",
    "\n",
    "\n",
    "# For clusters not in `unique_cluster_colors`, set them to grey\n",
    "default_color = \"#454143\"\n",
    "\n",
    "# Subset the AnnData object for the sample\n",
    "sample_id = \"s2r1_HV207\"  # Replace with your sample's ID\n",
    "adata_sample = adata[adata.obs['sample'] == sample_id].copy()  # Ensure `sample_id` exists in `adata.obs`\n",
    "\n",
    "# Ensure that the 'cell_id' column in `adata_sample.obs` matches the IDs from the CSV\n",
    "adata_cropped = adata_sample[adata_sample.obs.index.isin(deep_cell_ids)].copy()\n",
    "# Create a dictionary for cell_id -> lvl1_cluster from adata_cropped\n",
    "cluster_mapping = adata_cropped.obs['Lvl4'].to_dict()\n",
    "\n",
    "# Add a cluster column to the filtered CSV data\n",
    "deep_data = deep_vertices.copy()\n",
    "deep_data['Lvl4'] = deep_data['cell_id'].map(cluster_mapping)\n",
    "\n",
    "# Drop rows with missing clusters (if any cell IDs in the cropped area are not in adata_cropped)\n",
    "deep_data = deep_data.dropna(subset=['Lvl4'])\n",
    "\n",
    "# Group by cell_id and plot each mask\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Set the style for black background\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "# Plot the nuclei masks with unique colors for the specified clusters\n",
    "for cell_id, group in deep_data.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    cluster = group['Lvl4'].iloc[0]  # All rows for the same cell_id should have the same cluster\n",
    "    \n",
    "    # Get the color for the cluster or default to grey\n",
    "    color = unique_cluster_colors.get(cluster, default_color)\n",
    "    \n",
    "    # Plot the mask\n",
    "    plt.fill(x, y, color=color, edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "\n",
    "# Hide the axes for better visual presentation\n",
    "plt.axis('off')\n",
    "\n",
    "# Set the axis limits to match the cropped area\n",
    "plt.xlim(deep_xmin, deep_xmax)\n",
    "plt.ylim(deep_ymin, deep_ymax)\n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 100  \n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=20,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_Lvl4_TLS_{sample_id}.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black')\n",
    "\n",
    "# Show the plot with a black background\n",
    "plt.gcf().patch.set_facecolor('black')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b00cbe02-4618-4e93-9b99-d05e130a54aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_knn50k10_merged'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1799ad3-2aa1-4daa-a453-1fc8e26719d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "\n",
    "# Load the CSV file for transcripts\n",
    "transcript_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/transcripts.csv.gz\"\n",
    "transcripts = pd.read_csv(transcript_csv_path)\n",
    "\n",
    "# Load the CSV file for the nucleus boundaries\n",
    "nucleus_csv_path = \"/data/vasileiosionat2/Xenium/Plotting_Paper_figures/s2r1-HV207/cell_boundaries.csv.gz\"\n",
    "nuclei = pd.read_csv(nucleus_csv_path)\n",
    "\n",
    "# Generate boundaries of cropped area\n",
    "xmin, xmax = 4000, 4800  # Horizontal bounds\n",
    "ymin, ymax = 1850, 2950 # Vertical bounds\n",
    "\n",
    "# Filter the transcripts based on the cropped area\n",
    "filtered_transcripts = transcripts[\n",
    "    (transcripts['x_location'] >= xmin) & (transcripts['x_location'] <= xmax) &\n",
    "    (transcripts['y_location'] >= ymin) & (transcripts['y_location'] <= ymax)\n",
    "]\n",
    "\n",
    "# Filter the specific transcripts you want to plot\n",
    "transcript_of_interest = [\"KRT19\", \"PTPRC\", \"VCAN\", \"VWF\"]\n",
    "filtered_transcripts = filtered_transcripts[filtered_transcripts['feature_name'].isin(transcript_of_interest)]\n",
    "\n",
    "# Filter the nuclei data based on the cropped area\n",
    "filtered_nuclei = nuclei[\n",
    "    (nuclei['vertex_x'] >= xmin) & (nuclei['vertex_x'] <= xmax) &\n",
    "    (nuclei['vertex_y'] >= ymin) & (nuclei['vertex_y'] <= ymax)\n",
    "]\n",
    "\n",
    "# Manually define a colormap for each transcript\n",
    "transcript_color_map = {\n",
    "    \"KRT19\": \"#FFEA00\",  \n",
    "    \"PTPRC\": \"#008000\",  \n",
    "    \"VCAN\": \"#0000FF\",  \n",
    "    \"VWF\": \"#990F0FFF\",\n",
    "}\n",
    "\n",
    "# Start plotting with black background\n",
    "plt.style.use('dark_background')  # Set the background to black\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# Plot the nuclei masks in grey with black contours (no fill)\n",
    "for cell_id, group in filtered_nuclei.groupby('cell_id'):\n",
    "    x = group['vertex_x'].values\n",
    "    y = group['vertex_y'].values\n",
    "    plt.plot(x, y, color='grey', linewidth=0.25)  # Only outlines, no fill\n",
    "\n",
    "# Plot the selected transcripts as points with unique colors\n",
    "for idx, row in filtered_transcripts.iterrows():\n",
    "    if row['feature_name'] in transcript_of_interest:\n",
    "        # Get the color for the current transcript from the manual colormap\n",
    "        color = transcript_color_map[row['feature_name']]\n",
    "        plt.scatter(row['x_location'], row['y_location'], color=color, s=3, alpha=0.7)\n",
    "\n",
    "\n",
    "# Set the aspect ratio to be equal to avoid distortion\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Remove axes labels and ticks\n",
    "plt.xticks([])  \n",
    "plt.yticks([])  \n",
    "\n",
    "# Add the scale bar\n",
    "scale_bar_length_px = 40  # Calculated from 40 microns / 0.2125 microns per pixel\n",
    "scalebar = AnchoredSizeBar(\n",
    "    plt.gca().transData,\n",
    "    scale_bar_length_px,  # Length in pixels\n",
    "    \"\",  # Label\n",
    "    loc=\"lower left\",\n",
    "    pad=0.5,\n",
    "    color=\"white\",\n",
    "    frameon=False,\n",
    "    size_vertical=5,\n",
    "    label_top=True,\n",
    ")\n",
    "\n",
    "plt.gca().add_artist(scalebar)\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Xenium_major_transcripts_HV207.svg\"\n",
    "plt.savefig(output_path, format='svg', facecolor='black', bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "237b083a-a4f5-425b-b4bf-a97fa5aa2218",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure all observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 2: Log-transform the data if not already done\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# Step 3: Perform differential expression analysis using the Wilcoxon method\n",
    "sc.tl.rank_genes_groups(adata, 'niche_cc14', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for each cluster (optional for reference)\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata.obs['niche_cc14'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata.uns['rank_genes_groups']['names'][cluster][:10]  # Top 10 genes\n",
    "\n",
    "# Step 5: Generate dendrogram for cluster ordering\n",
    "sc.tl.dendrogram(adata, groupby='niche_cc14')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata.uns['dendrogram_niche_cc14']['categories_ordered']\n",
    "\n",
    "# Step 7: Define your custom list of genes for plotting\n",
    "custom_genes = ['ANPEP', 'ANXA1', 'APOA5', 'C3', 'CAMP', 'CCL19', 'CCL5', 'CCN1', 'CD274',\n",
    "                'CD28', 'CD34', 'CD70', 'CD86', 'CFH', 'COL16A1', 'COL17A1', 'COL5A2', 'COL8A1', \n",
    "                'CSF3', 'CXCL1', 'CXCL10', 'CXCL12', 'CXCL14', 'CXCL2', 'CXCL5', 'CXCL6', 'CXCL9', 'EDN1', \n",
    "                'GDF15', 'HLA-DMB', 'HLA-DQA2', 'HLA-DQB2', 'HLA-DRB5', 'ICAM1', 'IGF1', 'IL1A', 'IL1RN', \n",
    "                'IL23A', 'IL33', 'ITGB2', 'KNG1', 'LAMC2', 'LPL', 'MMRN2', 'PECAM1', 'POSTN', 'PRG4', \n",
    "                'PTN', 'PTPRC', 'RETN', 'S100A12', 'SAA1', 'SLAMF7', 'SLPI', 'TFF2', 'THBS2', 'THY1', 'TNC', 'VCAN']  # Replace with your gene list\n",
    "\n",
    "# Filter the custom list to include only genes present in the dataset\n",
    "genes_for_plot = [gene for gene in custom_genes if gene in adata.var_names]\n",
    "\n",
    "# Step 8: Plot the dotplot with your chosen genes\n",
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    var_names=genes_for_plot,\n",
    "    groupby='niche_cc14',\n",
    "    dendrogram=True,\n",
    "    use_raw=False,\n",
    "    cmap=\"vlag\",\n",
    "    standard_scale='var',\n",
    "    swap_axes=True, \n",
    "    dot_max= 0.7\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba43fae-01b7-4dc5-bf26-176d77c43133",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Step 1: Ensure all observation names are unique\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Step 2: Log-transform the data if not already done\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# Step 3: Perform differential expression analysis using the Wilcoxon method\n",
    "sc.tl.rank_genes_groups(adata, 'niche_cc14', method='wilcoxon', use_raw=False)\n",
    "\n",
    "# Step 4: Extract top marker genes for each cluster (optional for reference)\n",
    "top_genes_per_cluster = {}\n",
    "for cluster in adata.obs['niche_cc14'].cat.categories:\n",
    "    top_genes_per_cluster[cluster] = adata.uns['rank_genes_groups']['names'][cluster][:10]  # Top 10 genes\n",
    "\n",
    "# Step 5: Generate dendrogram for cluster ordering\n",
    "sc.tl.dendrogram(adata, groupby='niche_cc14')\n",
    "\n",
    "# Step 6: Retrieve the cluster order based on the dendrogram\n",
    "cluster_order = adata.uns['dendrogram_niche_cc14']['categories_ordered']\n",
    "\n",
    "# Step 7: Define your custom list of genes for plotting\n",
    "custom_genes = ['SELE', 'EGFR', 'FPR1', 'LPL', 'CD19', 'ITGB2', 'VSIG4', 'CCR7', \n",
    "                'ACKR1', 'SDC1', 'TLR2', 'ITGB5', 'TLR4', 'PDCD1', 'CD86', 'CD247', \n",
    "                'CD4', 'SELL', 'CD27', 'CD28', 'CTLA4', 'CFB', 'ITGA1', 'CD93', 'CSF3R',\n",
    "                'CSF2RA', 'CXCR4', 'EDNRB', 'ADGRL4', 'ERBB2', 'LAG3', 'IL2RA', 'ITGB4', 'IL1R2', \n",
    "                'IL23R', 'LILRB2', 'IL1RL1', 'ICAM1', 'THY1', 'CLEC14A', 'PECAM1', 'MRC1', 'SLAMF7']  # Replace with your gene list\n",
    "\n",
    "# Filter the custom list to include only genes present in the dataset\n",
    "genes_for_plot = [gene for gene in custom_genes if gene in adata.var_names]\n",
    "\n",
    "# Step 8: Plot the dotplot with your chosen genes\n",
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    var_names=genes_for_plot,\n",
    "    groupby='niche_cc14',\n",
    "    dendrogram=True,\n",
    "    use_raw=False,\n",
    "    cmap=\"vlag\",\n",
    "    standard_scale='var',\n",
    "    swap_axes=True, \n",
    "    dot_max=0.7\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "431fe6e8-d49b-4265-882d-0c46151fb6ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ba86ad0-939c-4cb4-8d93-072a823881bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Assuming adata is already loaded\n",
    "# List of ligand-receptor pairs\n",
    "ligand_receptor_pairs = [\n",
    "    ('ANPEP', 'SELE'),\n",
    "    ('ANXA1', 'EGFR'),\n",
    "    ('ANXA1', 'FPR1'),\n",
    "    ('APOA5', 'LPL'),\n",
    "    ('C3', 'CD19'),\n",
    "    ('C3', 'ITGB2'),\n",
    "    ('C3', 'VSIG4'),\n",
    "    ('CAMP', 'EGFR'),\n",
    "    ('CCL19', 'CCR7'),\n",
    "    ('CCL5', 'ACKR1'),\n",
    "    ('CCL5', 'SDC1'),\n",
    "    ('CCN1', 'TLR2'),\n",
    "    ('CCN1', 'ITGB5'),\n",
    "    ('CCN1', 'TLR4'),\n",
    "    ('CCN1', 'ITGB2'),\n",
    "    ('CD274', 'PDCD1'),\n",
    "    ('CD28', 'CD86'),\n",
    "    ('CD28', 'CD247'),\n",
    "    ('CD28', 'CD4'),\n",
    "    ('CD34', 'SELL'),\n",
    "    ('CD34', 'SELE'),\n",
    "    ('CD70', 'CD27'),\n",
    "    ('CD86', 'CD28'),\n",
    "    ('CD86', 'CTLA4'),\n",
    "    ('CFH', 'CFB'),\n",
    "    ('CFH', 'SELL'),\n",
    "    ('COL16A1', 'ITGA1'),\n",
    "    ('COL17A1', 'ITGA1'),\n",
    "    ('COL5A2', 'SDC1'),\n",
    "    ('COL5A2', 'CD93'),\n",
    "    ('COL5A2', 'ITGA1'),\n",
    "    ('COL8A1', 'ITGA1'),\n",
    "    ('CSF3', 'CSF3R'),\n",
    "    ('CSF3', 'CSF2RA'),\n",
    "    ('CXCL1', 'ACKR1'),\n",
    "    ('CXCL10', 'ACKR1'),\n",
    "    ('CXCL12', 'CXCR4'),\n",
    "    ('CXCL12', 'CD4'),\n",
    "    ('CXCL14', 'CXCR4'),\n",
    "    ('CXCL2', 'ACKR1'),\n",
    "    ('CXCL5', 'ACKR1'),\n",
    "    ('CXCL6', 'ACKR1'),\n",
    "    ('CXCL9', 'ACKR1'),\n",
    "    ('EDN1', 'EDNRB'),\n",
    "    ('EDN1', 'ADGRL4'),\n",
    "    ('EDN1', 'EGFR'),\n",
    "    ('GDF15', 'ERBB2'),\n",
    "    ('HLA-DMB', 'CD4'),\n",
    "    ('HLA-DQA2', 'LAG3'),\n",
    "    ('HLA-DQA2', 'CD4'),\n",
    "    ('HLA-DQB2', 'LAG3'),\n",
    "    ('HLA-DRB5', 'LAG3'),\n",
    "    ('HLA-DRB5', 'CD4'),\n",
    "    ('ICAM1', 'EGFR'),\n",
    "    ('ICAM1', 'IL2RA'),\n",
    "    ('ICAM1', 'ITGB2'),\n",
    "    ('IGF1', 'ITGB4'),\n",
    "    ('IL1A', 'IL1R2'),\n",
    "    ('IL1RN', 'IL1R2'),\n",
    "    ('IL23A', 'IL23R'),\n",
    "    ('IL23A', 'LILRB2'),\n",
    "    ('IL33', 'IL1RL1'),\n",
    "    ('ITGB2', 'ICAM1'),\n",
    "    ('ITGB2', 'THY1'),\n",
    "    ('KNG1', 'ITGB2'),\n",
    "    ('KNG1', 'CD93'),\n",
    "    ('LAMC2', 'ITGB4'),\n",
    "    ('LPL', 'SDC1'),\n",
    "    ('MMRN2', 'CD93'),\n",
    "    ('MMRN2', 'CLEC14A'),\n",
    "    ('PECAM1', 'PECAM1'),\n",
    "    ('POSTN', 'ITGB5'),\n",
    "    ('PRG4', 'TLR2'),\n",
    "    ('PRG4', 'TLR4'),\n",
    "    ('PTN', 'SDC1'),\n",
    "    ('PTPRC', 'MRC1'),\n",
    "    ('PTPRC', 'CD247'),\n",
    "    ('PTPRC', 'CD4'),\n",
    "    ('RETN', 'TLR4'),\n",
    "    ('S100A12', 'TLR4'),\n",
    "    ('SAA1', 'FPR1'),\n",
    "    ('SAA1', 'TLR2'),\n",
    "    ('SLAMF7', 'SLAMF7'),\n",
    "    ('SLPI', 'CD4'),\n",
    "    ('TFF2', 'CXCR4'),\n",
    "    ('THBS2', 'SDC1'),\n",
    "    ('THY1', 'ITGB2'),\n",
    "    ('TNC', 'SDC1'),\n",
    "    ('TNC', 'EGFR'),\n",
    "    ('VCAN', 'EGFR'),\n",
    "    ('VCAN', 'SELL'),\n",
    "    ('VCAN', 'TLR2')\n",
    "]\n",
    "\n",
    "# Directory to save files\n",
    "output_dir = \"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/\"\n",
    "\n",
    "# Create a dictionary to store ligand-receptor pairs\n",
    "ligand_dict = {}\n",
    "\n",
    "# Loop through the pairs to collect receptor data for each ligand\n",
    "for ligand, receptor in ligand_receptor_pairs:\n",
    "    if ligand not in ligand_dict:\n",
    "        ligand_dict[ligand] = []\n",
    "    ligand_dict[ligand].append(receptor)\n",
    "\n",
    "# Loop through each ligand and save its receptors in a separate CSV file\n",
    "for ligand, receptors in ligand_dict.items():\n",
    "    # Get the gene expression data for the ligand and its receptors from adata\n",
    "    data = adata.obs[['cell_id', 'x_centroid', 'y_centroid', 'sample', 'niche_cc14', 'Lvl4']].copy()\n",
    "\n",
    "    # Add a column for the ligand's expression\n",
    "    if ligand in adata.var_names:\n",
    "        ligand_expression = adata.raw[:, ligand].X if adata.raw is not None else adata.X[:, adata.var_names.get_loc(ligand)]\n",
    "        ligand_expression_dense = ligand_expression.toarray() if hasattr(ligand_expression, 'toarray') else ligand_expression\n",
    "        data[ligand] = ligand_expression_dense > 0\n",
    "\n",
    "    # Add a column for each receptor's expression\n",
    "    for receptor in receptors:\n",
    "        if receptor in adata.var_names:\n",
    "            # Check if raw data exists, otherwise use adata.X\n",
    "            expression_data = adata.raw[:, receptor].X if adata.raw is not None else adata.X[:, adata.var_names.get_loc(receptor)]\n",
    "            \n",
    "            # Convert sparse matrix to dense format and check for expression\n",
    "            expression_data_dense = expression_data.toarray() if hasattr(expression_data, 'toarray') else expression_data\n",
    "            data[receptor] = expression_data_dense > 0  # You can change this condition based on the threshold you want to use\n",
    "\n",
    "    # Save to a CSV file\n",
    "    file_name = os.path.join(output_dir, f\"{ligand}_receptors.csv\")\n",
    "    data.to_csv(file_name, index=False)\n",
    "    print(f\"Saved {ligand} receptors to {file_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d20a416-46d6-4aa1-b6d0-6598ac32761f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Directory where ligand-receptor CSV files are saved\n",
    "input_dir = \"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/\"\n",
    "\n",
    "# Directory to save files for each sample\n",
    "output_base_dir = \"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Samples/\"\n",
    "\n",
    "# List of ligand-receptor files (you can get this from the directory)\n",
    "ligand_files = [f for f in os.listdir(input_dir) if f.endswith('_receptors.csv')]\n",
    "\n",
    "# Loop through each ligand-receptor file\n",
    "for ligand_file in ligand_files:\n",
    "    ligand_path = os.path.join(input_dir, ligand_file)\n",
    "    \n",
    "    # Load the ligand-receptor data into a DataFrame\n",
    "    data = pd.read_csv(ligand_path)\n",
    "    \n",
    "    # Get the unique samples from the 'sample' column\n",
    "    unique_samples = data['sample'].unique()\n",
    "    \n",
    "    # Loop through each unique sample and save the filtered data\n",
    "    for sample in unique_samples:\n",
    "        # Filter the data for the current sample\n",
    "        sample_data = data[data['sample'] == sample]\n",
    "        \n",
    "        # Create a folder for the sample if it doesn't exist\n",
    "        sample_folder = os.path.join(output_base_dir, sample)\n",
    "        os.makedirs(sample_folder, exist_ok=True)\n",
    "        \n",
    "        # Save the filtered data to a new CSV file\n",
    "        sample_file_name = f\"{ligand_file.replace('.csv', f'_{sample}.csv')}\"\n",
    "        sample_file_path = os.path.join(sample_folder, sample_file_name)\n",
    "        sample_data.to_csv(sample_file_path, index=False)\n",
    "        \n",
    "        print(f\"Saved {sample_file_name} to {sample_folder}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c885c03-b648-480d-8876-1ed76f90429d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Define the ligands and receptors\n",
    "ligands = ['ANPEP', 'ANXA1', 'APOA5', 'C3', 'CAMP', 'CCL19', 'CCL5', 'CCN1', 'CD274',\n",
    "                'CD28', 'CD34', 'CD70', 'CD86', 'CFH', 'COL16A1', 'COL17A1', 'COL5A2', 'COL8A1', \n",
    "                'CSF3', 'CXCL1', 'CXCL10', 'CXCL12', 'CXCL14', 'CXCL2', 'CXCL5', 'CXCL6', 'CXCL9', 'EDN1', \n",
    "                'GDF15', 'HLA-DMB', 'HLA-DQA2', 'HLA-DQB2', 'HLA-DRB5', 'ICAM1', 'IGF1', 'IL1A', 'IL1RN', \n",
    "                'IL23A', 'IL33', 'ITGB2', 'KNG1', 'LAMC2', 'LPL', 'MMRN2', 'PECAM1', 'POSTN', 'PRG4', \n",
    "                'PTN', 'PTPRC', 'RETN', 'S100A12', 'SAA1', 'SLAMF7', 'SLPI', 'TFF2', 'THBS2', 'THY1', 'TNC', 'VCAN']  # Replace with your actual list of ligands \n",
    "\n",
    "# Define the directory where your receptor CSVs are stored\n",
    "csv_directory = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Samples/HV220A_B'\n",
    "\n",
    "# Initialize a dictionary to store results\n",
    "ligand_table = []\n",
    "\n",
    "# Iterate through each ligand\n",
    "for ligand in ligands:\n",
    "    print(f\"Processing ligand: {ligand}\")\n",
    "    \n",
    "    # Path to the ligand's receptor CSV\n",
    "    receptor_csv_path = os.path.join(csv_directory, f\"{ligand}_receptors_HV220A_B.csv\")\n",
    "    \n",
    "    if os.path.exists(receptor_csv_path):\n",
    "        # Read the receptor CSV\n",
    "        receptor_df = pd.read_csv(receptor_csv_path)\n",
    "        \n",
    "        # Count the number of cells that express the ligand (i.e., the number of \"True\" in the ligand column)\n",
    "        ligand_cell_count = receptor_df[ligand].sum()  # Assuming ligand column has boolean values (True/False)\n",
    "        \n",
    "        # Create a row for this ligand with the ligand name and ligand cell count\n",
    "        ligand_row = [ligand, ligand_cell_count]\n",
    "        \n",
    "        # Append the row to the ligand_table list\n",
    "        ligand_table.append(ligand_row)\n",
    "    \n",
    "\n",
    "# Convert the ligand_table to a DataFrame\n",
    "ligand_table_df = pd.DataFrame(ligand_table, columns=['Ligand Name', 'Ligand Cell Count'])\n",
    "\n",
    "# Save the table to CSV\n",
    "output_table_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Stats.csv'\n",
    "#ligand_table_df.to_csv(output_table_path, index=False)\n",
    "\n",
    "# Print the saved table\n",
    "ligand_table_df\n",
    "print(f\"Ligand table saved to {output_table_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e51bf2e9-e070-4f4b-a5cd-d7d436d0ee56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "from sklearn.neighbors import BallTree\n",
    "\n",
    "# Define the ligands and directory\n",
    "ligands = ['ANPEP', 'ANXA1', 'APOA5', 'C3', 'CAMP', 'CCL19', 'CCL5', 'CCN1', 'CD274',\n",
    "           'CD28', 'CD34', 'CD70', 'CD86', 'CFH', 'COL16A1', 'COL17A1', 'COL5A2', 'COL8A1',\n",
    "           'CSF3', 'CXCL1', 'CXCL10', 'CXCL12', 'CXCL14', 'CXCL2', 'CXCL5', 'CXCL6', 'CXCL9',\n",
    "           'EDN1', 'GDF15', 'HLA-DMB', 'HLA-DQA2', 'HLA-DQB2', 'HLA-DRB5', 'ICAM1', 'IGF1',\n",
    "           'IL1A', 'IL1RN', 'IL23A', 'IL33', 'ITGB2', 'KNG1', 'LAMC2', 'LPL', 'MMRN2',\n",
    "           'PECAM1', 'POSTN', 'PRG4', 'PTN', 'PTPRC', 'RETN', 'S100A12', 'SAA1', 'SLAMF7',\n",
    "           'SLPI', 'TFF2', 'THBS2', 'THY1', 'TNC', 'VCAN']\n",
    "\n",
    "csv_directory = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Samples/HV220A_B'\n",
    "output_table_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Stats.csv'\n",
    "\n",
    "ligand_table = []\n",
    "\n",
    "# Process each ligand CSV\n",
    "for ligand in ligands:\n",
    "    print(f\"Processing ligand: {ligand}\")\n",
    "\n",
    "    receptor_csv_path = os.path.join(csv_directory, f\"{ligand}_receptors_HV220A_B.csv\")\n",
    "\n",
    "    if not os.path.exists(receptor_csv_path):\n",
    "        print(f\"File for {ligand} not found.\")\n",
    "        continue\n",
    "\n",
    "    receptor_df = pd.read_csv(receptor_csv_path)\n",
    "\n",
    "    # Ensure coordinate columns exist\n",
    "    if not {'x_centroid', 'y_centroid'}.issubset(receptor_df.columns):\n",
    "        print(f\"Missing coordinates for ligand {ligand}. Skipping.\")\n",
    "        continue\n",
    "\n",
    "    # Count ligand-positive cells\n",
    "    ligand_cell_count = receptor_df[ligand].sum() if ligand in receptor_df else 0\n",
    "\n",
    "    # Handle NaN in coordinates\n",
    "    receptor_df['z'] = receptor_df['z'].fillna(0) if 'z' in receptor_df.columns else 0\n",
    "    receptor_df[['x_centroid', 'y_centroid', 'z']] = receptor_df[['x_centroid', 'y_centroid', 'z']].fillna(0)\n",
    "\n",
    "    # Filter cells with ligand == True\n",
    "    data_img = receptor_df.loc[receptor_df[ligand] == True, ['x_centroid', 'y_centroid', 'z']]\n",
    "    \n",
    "    if data_img.empty:\n",
    "        print(f\"No true ligand cells for {ligand}.\")\n",
    "        ligand_table.append([ligand, ligand_cell_count, 0])\n",
    "        continue\n",
    "\n",
    "    # Build BallTree and query neighbors\n",
    "    kdt = BallTree(receptor_df[['x_centroid', 'y_centroid', 'z']], metric='euclidean')\n",
    "    ind = kdt.query_radius(data_img, r=10, return_distance=False)\n",
    "\n",
    "    # Count neighbors for each true ligand cell (excluding self)\n",
    "    num_total_neighbors = sum(len(neighbors) - 1 for neighbors in ind)\n",
    "\n",
    "    ligand_table.append([ligand, ligand_cell_count, num_total_neighbors])\n",
    "\n",
    "# Save results\n",
    "ligand_table_df = pd.DataFrame(ligand_table, columns=['Ligand Name', 'Ligand Cell Count', 'Total Neighbors'])\n",
    "ligand_table_df.to_csv(output_table_path, index=False)\n",
    "\n",
    "print(f\"Ligand table saved to {output_table_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb4411d-e725-4ce1-be52-4ec8fb643f6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "from sklearn.neighbors import BallTree\n",
    "\n",
    "# Define ligands and receptors\n",
    "ligands = ['ANPEP', 'ANXA1', 'APOA5', 'C3', 'CAMP', 'CCL19', 'CCL5', 'CCN1', 'CD274',\n",
    "           'CD28', 'CD34', 'CD70', 'CD86', 'CFH', 'COL16A1', 'COL17A1', 'COL5A2', 'COL8A1',\n",
    "           'CSF3', 'CXCL1', 'CXCL10', 'CXCL12', 'CXCL14', 'CXCL2', 'CXCL5', 'CXCL6', 'CXCL9',\n",
    "           'EDN1', 'GDF15', 'HLA-DMB', 'HLA-DQA2', 'HLA-DQB2', 'HLA-DRB5', 'ICAM1', 'IGF1',\n",
    "           'IL1A', 'IL1RN', 'IL23A', 'IL33', 'ITGB2', 'KNG1', 'LAMC2', 'LPL', 'MMRN2',\n",
    "           'PECAM1', 'POSTN', 'PRG4', 'PTN', 'PTPRC', 'RETN', 'S100A12', 'SAA1', 'SLAMF7',\n",
    "           'SLPI', 'TFF2', 'THBS2', 'THY1', 'TNC', 'VCAN']\n",
    "\n",
    "receptors = ['SELE', 'EGFR', 'FPR1', 'LPL', 'CD19', 'ITGB2', 'VSIG4', 'CCR7', 'ACKR1', 'SDC1',\n",
    "             'TLR2', 'ITGB5', 'TLR4', 'PDCD1', 'CD86', 'CD247', 'CD4', 'SELL', 'CD27', 'CD28',\n",
    "             'CTLA4', 'CFB', 'ITGA1', 'CD93', 'CSF3R', 'CSF2RA', 'CXCR4', 'EDNRB', 'ADGRL4',\n",
    "             'ERBB2', 'LAG3', 'IL2RA', 'ITGB4', 'IL1R2', 'IL23R', 'LILRB2', 'IL1RL1', 'ICAM1',\n",
    "             'THY1', 'CLEC14A', 'PECAM1', 'MRC1', 'SLAMF7']\n",
    "\n",
    "base_directory = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Samples/'\n",
    "output_directory = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Neighbors/'\n",
    "\n",
    "# Iterate over each sample folder\n",
    "for sample_folder in os.listdir(base_directory):\n",
    "    sample_path = os.path.join(base_directory, sample_folder)\n",
    "    \n",
    "    # Check if it's a folder\n",
    "    if not os.path.isdir(sample_path):\n",
    "        continue\n",
    "\n",
    "    ligand_table = []\n",
    "\n",
    "    # Process each ligand CSV in the current sample\n",
    "    for ligand in ligands:\n",
    "        print(f\"Processing ligand: {ligand} for sample: {sample_folder}\")\n",
    "\n",
    "        receptor_csv_path = os.path.join(sample_path, f\"{ligand}_receptors_{sample_folder}.csv\")\n",
    "\n",
    "        if not os.path.exists(receptor_csv_path):\n",
    "            print(f\"File for {ligand} not found in sample {sample_folder}.\")\n",
    "            continue\n",
    "\n",
    "        receptor_df = pd.read_csv(receptor_csv_path)\n",
    "\n",
    "        # Ensure coordinate columns exist\n",
    "        if not {'x_centroid', 'y_centroid'}.issubset(receptor_df.columns):\n",
    "            print(f\"Missing coordinates for ligand {ligand} in sample {sample_folder}. Skipping.\")\n",
    "            continue\n",
    "\n",
    "        # Count ligand-positive cells\n",
    "        ligand_cell_count = receptor_df[ligand].sum() if ligand in receptor_df else 0\n",
    "\n",
    "        # Handle NaN in coordinates\n",
    "        receptor_df['z'] = receptor_df['z'].fillna(0) if 'z' in receptor_df.columns else 0\n",
    "        receptor_df[['x_centroid', 'y_centroid', 'z']] = receptor_df[['x_centroid', 'y_centroid', 'z']].fillna(0)\n",
    "\n",
    "        # Filter cells with ligand == True\n",
    "        data_img = receptor_df.loc[receptor_df[ligand] == True, ['x_centroid', 'y_centroid', 'z']]\n",
    "\n",
    "        if data_img.empty:\n",
    "            print(f\"No true ligand cells for {ligand} in sample {sample_folder}.\")\n",
    "            ligand_row = [ligand, ligand_cell_count, 0] + [0] * len(receptors)\n",
    "            ligand_table.append(ligand_row)\n",
    "            continue\n",
    "\n",
    "        # Build BallTree and query neighbors\n",
    "        kdt = BallTree(receptor_df[['x_centroid', 'y_centroid', 'z']], metric='euclidean')\n",
    "        ind = kdt.query_radius(data_img, r=40, return_distance=False)\n",
    "\n",
    "        # Count total neighbors and receptor-specific neighbors\n",
    "        total_neighbors_per_ligand_cell = [max(len(neighbors) - 1, 0) for neighbors in ind]\n",
    "        avg_total_neighbors = np.mean(total_neighbors_per_ligand_cell) if total_neighbors_per_ligand_cell else 0\n",
    "\n",
    "        receptor_neighbor_counts_per_ligand_cell = []\n",
    "        for receptor in receptors:\n",
    "            if receptor not in receptor_df.columns:\n",
    "                receptor_neighbor_counts_per_ligand_cell.append(0)\n",
    "                continue\n",
    "\n",
    "            receptor_positive_cells = receptor_df[receptor] == True\n",
    "            receptor_indices = receptor_df.index[receptor_positive_cells].tolist()\n",
    "            receptor_counts_per_ligand_cell = [\n",
    "                sum(neighbor in receptor_indices for neighbor in neighbors) - 1 for neighbors in ind\n",
    "            ]\n",
    "            avg_receptor_neighbors = np.mean(receptor_counts_per_ligand_cell) if receptor_counts_per_ligand_cell else 0\n",
    "            receptor_neighbor_counts_per_ligand_cell.append(max(avg_receptor_neighbors, 0))\n",
    "\n",
    "        ligand_row = [ligand, ligand_cell_count, avg_total_neighbors] + receptor_neighbor_counts_per_ligand_cell\n",
    "        ligand_table.append(ligand_row)\n",
    "\n",
    "    # Save results for the current sample\n",
    "    columns = ['Ligand Name', 'Ligand Cell Count', 'Average Total Neighbors per Ligand Cell'] + [f'Average {receptor} Neighbors per Ligand Cell' for receptor in receptors]\n",
    "    ligand_table_df = pd.DataFrame(ligand_table, columns=columns)\n",
    "\n",
    "    # Ensure output folder exists\n",
    "    output_sample_path = os.path.join(output_directory, f\"{sample_folder}\")\n",
    "    os.makedirs(output_sample_path, exist_ok=True)\n",
    "\n",
    "    output_file_path = os.path.join(output_sample_path, f\"Ligand_Stats_with_receptors_{sample_folder}.csv\")\n",
    "    ligand_table_df.to_csv(output_file_path, index=False)\n",
    "\n",
    "    print(f\"Ligand table with receptor neighbors for sample {sample_folder} saved to {output_file_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d7cd66e-93fc-461b-bfeb-80301edf08f5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ligand_table_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9758f1a2-2c8c-4038-8ccc-6e31a9a47c82",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Define the directory containing the CSV files\n",
    "root_dir = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Neighbors'  # Change this to the folder path\n",
    "\n",
    "\n",
    "# Function to process each CSV file\n",
    "def process_csv(file_path):\n",
    "    try:\n",
    "        # Load the CSV file as a DataFrame\n",
    "        df = pd.read_csv(file_path)\n",
    "        \n",
    "        # Select columns that match the pattern 'Average ... Neighbors per Ligand Cell'\n",
    "        pattern = r'^Average .+ Neighbors per Ligand Cell$'\n",
    "        selected_columns = df.filter(regex=pattern)\n",
    "\n",
    "        # Extract gene names by removing the prefix 'Average ' and the suffix ' Neighbors per Ligand Cell'\n",
    "        gene_names = selected_columns.columns.str.replace(r'^Average ', '', regex=True)\n",
    "        gene_names = gene_names.str.replace(r' Neighbors per Ligand Cell$', '', regex=True)\n",
    "\n",
    "        # Replace the selected columns in the original DataFrame with the modified columns (gene names)\n",
    "        df[gene_names] = selected_columns\n",
    "        df = df.drop(selected_columns.columns, axis=1)  # Drop the original columns with old names\n",
    "\n",
    "        # Print the modified DataFrame to check\n",
    "        print(f\"Modified DataFrame for {file_path}:\")\n",
    "        print(df.head())  # Show the first few rows to check\n",
    "\n",
    "        # Save the updated CSV file if needed\n",
    "        df.to_csv(file_path, index=False)  # Uncomment this line to save the modified file\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {file_path}: {e}\")\n",
    "\n",
    "# Walk through all files in the directory and subdirectories\n",
    "for root, dirs, files in os.walk(root_dir):\n",
    "    for file in files:\n",
    "        if file.endswith('.csv'):\n",
    "            file_path = os.path.join(root, file)\n",
    "            process_csv(file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a10d723b-b0c9-42f8-8d08-a30e240f2876",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139733d6-7bce-4b8b-8208-7b6b2c2ba3e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "df = pd.read_csv('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Neighbors/s1r1_b/Ligand_Stats_with_receptors_s1r1_b.csv')\n",
    "\n",
    "# Create a list of ligand-receptor pairs\n",
    "ligand_receptor_pairs = [\n",
    "    ('ANPEP', 'SELE'), ('ANXA1', 'EGFR'), ('ANXA1', 'FPR1'), ('APOA5', 'LPL'),\n",
    "    ('C3', 'CD19'), ('C3', 'ITGB2'), ('C3', 'VSIG4'), ('CAMP', 'EGFR'),\n",
    "    ('CCL19', 'CCR7'), ('CCL5', 'ACKR1'), ('CCL5', 'SDC1'), ('CCN1', 'TLR2'),\n",
    "    ('CCN1', 'ITGB5'), ('CCN1', 'TLR4'), ('CCN1', 'ITGB2'), ('CD274', 'PDCD1'),\n",
    "    ('CD28', 'CD86'), ('CD28', 'CD247'), ('CD28', 'CD4'), ('CD34', 'SELL'),\n",
    "    ('CD34', 'SELE'), ('CD70', 'CD27'), ('CD86', 'CD28'), ('CD86', 'CTLA4'),\n",
    "    ('CFH', 'CFB'), ('CFH', 'SELL'), ('COL16A1', 'ITGA1'), ('COL17A1', 'ITGA1'),\n",
    "    ('COL5A2', 'SDC1'), ('COL5A2', 'CD93'), ('COL5A2', 'ITGA1'), ('COL8A1', 'ITGA1'),\n",
    "    ('CSF3', 'CSF3R'), ('CSF3', 'CSF2RA'), ('CXCL1', 'ACKR1'), ('CXCL10', 'ACKR1'),\n",
    "    ('CXCL12', 'CXCR4'), ('CXCL12', 'CD4'), ('CXCL14', 'CXCR4'), ('CXCL2', 'ACKR1'),\n",
    "    ('CXCL5', 'ACKR1'), ('CXCL6', 'ACKR1'), ('CXCL9', 'ACKR1'), ('EDN1', 'EDNRB'),\n",
    "    ('EDN1', 'ADGRL4'), ('EDN1', 'EGFR'), ('GDF15', 'ERBB2'), ('HLA-DMB', 'CD4'),\n",
    "    ('HLA-DQA2', 'LAG3'), ('HLA-DQA2', 'CD4'), ('HLA-DQB2', 'LAG3'), ('HLA-DRB5', 'LAG3'),\n",
    "    ('HLA-DRB5', 'CD4'), ('ICAM1', 'EGFR'), ('ICAM1', 'IL2RA'), ('ICAM1', 'ITGB2'),\n",
    "    ('IGF1', 'ITGB4'), ('IL1A', 'IL1R2'), ('IL1RN', 'IL1R2'), ('IL23A', 'IL23R'),\n",
    "    ('IL23A', 'LILRB2'), ('IL33', 'IL1RL1'), ('ITGB2', 'ICAM1'), ('ITGB2', 'THY1'),\n",
    "    ('KNG1', 'ITGB2'), ('KNG1', 'CD93'), ('LAMC2', 'ITGB4'), ('LPL', 'SDC1'),\n",
    "    ('MMRN2', 'CD93'), ('MMRN2', 'CLEC14A'), ('PECAM1', 'PECAM1'), ('POSTN', 'ITGB5'),\n",
    "    ('PRG4', 'TLR2'), ('PRG4', 'TLR4'), ('PTN', 'SDC1'), ('PTPRC', 'MRC1'),\n",
    "    ('PTPRC', 'CD247'), ('PTPRC', 'CD4'), ('RETN', 'TLR4'), ('S100A12', 'TLR4'),\n",
    "    ('SAA1', 'FPR1'), ('SAA1', 'TLR2'), ('SLAMF7', 'SLAMF7'), ('SLPI', 'CD4'),\n",
    "    ('TFF2', 'CXCR4'), ('THBS2', 'SDC1'), ('THY1', 'ITGB2'), ('TNC', 'SDC1'),\n",
    "    ('TNC', 'EGFR'), ('VCAN', 'EGFR'), ('VCAN', 'SELL'), ('VCAN', 'TLR2')\n",
    "]\n",
    "\n",
    "# Create a DataFrame that only contains the ligand-receptor pairs you are interested in\n",
    "heatmap_data = []\n",
    "\n",
    "for ligand, receptor in ligand_receptor_pairs:\n",
    "    if ligand in df['Ligand Name'].values and receptor in df.columns:\n",
    "        value = df.loc[df['Ligand Name'] == ligand, receptor].values[0]\n",
    "        heatmap_data.append([ligand, receptor, value])\n",
    "\n",
    "# Convert to DataFrame\n",
    "heatmap_df = pd.DataFrame(heatmap_data, columns=['Ligand', 'Receptor', 'Interaction'])\n",
    "\n",
    "# Pivot the DataFrame to get ligands as rows and receptors as columns\n",
    "heatmap_df_pivot = heatmap_df.pivot(index='Ligand', columns='Receptor', values='Interaction')\n",
    "\n",
    "# Plot the heatmap\n",
    "plt.figure(figsize=(25, 20))\n",
    "sns.heatmap(heatmap_df_pivot, cmap='coolwarm', annot=True, fmt='.2f', cbar=True)\n",
    "plt.title('Ligand-Receptor Interaction Heatmap')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04665a10-deb7-49c8-8597-7ecb42e3cb1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List samples where 'status.3' is 'perio'\n",
    "healthy_samples = adata.obs.loc[adata.obs['status.3'] == 'healthy', 'sample'].unique().tolist()\n",
    "print(healthy_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "278adf2a-8a78-4d26-9756-76781866c14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List samples where 'status.3' is 'perio'\n",
    "perio_samples = adata.obs.loc[adata.obs['status.3'] == 'perio', 'sample'].unique().tolist()\n",
    "print(perio_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26efb691-93d9-4ba0-853f-4b8e531caeb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Directory containing the subfolders\n",
    "base_dir = \"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Neighbors/\"\n",
    "\n",
    "# Ligand-receptor pairs of interest\n",
    "ligand_receptor_pairs = [\n",
    "    ('ANPEP', 'SELE'),  ('ANXA1', 'EGFR'),  ('ANXA1', 'FPR1'),  ('APOA5', 'LPL'),   ('C3', 'CD19'),   ('C3', 'ITGB2'),  ('C3', 'VSIG4'),\n",
    "    ('CAMP', 'EGFR'),  ('CCL19', 'CCR7'),  ('CCL5', 'ACKR1'),  ('CCL5', 'SDC1'),  ('CCN1', 'TLR2'),  ('CCN1', 'ITGB5'),  ('CCN1', 'TLR4'),\n",
    "    ('CCN1', 'ITGB2'), ('CD274', 'PDCD1'), ('CD28', 'CD86'), ('CD28', 'CD247'), ('CD28', 'CD4'), ('CD34', 'SELL'), ('CD34', 'SELE'),\n",
    "    ('CD70', 'CD27'), ('CD86', 'CD28'), ('CD86', 'CTLA4'),  ('CFH', 'CFB'),  ('CFH', 'SELL'),  ('COL16A1', 'ITGA1'),  ('COL17A1', 'ITGA1'),\n",
    "    ('COL5A2', 'SDC1'), ('COL5A2', 'CD93'), ('COL5A2', 'ITGA1'), ('COL8A1', 'ITGA1'), ('CSF3', 'CSF3R'), ('CSF3', 'CSF2RA'),  ('CXCL1', 'ACKR1'),\n",
    "    ('CXCL10', 'ACKR1'), ('CXCL12', 'CXCR4'), ('CXCL12', 'CD4'), ('CXCL14', 'CXCR4'), ('CXCL2', 'ACKR1'), ('CXCL5', 'ACKR1'),  ('CXCL6', 'ACKR1'),\n",
    "    ('CXCL9', 'ACKR1'), ('EDN1', 'EDNRB'), ('EDN1', 'ADGRL4'), ('EDN1', 'EGFR'), ('GDF15', 'ERBB2'), ('HLA-DMB', 'CD4'),  ('HLA-DQA2', 'LAG3'),\n",
    "    ('HLA-DQA2', 'CD4'), ('HLA-DQB2', 'LAG3'), ('HLA-DRB5', 'LAG3'), ('HLA-DRB5', 'CD4'), ('ICAM1', 'EGFR'), ('ICAM1', 'IL2RA'),  ('ICAM1', 'ITGB2'),\n",
    "    ('IGF1', 'ITGB4'),  ('IL1A', 'IL1R2'), ('IL1RN', 'IL1R2'), ('IL23A', 'IL23R'), ('IL23A', 'LILRB2'), ('IL33', 'IL1RL1'),  ('ITGB2', 'ICAM1'),\n",
    "    ('ITGB2', 'THY1'),  ('KNG1', 'ITGB2'),  ('KNG1', 'CD93'),  ('LAMC2', 'ITGB4'),  ('LPL', 'SDC1'),  ('MMRN2', 'CD93'),  ('MMRN2', 'CLEC14A'),\n",
    "    ('PECAM1', 'PECAM1'),  ('POSTN', 'ITGB5'),  ('PRG4', 'TLR2'),  ('PRG4', 'TLR4'),  ('PTN', 'SDC1'),  ('PTPRC', 'MRC1'),  ('PTPRC', 'CD247'),\n",
    "    ('PTPRC', 'CD4'),  ('RETN', 'TLR4'),  ('S100A12', 'TLR4'),  ('SAA1', 'FPR1'),  ('SAA1', 'TLR2'),  ('SLAMF7', 'SLAMF7'),  ('SLPI', 'CD4'),\n",
    "    ('TFF2', 'CXCR4'),  ('THBS2', 'SDC1'), ('THY1', 'ITGB2'), ('TNC', 'SDC1'),  ('TNC', 'EGFR'),  ('VCAN', 'EGFR'),  ('VCAN', 'SELL'),  ('VCAN', 'TLR2')\n",
    "]\n",
    "\n",
    "# Define your custom sample order\n",
    "custom_sample_order = [\n",
    "    's1r2', 's1r6', 's2r1_HV203', 's2r1_HV207', 's2r7_a', 's2r7_b', 's2r9_a', 's2r9_b', \n",
    "    'HV171A', 'HV191A_A', 'HV191A_B', 'HV205A', 'HV205B', 'HV220A_A', 'HV220A_B',\n",
    "    's1r1_a', 's1r1_b', 's1r3', 's1r5', 's2r2_HV137', 's2r2_HV184', 's2r3', 's2r4_a', 's2r4_b', \n",
    "    's2r5', 's2r8_HV185', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', 'HV137A', 'HV140A_A', \n",
    "    'HV140A_B', 'HV140A_C', 'HV140A_D', 'HV160C_A', 'HV160C_B', 'HV192A_A', 'HV192A_B', 'HV192A_C', 'HV198B_A', 'HV198B_B', \n",
    "    'HV219AB_A', 'HV219AB_B', 'HV219AS_A', 'HV219AS_B', 'HV219AS_C'\n",
    "]\n",
    "\n",
    "# Initialize a results DataFrame with ligand-receptor combinations as the index\n",
    "index_pairs = [f\"{ligand}-{receptor}\" for ligand, receptor in ligand_receptor_pairs]\n",
    "results = pd.DataFrame(index=index_pairs)\n",
    "\n",
    "# Collect sample data in a dictionary for sorting later\n",
    "sample_data_dict = {}\n",
    "\n",
    "for subfolder_name in os.listdir(base_dir):\n",
    "    subfolder_path = os.path.join(base_dir, subfolder_name)\n",
    "    \n",
    "    if os.path.isdir(subfolder_path):\n",
    "        # Find any CSV file in the subfolder\n",
    "        csv_files = [f for f in os.listdir(subfolder_path) if f.endswith(\".csv\")]\n",
    "        \n",
    "        if csv_files:\n",
    "            sample_csv = os.path.join(subfolder_path, csv_files[0])  # Use the first CSV found\n",
    "            \n",
    "            # Read the CSV file (including \"Ligand Cell Count\")\n",
    "            data = pd.read_csv(sample_csv, index_col=0)\n",
    "            \n",
    "            # Check if the \"Ligand Cell Count\" is at least 100\n",
    "            if data['Ligand Cell Count'].max() >= 100:\n",
    "                # Prepare data for the current sample\n",
    "                sample_data = {}\n",
    "\n",
    "                # For each ligand-receptor pair, check if the ligand cell count is >= 100\n",
    "                for ligand, receptor in ligand_receptor_pairs:\n",
    "                    # Ensure both ligand and receptor exist\n",
    "                    count = (\n",
    "                        data.loc[ligand, receptor] if ligand in data.index and receptor in data.columns else pd.NA\n",
    "                    )\n",
    "                    # If the ligand cell count is less than 100, set the count to NaN (pd.NA)\n",
    "                    if data.loc[ligand, 'Ligand Cell Count'] < 100:\n",
    "                        count = pd.NA\n",
    "\n",
    "                    sample_data[f\"{ligand}-{receptor}\"] = count\n",
    "\n",
    "                sample_data_dict[subfolder_name] = sample_data\n",
    "\n",
    "# Add the data in the custom order\n",
    "for sample_name in custom_sample_order:\n",
    "    if sample_name in sample_data_dict:\n",
    "        results[sample_name] = pd.Series(sample_data_dict[sample_name])\n",
    "\n",
    "# Save the output to an Excel file\n",
    "output_path = \"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Ligand-Receptor/Ligand_Neighbors-per-sample/Ligand-neighbors-per-sample.csv\"\n",
    "results.to_csv(output_path, index=True)\n",
    "print(f\"Data saved successfully to {output_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c128a6c-4e63-4560-a2ae-4c1ea41596bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Reset to default Matplotlib style\n",
    "plt.style.use(\"default\")\n",
    "\n",
    "# Replace problematic values and ensure mask is boolean\n",
    "results = results.apply(pd.to_numeric, errors='coerce')\n",
    "nan_mask = results.isna()\n",
    "\n",
    "plt.figure(figsize=(30, 20))\n",
    "\n",
    "sns.heatmap(\n",
    "    results,\n",
    "    cmap=\"YlGnBu\",\n",
    "    annot=True,\n",
    "    fmt=\".2f\",\n",
    "    vmax=20,\n",
    "    cbar_kws={'label': 'Interaction Count'},\n",
    "    mask=nan_mask,\n",
    "    linewidths=0.5,\n",
    "    linecolor=\"white\"\n",
    ")\n",
    "\n",
    "grey_patch = plt.Line2D([0], [0], marker='o', color='w', label='NaN',\n",
    "                         markerfacecolor='lightgrey', markersize=10)\n",
    "plt.legend(handles=[grey_patch], loc='upper right')\n",
    "\n",
    "plt.title(\"Ligand-Receptor Interaction Heatmap\")\n",
    "plt.xlabel(\"Samples\")\n",
    "plt.ylabel(\"Ligand-Receptor Pairs\")\n",
    "plt.xticks(rotation=90)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3265f2f-ffda-4143-b989-030f0d05be47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Healthy samples\n",
    "healthy_samples = [\n",
    "      's1r2', 's1r6', 's2r1_HV203', 's2r1_HV207', 's2r7_a', 's2r7_b', 's2r9_a', 's2r9_b', \n",
    "    'HV171A', 'HV191A_A', 'HV191A_B', 'HV205A', 'HV205B', 'HV220A_A', 'HV220A_B',  \n",
    "]\n",
    "\n",
    "# Perio samples\n",
    "perio_samples = [\n",
    "     's1r1_a', 's1r1_b', 's1r3', 's1r5', 's2r2_HV137', 's2r2_HV184', 's2r3', 's2r4_a', 's2r4_b', \n",
    "    's2r5', 's2r8_HV185', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', 'HV137A', 'HV140A_A', \n",
    "    'HV140A_B', 'HV140A_C', 'HV140A_D', 'HV160C_A', 'HV160C_B', 'HV192A_A', 'HV192A_B', 'HV192A_C', 'HV198B_A', 'HV198B_B', \n",
    "    'HV219AB_A', 'HV219AB_B', 'HV219AS_A', 'HV219AS_B', 'HV219AS_C'  # Replace with your actual sample names\n",
    "]\n",
    "\n",
    "# Output the lists\n",
    "print(\"Healthy Samples:\", healthy_samples)\n",
    "print(\"Perio Samples:\", perio_samples)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "570efeed-da89-4466-943c-1c67696e694b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming results_grouped is your DataFrame, and the columns are your samples\n",
    "healthy_data = results[healthy_samples]\n",
    "perio_data = results[perio_samples]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6ebfdd1-3df7-4261-babf-ab32089f8390",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "import pandas as pd\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Healthy samples\n",
    "healthy_samples = [\n",
    "    's1r2', 's1r6', 's2r1_HV203', 's2r1_HV207', 's2r7_a', 's2r7_b', 's2r9_a', 's2r9_b', \n",
    "    'HV171A', 'HV191A_A', 'HV191A_B', 'HV205A', 'HV205B', 'HV220A_A', 'HV220A_B',  \n",
    "]\n",
    "\n",
    "# Perio samples\n",
    "perio_samples = [\n",
    "    's1r1_a', 's1r1_b', 's1r3', 's1r5', 's2r2_HV137', 's2r2_HV184', 's2r3', 's2r4_a', 's2r4_b', \n",
    "    's2r5', 's2r8_HV185', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', 'HV137A', 'HV140A_A', \n",
    "    'HV140A_B', 'HV140A_C', 'HV140A_D', 'HV160C_A', 'HV160C_B', 'HV192A_A', 'HV192A_B', 'HV192A_C', 'HV198B_A', 'HV198B_B', \n",
    "    'HV219AB_A', 'HV219AB_B', 'HV219AS_A', 'HV219AS_B', 'HV219AS_C'  # Replace with your actual sample names\n",
    "]\n",
    "\n",
    "# Assuming `results` is the DataFrame that contains the ligand-receptor expression data\n",
    "# Extract Healthy and Perio data columns using .loc or directly by column names\n",
    "healthy_data = results[healthy_samples]  # Accessing columns based on the list\n",
    "perio_data = results[perio_samples]  # Same here\n",
    "\n",
    "# Initialize an empty list to store results\n",
    "stats_results = []  # Renamed to avoid conflict with the DataFrame\n",
    "\n",
    "# Loop through each row (ligand-receptor pair) in the DataFrame\n",
    "for index, row in results.iterrows():  # 'index' is the ligand-receptor pair\n",
    "    # Extract the expression values for the current ligand-receptor pair\n",
    "    healthy_expression = healthy_data.loc[index]  # Get Healthy expression values\n",
    "    perio_expression = perio_data.loc[index]  # Get Perio expression values\n",
    "\n",
    "    # Perform t-test to compare Healthy and Perio samples\n",
    "    t_stat, p_value = stats.ttest_ind(healthy_expression, perio_expression)\n",
    "\n",
    "    # Store the result in the stats_results list\n",
    "    stats_results.append({\n",
    "        'Ligand-Receptor Pair': index,  # Store ligand-receptor pair name\n",
    "        'T-statistic': t_stat,\n",
    "        'P-value': p_value\n",
    "    })\n",
    "\n",
    "# Convert stats_results list to a DataFrame\n",
    "stats_results_df = pd.DataFrame(stats_results)\n",
    "\n",
    "# Correct p-values using Bonferroni correction (for multiple comparisons)\n",
    "stats_results_df['Corrected P-value'] = multipletests(stats_results_df['P-value'], method='bonferroni')[1]\n",
    "\n",
    "# Print the results DataFrame\n",
    "print(stats_results_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb5ea253-2e4e-413e-9c42-457d302cc862",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Plot p-values for each ligand-receptor pair\n",
    "plt.figure(figsize=(20, 8))\n",
    "sns.barplot(x='Ligand-Receptor Pair', y='Corrected P-value', data=stats_results_df)\n",
    "plt.xticks(rotation=90)\n",
    "plt.title('P-values for Ligand-Receptor Pairs: Healthy vs Perio')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f996bcac-945d-4585-931e-41176f10d1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1ef2fdb-6410-47d1-8a5d-d18ce7ca6eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ligand_table_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d190bef-f5c8-453c-9b1e-4b1750fb862a",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = sc.read_h5ad('/data/vasileiosionat2/Xenium/Integration/NEW_CITE_ seq_files_ 24.12.23/Xenium_citeSeq_harmony.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0de052b6-277b-4178-955f-efd2fb6dec0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['Lvl4'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d304242-da5a-442d-99b6-a7d21cffa3d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Filter Xenium assay cells from the merged object\n",
    "xen_obj = merged[merged.obs['assay'] == 'Xenium'].copy()  # Ensure a copy to avoid warnings\n",
    "\n",
    "# Define the Lvl4 column as string type\n",
    "xen_obj.obs['Lvl4'] = xen_obj.obs['Lvl4'].astype('str')\n",
    "\n",
    "# Compute cell counts for all categories without filtering\n",
    "celltype_counts = pd.DataFrame(xen_obj.obs.groupby(['Lvl4', 'citeSeq_to_Xenium_label']).size()).unstack()\n",
    "celltype_counts.columns = celltype_counts.columns.droplevel()\n",
    "celltype_counts.index.name = 'Xenium cell type'\n",
    "celltype_counts.columns.name = 'Predicted citeSeq cell type'\n",
    "celltype_counts = celltype_counts.T\n",
    "\n",
    "# Normalize row-wise and handle NaNs\n",
    "celltype_counts = celltype_counts.div(celltype_counts.sum(axis=1), axis=0).fillna(0)\n",
    "\n",
    "# Specify clusters to plot (leave analysis intact)\n",
    "# Define filters\n",
    "lvl4_clusters_to_plot = ['B', 'Pl.1', 'Pl.2', 'PB', 'Th', 'Th_proximity_to_B', 'Treg', 'Tc', 'NK', 'T.IL7Rhi', 'T_proximity_to_Fib', 'T.IE',\n",
    "                'Mac', 'Mast', 'Mono', 'Mac.Neut.Mix', 'cDC1', 'Lang', 'cDC2', 'mregDC', 'pDC']  # Replace with your desired Lvl4 categories\n",
    "citeseq_clusters_to_plot = ['B Atypical', 'B Memory 1', 'B Memory 2', 'B Naive', 'Plasma 1', 'Plasma 2', \n",
    "                   'Plasma 3', 'Plasmablast', 'CD4 IL17A+ FOXP3+', 'CD4 T naive', 'CD4 TCM', 'CD4 TFH',\n",
    "                  'CD4 TFreg', 'CD4 TH17', 'CD4 Treg', 'CD8 T', 'CD8 T IKZF2+', 'CD8 T TNF+', 'CD8 T naive', 'CD8 TRM', 'CD16+ NK', 'NK', 'T gd',\n",
    "                  'ILC', 'MAIT', 'Macrophage', 'Mast', 'Monocyte', 'Monocyte/Macrophage', 'Neutrophil', 'cDC1', 'cDC2', 'mregDC', 'pDC']  # Replace with your desired citeSeq labels\n",
    "\n",
    "\n",
    "# Filter for plotting only\n",
    "filtered_counts = celltype_counts.loc[\n",
    "    citeseq_clusters_to_plot, lvl4_clusters_to_plot\n",
    "]\n",
    "\n",
    "# Plot heatmap for selected clusters\n",
    "plt.figure(figsize=(12, 10))\n",
    "sns.heatmap(filtered_counts, cmap='YlGnBu', fmt=\".2f\", cbar=True)\n",
    "plt.title(\"Xenium vs Predicted scRNAseq Heatmap (Selected Clusters)\")\n",
    "# Save the plot as TIFF file\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_Citeseq_clusters.svg', format='svg')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a230311a-1713-41f0-a442-90959bda4846",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11851cc6-12b7-4a48-ac74-c5c2dff8d1c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_cc14'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01c299e3-f85f-427a-ab28-2073e2942270",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl4'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e024990-54aa-47c8-b9d3-cdef9ea9bdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = ['#e1e1e1', 'yellow', '#7F00FF', 'red', 'magenta']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=3,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/niches.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': 'red', 'Plasma': '#7F00FF', 'Crevicular': 'yellow', 'Plasma-Fib CT' : 'magenta'}\n",
    "lvl1_cluster_key = 'niche_knn50k10_merged'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bcbd588-183d-483d-8305-d09da93c50e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = ['#e1e1e1', 'yellow', '#7F00FF', 'red']\n",
    "        alpha_value = 0.1  # Transparency level (adjust as needed)\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                alpha=alpha_value,  # Set transparency\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/niches_faint.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': 'red', 'Plasma': '#7F00FF', 'Crevicular': 'yellow'}\n",
    "lvl1_cluster_key = 'niche_cc14'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5807bc-f07b-4569-a5f8-a56ce257fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = ['#e1e1e1', '#FFFDD0', 'purple', '#F88379']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/niches_faint.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': '#F88379', 'Plasma': 'purple', 'Crevicular': '#FFFDD0'}\n",
    "lvl1_cluster_key = 'niche_cc14'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3cdec5-2741-485b-9ec8-3e5b4748019b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#808080')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#808080', '#0000FF',  'red', 'magenta', '#39e75f', 'cyan',]\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_B-Pl.svg'\n",
    "lvl1_cluster_colors = {'B' : 'magenta', 'pDC' : 'cyan', 'PB': 'red', 'Pl.1': '#0000FF', 'Pl.2': '#0000FF', 'Fib.2': '#39e75f'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a92e50de-0a32-4a01-a689-6c72cd4d5456",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#808080')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = ['#808080', 'yellow', 'purple', '#008000', 'red', 'cyan']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_T-Fib.svg'\n",
    "lvl1_cluster_colors = {'Treg' : 'red', 'mregDC' : 'cyan', 'Tc': 'yellow', 'Th_proximity_to_B': 'purple'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba45220a-f75c-40dc-9a94-fd8285ec57b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742e1a07-6983-4daf-bbbb-442a05afe900",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['citeSeq_to_Xenium_label'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cdea4f1-fafb-43de-aa2b-c55924eed2e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#808080')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = ['#808080', 'yellow', 'purple', 'red', 'orange', 'cyan', 'magenta']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Citeseq_T-Fib.svg'\n",
    "lvl1_cluster_colors = {'CD4 TFreg' : 'orange', 'mregDC' : 'cyan', 'CD4 IL17A+ FOXP3+': 'purple',  'CD4 T naive': 'purple','CD4 Treg': 'red', 'CD4 TCM': 'purple', 'CD4 T naive': 'purple', 'CD8 T': 'yellow', 'CD8 T IKZF2+': 'yellow', 'CD8 T naive': 'yellow',\n",
    "                       'CD8 T TNF+': 'yellow', 'CD8 TRM': 'yellow' , 'CD4 TFH': 'magenta'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f2210e-e209-4e2b-8955-6e2a78c4f7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Fib.2.svg'\n",
    "lvl1_cluster_colors = {'Fibroblast 2' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b82d975a-e093-406f-b423-38d84cf1b3ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/TFH.svg'\n",
    "lvl1_cluster_colors = {'CD4 TFH' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06b5d1b1-1297-454d-b3d9-1eb119872527",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/TFReg.svg'\n",
    "lvl1_cluster_colors = {'CD4 TFreg' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f070423f-c62f-4ff0-8c18-c93bd6a10098",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/mregDC.svg'\n",
    "lvl1_cluster_colors = {'mregDC' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "963df53f-8678-4f30-bcf7-45c5a46b5762",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Plasma.svg'\n",
    "lvl1_cluster_colors = {'Plasma 1' : '#39e75f', 'Plasma 2' : '#39e75f', 'Plasma 3' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "febc2e00-e41a-45f5-8127-64be5069c00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/PB.svg'\n",
    "lvl1_cluster_colors = {'Plasmablast' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1920e18b-59b6-4e52-a9cc-b2c24b5da314",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Bnaive.svg'\n",
    "lvl1_cluster_colors = {'B Naive' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab73cf3e-7019-418c-a995-310261bd68a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Bmem.svg'\n",
    "lvl1_cluster_colors = {'B Memory 1' : '#39e75f', 'B Memory 2' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea07b77c-c0dc-4bae-b8fd-9942f3fe6d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Batyp.svg'\n",
    "lvl1_cluster_colors = {'B Atypical' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a8d2287-8b9e-4bd1-8c7f-1b9e1db7a880",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/pDC.svg'\n",
    "lvl1_cluster_colors = {'pDC' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bd2da20-c462-4108-a1e2-d64358ae67bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['citeSeq_to_Xenium_label'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b90e677a-ea5d-4c7a-b1f5-536a0b17acd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = merged[merged.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', '#39e75f']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/cMyel.svg'\n",
    "lvl1_cluster_colors = {'Cycling myeloid' : '#39e75f'}\n",
    "lvl1_cluster_key = 'citeSeq_to_Xenium_label'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d40f5680-507c-4471-b7d5-197da0016167",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['citeSeq_to_Xenium_label'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec861039-b73a-4b7e-b567-4dd9f3dcd1c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter Xenium data\n",
    "xen_obj = merged[merged.obs['assay'] == 'Xenium']\n",
    "xen_obj.obs['Lvl4'] = xen_obj.obs['Lvl4'].astype('str')\n",
    "\n",
    "# Group by Xenium and predicted CITE-seq cell types, count occurrences\n",
    "celltype_counts = pd.DataFrame(\n",
    "    xen_obj.obs.groupby(['Lvl4', 'citeSeq_to_Xenium_label']).size()\n",
    ").unstack()\n",
    "\n",
    "# Clean DataFrame\n",
    "celltype_counts.columns = celltype_counts.columns.droplevel()\n",
    "celltype_counts.index.name = 'Xenium cell type'\n",
    "celltype_counts.columns.name = 'Predicted CITE-seq cell type'\n",
    "celltype_counts = celltype_counts.T\n",
    "\n",
    "# Normalize rows by their sums\n",
    "celltype_counts = celltype_counts.div(celltype_counts.sum(axis=1), axis=0)\n",
    "celltype_counts = celltype_counts.loc[:, celltype_counts.idxmax(axis=0).sort_values().index].fillna(0)\n",
    "\n",
    "# Compute column sums to identify significant columns\n",
    "column_sums = celltype_counts.abs().sum(axis=0)\n",
    "\n",
    "# Threshold for important Xenium clusters\n",
    "columns_to_keep = column_sums[column_sums >= 0.1].index\n",
    "celltype_counts2 = celltype_counts[columns_to_keep]\n",
    "\n",
    "# Specify desired CITE-seq clusters to visualize on the y-axis\n",
    "desired_cite_clusters = ['CD4 TFreg', 'mregDC', 'CD4 TFH', 'B Naive', 'B Memory 1', 'B Memory 2', 'Plasma 1', 'Plasma 2', 'Plasma 3', \n",
    "                         'Plasmablast', 'pDC', 'B Atypical']  # Replace with your actual cluster names\n",
    "celltype_counts2 = celltype_counts2.loc[celltype_counts2.index.intersection(desired_cite_clusters)]\n",
    "\n",
    "# Exclude Xenium Lvl4 clusters starting with \"VEC\" or \"Ep\" on the x-axis\n",
    "exclude_patterns = ('VEC', 'Ep', 'T.', 'T_', 'F', 'Tc', 'L', 'Tr', 'NK', 'SM', 'M', 'cD')\n",
    "columns_to_exclude = [col for col in celltype_counts2.columns if col.startswith(exclude_patterns)]\n",
    "celltype_counts2 = celltype_counts2.drop(columns=columns_to_exclude)\n",
    "\n",
    "# Plot heatmap\n",
    "plt.figure(figsize=(5, 4))\n",
    "sns.heatmap(celltype_counts2, cmap='YlGnBu', annot=False, fmt=\".2f\", linewidths=.5, \n",
    "            cbar_kws={'label': 'Co-occurrence Frequency'})\n",
    "plt.title(\"Filtered Co-occurrence Frequencies Between Xenium and Predicted CITE-seq Clusters\")\n",
    "plt.xlabel(\"Xenium Cell Types (Lvl4)\")\n",
    "plt.ylabel(\"Predicted CITE-seq Cell Types\")\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "# Save the plot\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/citeseq-to-xen.svg'\n",
    "fig = plt.gcf()  # Get the current figure\n",
    "plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a86591c-4d7c-42d0-af81-58c91d3460e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/ThB.svg'\n",
    "lvl1_cluster_colors = {'Th_proximity_to_B' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74ee56ca-cc5a-495b-8d96-ee42092d4d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Th.svg'\n",
    "lvl1_cluster_colors = {'Th_proximity_to_B' : 'black', 'Th' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03840a2c-bb87-4253-ae1c-7cb4bf35f260",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Tc.svg'\n",
    "lvl1_cluster_colors = {'Tc' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a9193af-af7a-4a29-a41b-6fd7ba14ec08",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/B.svg'\n",
    "lvl1_cluster_colors = {'B' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ffa3f9-7153-49a3-ab44-7bbcc21539e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/mregDC.svg'\n",
    "lvl1_cluster_colors = {'mregDC' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7658b3a-4ffd-4375-9969-de0eb46e46b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Pl.svg'\n",
    "lvl1_cluster_colors = {'Pl.1' : 'black', 'Pl.2' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "246c2a51-8fcf-42aa-b222-c428ffca2d61",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/PB.svg'\n",
    "lvl1_cluster_colors = {'PB' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "319ad043-2b18-4bb0-8e94-38a1a7fa9120",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Fib2.svg'\n",
    "lvl1_cluster_colors = {'Fib.2' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4398a659-df01-458e-9562-04374978a102",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/pDC.svg'\n",
    "lvl1_cluster_colors = {'pDC' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90eaa911-9e61-4c12-8b81-c6361fad1625",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'black']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Treg.svg'\n",
    "lvl1_cluster_colors = {'Treg' : 'black'}\n",
    "lvl1_cluster_key = 'Lvl4'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06bf6bdb-22b3-4f8c-86f5-9d8ba3c6c264",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'yellow', 'red', '#7F00FF']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/niches.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': 'red', 'Plasma': '#7F00FF', 'Crevicular': 'yellow'}\n",
    "lvl1_cluster_key = 'niche_cc14'\n",
    "generate_spatial_plot(adata, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9b1a683-0a5f-448b-a10d-21428119f54a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_perio = adata[adata.obs['status.3'] == \"perio\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96115cf4-39e1-4207-8a0f-bb4c97b9ad05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata_perio[adata_perio.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'yellow', 'red', '#7F00FF', 'magenta']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=8,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        #plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/niches.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': 'red', 'Plasma': '#7F00FF', 'Crevicular': 'yellow', 'Plasma-Fib CT': 'magenta'}\n",
    "lvl1_cluster_key = 'perio_niche_knn50k10'\n",
    "generate_spatial_plot(adata_perio, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8e6b5c-96be-4d4c-a8ec-a6193dd52107",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, cluster_colors, output_path):\n",
    "    # Define rotation angle in degrees\n",
    "    rotation_angle = 90\n",
    "    rotation_rad = np.deg2rad(rotation_angle)\n",
    "    \n",
    "    # Rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(rotation_rad), -np.sin(rotation_rad)],\n",
    "        [np.sin(rotation_rad), np.cos(rotation_rad)]\n",
    "    ])\n",
    "    \n",
    "    # Loop over each sample to generate a spatial scatter plot\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata_perio[adata_perio.obs['sample'] == sample_name]\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Apply rotation transformation\n",
    "        coords = np.vstack((x_coords, y_coords))\n",
    "        transformed_coords = rotation_matrix @ coords\n",
    "        new_x_coords = transformed_coords[0, :]\n",
    "        new_y_coords = transformed_coords[1, :]\n",
    "\n",
    "        # Calculate aspect ratio\n",
    "        x_range = new_x_coords.max() - new_x_coords.min()\n",
    "        y_range = new_y_coords.max() - new_y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        plt.figure(figsize=figsize)\n",
    "        fig = plt.gcf()\n",
    "        ax = plt.gca()\n",
    "        fig.patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Map cluster colors, set unmapped categories to grey\n",
    "        subset_adata.obs[f'{cluster_key}_color'] = subset_adata.obs[cluster_key].map(cluster_colors).fillna('#e1e1e1')\n",
    "\n",
    "        # Color priority order: Yellow -> Blue -> Green\n",
    "        color_order = [ '#e1e1e1', 'yellow', 'red', '#7F00FF', 'magenta']\n",
    "        for color in color_order:\n",
    "            mask = subset_adata.obs[f'{cluster_key}_color'] == color\n",
    "            plt.scatter(\n",
    "                x=new_x_coords[mask],\n",
    "                y=new_y_coords[mask],\n",
    "                c=color,\n",
    "                s=10,\n",
    "                label=color\n",
    "            )\n",
    "\n",
    "        # Remove grids, ticks, and spines\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "\n",
    "        # Add a 200 μm scale bar\n",
    "        scale_bar_length = 200\n",
    "        scale_bar_x_start = new_x_coords.min() + 0.01 * (new_x_coords.max() - new_x_coords.min())\n",
    "        scale_bar_x_end = scale_bar_x_start + scale_bar_length\n",
    "        scale_bar_y = new_y_coords.min() + 0.3 * (new_y_coords.max() - new_y_coords.min())\n",
    "        ax.plot(\n",
    "            [scale_bar_x_start, scale_bar_x_end],\n",
    "            [scale_bar_y, scale_bar_y],\n",
    "            color='black',\n",
    "            linewidth=8\n",
    "        )\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s2r2_HV184']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_niches.svg'\n",
    "lvl1_cluster_colors = {'Lymphoid': 'red', 'Plasma': '#7F00FF', 'Crevicular': 'yellow', 'Plasma-Fib CT': 'magenta'}\n",
    "lvl1_cluster_key = 'perio_niche_knn50k10'\n",
    "generate_spatial_plot(adata_perio, sample_names, lvl1_cluster_key, lvl1_cluster_colors, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2282a1d1-bcaa-4b73-b3a2-8a7aaab805bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(df['perio_niche_knn50k10'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2f7c0fa-961b-4b6e-b856-e6ad3666c017",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_perio.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts for all clusters and niches\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Define cluster (y-axis) and niche (x-axis) orders\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'Mac', 'MyoF']\n",
    "niche_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']  # Niches to plot\n",
    "\n",
    "# **Normalize using all niches, not just the selected ones**\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across ALL niches per cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Normalize\n",
    "\n",
    "# **Filter only the selected niches for plotting**\n",
    "niche_groups_relative_filtered = niche_groups_relative[niche_order]\n",
    "\n",
    "# Reindex to ensure correct order\n",
    "niche_groups_relative_filtered = niche_groups_relative_filtered.reindex(index=cluster_order, columns=niche_order)\n",
    "\n",
    "# **Plot heatmap: Niches (x-axis) and Clusters (y-axis)**\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered,  # Pre-filtered for plotting\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'Relative Frequency'},\n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.5,\n",
    "    vmin=0.05,\n",
    "    xticklabels=niche_order,\n",
    "    yticklabels=cluster_order\n",
    ")\n",
    "plt.title('Relative Abundance of Cell Clusters Across Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    ".\n",
    "# Save output\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_clusters_niches.pdf'\n",
    "fig = plt.gcf()\n",
    "plt.savefig(output_path, dpi=300, format='pdf', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b886177-b4da-4651-9325-b12a8c377efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams.update(plt.rcParamsDefault)  # Reset to default settings\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures text is stored as an actual font\n",
    "plt.rcParams['ps.fonttype'] = 42  # If exporting to PostScript (PS) as well\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_healthy.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts for all clusters and niches\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Define cluster (y-axis) and niche (x-axis) orders\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'Treg', 'Tc', 'NK', 'T.IL7Rhi', \n",
    "                 'T_proximity_to_Fib', 'cDC1', 'cDC2', 'mregDC', 'pDC', 'Pl.1', 'Pl.2', 'PB', \n",
    "                 'Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', 'Mac', 'Mast', 'Mono', 'MyoF']\n",
    "niche_order = ['Epi-CT', 'Lymphoid', 'Plasma', 'Plasma-Fib CT', 'Fibrous CT']  # Niches to plot\n",
    "\n",
    "# **Normalize using all niches, not just the selected ones**\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across ALL niches per cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Normalize\n",
    "\n",
    "# **Filter only the selected niches for plotting**\n",
    "niche_groups_relative_filtered = niche_groups_relative[niche_order]\n",
    "\n",
    "# Reindex to ensure correct order\n",
    "niche_groups_relative_filtered = niche_groups_relative_filtered.reindex(index=cluster_order, columns=niche_order)\n",
    "\n",
    "# **Plot heatmap: Niches (x-axis) and Clusters (y-axis)**\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered,  # Pre-filtered for plotting\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'Relative Frequency'},\n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.5,\n",
    "    vmin=0.05,\n",
    "    xticklabels=niche_order,\n",
    "    yticklabels=cluster_order\n",
    ")\n",
    "plt.title('Relative Abundance of Cell Clusters Across Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save output\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_healthy_clusters_niches_suppl.pdf'\n",
    "fig = plt.gcf()\n",
    "plt.savefig(output_path, dpi=300, format='pdf', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb9c9f40-8328-458c-8b2c-94c8d68fb834",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams.update(plt.rcParamsDefault)  # Reset to default settings\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Ensures text is stored as an actual font\n",
    "plt.rcParams['ps.fonttype'] = 42  # If exporting to PostScript (PS) as well\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_perio.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts for all clusters and niches\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Define cluster (y-axis) and niche (x-axis) orders\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'Treg', 'Tc', 'NK', 'T.IL7Rhi', \n",
    "                 'T_proximity_to_Fib', 'cDC1', 'cDC2', 'mregDC', 'pDC', 'Pl.1', 'Pl.2', 'PB', \n",
    "                 'Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', 'Mac', 'Mast', 'Mono', 'MyoF']\n",
    "niche_order = ['Epi-CT', 'Lymphoid', 'Plasma', 'Plasma-Fib CT', 'Fibrous CT']  # Niches to plot\n",
    "\n",
    "# **Normalize using all niches, not just the selected ones**\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across ALL niches per cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Normalize\n",
    "\n",
    "# **Filter only the selected niches for plotting**\n",
    "niche_groups_relative_filtered = niche_groups_relative[niche_order]\n",
    "\n",
    "# Reindex to ensure correct order\n",
    "niche_groups_relative_filtered = niche_groups_relative_filtered.reindex(index=cluster_order, columns=niche_order)\n",
    "\n",
    "# **Plot heatmap: Niches (x-axis) and Clusters (y-axis)**\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered,  # Pre-filtered for plotting\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'Relative Frequency'},\n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.5,\n",
    "    vmin=0.05,\n",
    "    xticklabels=niche_order,\n",
    "    yticklabels=cluster_order\n",
    ")\n",
    "plt.title('Relative Abundance of Cell Clusters Across Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save output\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_clusters_niches_suppl.pdf'\n",
    "fig = plt.gcf()\n",
    "plt.savefig(output_path, dpi=300, format='pdf', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e7c1fb9-361e-4a09-b602-f5e437a181b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f36eb37f-6a6d-4b13-bf70-00b2903d023e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37f92551-2323-45c3-9239-7f243973556d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_healthy.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Manually specify which clusters (x-axis) and niches (y-axis) to show, with order\n",
    "selected_clusters = ['B', 'Th', 'Th_proximity_to_B', 'Treg', 'Tc', 'NK', 'T.IL7Rhi', \n",
    "                    'T_proximity_to_Fib', 'cDC1','cDC2', 'mregDC', 'pDC', 'Pl.1', 'Pl.2', 'PB', \n",
    "                    'Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', 'Mac', 'Mast', 'Mono', 'MyoF']\n",
    "selected_niches = ['Lymphoid', 'Plasma', 'Plasma-Fib CT', 'Fibrous CT']  # Replace with actual niche names\n",
    "\n",
    "# Ensure selected clusters and niches exist in the DataFrame\n",
    "niche_groups_filtered = niche_groups.loc[\n",
    "    niche_groups.index.intersection(selected_clusters),  # Use intersection() to filter selected clusters\n",
    "    niche_groups.columns.intersection(selected_niches)   # Use intersection() to filter selected niches\n",
    "]\n",
    "\n",
    "# Specify the order of clusters and niches\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'MyoF',\n",
    "                    'Mac', 'Mast']  # Adjust the order as needed\n",
    "\n",
    "niche_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']  # Adjust the order as needed\n",
    "\n",
    "# Plot heatmap with niches on y-axis and clusters on x-axis using raw counts\n",
    "plt.figure(figsize=(8, 3))  # Adjusted figsize for readability\n",
    "sns.heatmap(\n",
    "    niche_groups_filtered[niche_order].T[cluster_order], cmap='plasma',  # Apply niche_order on y-axis and cluster_order on x-axis\n",
    "    cbar_kws={'label': 'Cell Count'}, \n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=5000,\n",
    "    xticklabels=cluster_order,  # Apply cluster_order on x-axis\n",
    "    yticklabels=niche_order  # Apply niche_order on y-axis\n",
    ")\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Cell Cluster')\n",
    "plt.ylabel('Niche')\n",
    "plt.xticks(rotation=45, ha='right')  # Improved alignment\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Prevents clipping\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_clusters_niches.svg'\n",
    "fig = plt.gcf()  # Get current figure\n",
    "#plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed5b3499-2d2b-4664-89a3-8c750cd81322",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy = adata[adata.obs['status.3'] == \"healthy\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebde99ce-25fd-45be-afdd-c1fcb357b31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_knn50k10_merged'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "278b9561-5f62-42c9-a14f-5f18b882e658",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_healthy.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts for all clusters and niches\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Define cluster (y-axis) and niche (x-axis) orders\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'Mac', 'MyoF']\n",
    "niche_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']  # Niches to plot\n",
    "\n",
    "# **Normalize using all niches, not just the selected ones**\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across ALL niches per cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Normalize\n",
    "\n",
    "# **Filter only the selected niches for plotting**\n",
    "niche_groups_relative_filtered = niche_groups_relative[niche_order]\n",
    "\n",
    "# Reindex to ensure correct order\n",
    "niche_groups_relative_filtered = niche_groups_relative_filtered.reindex(index=cluster_order, columns=niche_order)\n",
    "\n",
    "# **Plot heatmap: Niches (x-axis) and Clusters (y-axis)**\n",
    "plt.figure(figsize=(4, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered,  # Pre-filtered for plotting\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'Relative Frequency'},\n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.5,\n",
    "    vmin=0.05,\n",
    "    xticklabels=niche_order,\n",
    "    yticklabels=cluster_order\n",
    ")\n",
    "plt.title('Relative Abundance of Cell Clusters Across Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save output\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_healthy_clusters_niches.pdf'\n",
    "fig = plt.gcf()\n",
    "plt.savefig(output_path, dpi=300, format='pdf', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62ee801-9089-4178-a9a6-e7e83d04741d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np  # You need this for np.log1p\n",
    "\n",
    "# Create DataFrame from adata.obs\n",
    "df = pd.DataFrame(adata_healthy.obs)\n",
    "df['niche_knn50k10_merged'] = df['niche_knn50k10_merged'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_knn50k10_merged', 'Lvl4']).size().unstack(fill_value=0)\n",
    "\n",
    "# Transpose the DataFrame\n",
    "niche_groups = niche_groups.T\n",
    "\n",
    "# Custom x-axis (niche) order\n",
    "custom_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT','Fibrous CT']\n",
    "niche_groups = niche_groups.reindex(columns=custom_order)\n",
    "\n",
    "# Custom y-axis (cell type) selection\n",
    "y_axis_clusters = ['B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'Mac', 'MyoF']\n",
    "niche_groups_filtered = niche_groups.loc[y_axis_clusters]\n",
    "\n",
    "# Transpose so clusters are columns (x-axis), and niches are rows (y-axis)\n",
    "niche_groups_flipped = niche_groups_filtered.T  # Now niches are rows, clusters are columns\n",
    "\n",
    "# Apply log1p transformation\n",
    "niche_groups_log = niche_groups_flipped.applymap(lambda x: np.log1p(x))\n",
    "\n",
    "# Plot heatmap\n",
    "plt.figure(figsize=(7, 5))\n",
    "sns.heatmap(\n",
    "    niche_groups_log,\n",
    "    vmax=np.log1p(1000), \n",
    "    vmin=np.log1p(50),       # log(0+1)\n",
    "    cmap='plasma',\n",
    "    cbar_kws={'label': 'log(Count + 1)'},\n",
    "    linewidths=0.05,\n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Adjust colorbar ticks (log scale)\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "log_ticks = [np.log1p(500), np.log1p(1000), np.log1p(10000)]\n",
    "cbar.set_ticks(log_ticks)\n",
    "cbar.ax.set_yticklabels(['500', '1,000', '10,000'])\n",
    "\n",
    "plt.tight_layout()\n",
    "output_path = f\"/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/IBEX_LV3_Clusters_Niches_Log.pdf\"\n",
    "#plt.savefig(output_path, format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd81ea12-c097-4796-99df-1cd0bbf7ef99",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(adata_healthy.obs)\n",
    "df['Lvl4'] = df['Lvl4'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['Lvl4', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Manually specify which clusters (x-axis) and niches (y-axis) to show, with order\n",
    "selected_clusters = ['B', 'Th', 'Th_proximity_to_B', 'Treg', 'Tc', 'NK', 'T.IL7Rhi', \n",
    "                    'T_proximity_to_Fib', 'cDC1','cDC2', 'mregDC', 'pDC', 'Pl.1', 'Pl.2', 'PB', \n",
    "                    'Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', 'Mac', 'Mast', 'Mono', 'MyoF']\n",
    "selected_niches = ['Lymphoid', 'Plasma', 'Plasma-Fib CT', 'Fibrous CT']  # Replace with actual niche names\n",
    "\n",
    "# Ensure selected clusters and niches exist in the DataFrame\n",
    "niche_groups_filtered = niche_groups.loc[\n",
    "    niche_groups.index.intersection(selected_clusters),  # Use intersection() to filter selected clusters\n",
    "    niche_groups.columns.intersection(selected_niches)   # Use intersection() to filter selected niches\n",
    "]\n",
    "\n",
    "# Specify the order of clusters and niches\n",
    "cluster_order = ['B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'MyoF',\n",
    "                    'Mac', 'Mast']  # Adjust the order as needed\n",
    "\n",
    "niche_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']  # Adjust the order as needed\n",
    "\n",
    "# Plot heatmap with niches on y-axis and clusters on x-axis using raw counts\n",
    "plt.figure(figsize=(8, 3))  # Adjusted figsize for readability\n",
    "sns.heatmap(\n",
    "    niche_groups_filtered[niche_order].T[cluster_order], cmap='plasma',  # Apply niche_order on y-axis and cluster_order on x-axis\n",
    "    cbar_kws={'label': 'Cell Count'}, \n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=700,\n",
    "    xticklabels=cluster_order,  # Apply cluster_order on x-axis\n",
    "    yticklabels=niche_order  # Apply niche_order on y-axis\n",
    ")\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Cell Cluster')\n",
    "plt.ylabel('Niche')\n",
    "plt.xticks(rotation=45, ha='right')  # Improved alignment\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Prevents clipping\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_perio_clusters_niches.svg'\n",
    "fig = plt.gcf()  # Get current figure\n",
    "#plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8154c20-50f5-4b5e-b1e1-4995895e64a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.read_h5ad('/data/vasileiosionat2/Xenium/Drake_outputs/ccProcessed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8326d212-eec8-4341-a68e-884f91b138ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "756e53e5-a8f2-4ddd-9740-09230e0eafed",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = sc.read_h5ad('/data/vasileiosionat2/Xenium/Integration_2025_only_stromal/Modified_citeseq_files/Xenium_citeSeq_total_harmony.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b445cf64-459c-4a8f-b4d5-423d070d8ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcb85c98-7e61-49ab-93cb-928582539f6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "\n",
    "# Assuming adata_perio and merged_perio are already loaded\n",
    "# Initialize a list to store the processed samples\n",
    "merged_samples = []\n",
    "\n",
    "# Get all unique samples\n",
    "samples = adata.obs['sample'].unique()\n",
    "\n",
    "for sample in samples:\n",
    "    # Subset both adata_perio and merged_perio based on the current sample\n",
    "    adata_sample_subset = adata[adata.obs['sample'] == sample]\n",
    "    merged_sample_subset = merged[merged.obs['sample'] == sample]\n",
    "\n",
    "    # Sort by the index (this ensures the rows align correctly)\n",
    "    adata_sample_subset = adata_sample_subset[adata_sample_subset.obs.sort_index().index]\n",
    "    merged_sample_subset = merged_sample_subset[merged_sample_subset.obs.sort_index().index]\n",
    "\n",
    "    # Transfer the columns from adata_perio to merged_perio\n",
    "    merged_sample_subset.obs['niche_knn50k10'] = adata_sample_subset.obs['niche_knn50k10']\n",
    "    merged_sample_subset.obs['niche_knn50k10_merged'] = adata_sample_subset.obs['niche_knn50k10_merged']\n",
    "\n",
    "    # Add the modified sample to the list\n",
    "    merged_samples.append(merged_sample_subset)\n",
    "\n",
    "# After processing all samples, concatenate them into one AnnData object\n",
    "combined = merged_samples[0].concatenate(*merged_samples[1:], join='outer')\n",
    "\n",
    "# Now, 'combined' holds the result with the transferred columns for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9fe4746-8c89-4e0e-873e-81d2b951bc62",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_perio  = combined[combined.obs['status.3'].str.contains('perio', na=False)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd6c3539-7905-4762-bf4a-25d693118d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_perio.obs['niche_knn50k10_merged'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a834316b-feec-496b-8afd-ae47a85ab04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert AnnData obs to DataFrame\n",
    "df = pd.DataFrame(combined_perio.obs)\n",
    "df['final_label_citeSeq'] = df['final_label_citeSeq'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['final_label_citeSeq', 'niche_knn50k10_merged']).size().unstack(fill_value=0)\n",
    "\n",
    "# Manually specify which clusters (y-axis) and niches (x-axis) to show, with order\n",
    "selected_clusters = ['B Atypical', 'B Memory 1', 'B Memory 2', 'B Naive', 'Plasma 1', 'Plasma 2', \n",
    "                   'Plasma 3', 'Plasmablast', 'CD4 TFH',\n",
    "                  'CD4 TFreg', 'mregDC', 'pDC', 'Fibroblast 1', 'Fibroblast 2', 'Endothelial 2', 'Endothelial 1', 'Endothelial 3', 'Endothelial 4', 'Endothelial 5']\n",
    "selected_niches = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']  # Replace with actual niche names\n",
    "\n",
    "# Ensure selected clusters and niches exist in the DataFrame\n",
    "niche_groups_filtered = niche_groups.loc[\n",
    "    niche_groups.index.intersection(selected_clusters),  \n",
    "    niche_groups.columns.intersection(selected_niches)   \n",
    "]\n",
    "\n",
    "# Calculate relative frequencies\n",
    "niche_groups_total = niche_groups.sum(axis=1)  # Sum across all niches for each cluster\n",
    "niche_groups_relative = niche_groups.div(niche_groups_total, axis=0)  # Compute relative frequencies\n",
    "\n",
    "# Filter the relative frequencies\n",
    "niche_groups_relative_filtered = niche_groups_relative.loc[\n",
    "    niche_groups_relative.index.intersection(selected_clusters),\n",
    "    niche_groups_relative.columns.intersection(selected_niches)\n",
    "]\n",
    "\n",
    "# Define custom order\n",
    "cluster_order = ['B Atypical', 'B Memory 1', 'B Memory 2', 'B Naive', 'CD4 TFH', 'CD4 TFreg',  'mregDC', 'Plasmablast', 'Plasma 1', 'Plasma 2',\n",
    "                   'Plasma 3',  'Fibroblast 2']\n",
    "\n",
    "niche_order = ['Lymphoid', 'Plasma', 'Plasma-Fib CT']\n",
    "\n",
    "# Apply custom order safely\n",
    "niche_groups_relative_filtered = niche_groups_relative_filtered.reindex(index=cluster_order, columns=niche_order)\n",
    "\n",
    "# Plot heatmap with clusters on y-axis and niches on x-axis\n",
    "plt.figure(figsize=(2.8, 4))  # Adjusted for better readability\n",
    "sns.heatmap(\n",
    "    niche_groups_relative_filtered, cmap='plasma',  \n",
    "    cbar_kws={'label': 'Relative Frequency'}, \n",
    "    linewidths=0.05, linecolor='black',\n",
    "    vmax=0.6,\n",
    "    xticklabels=niche_order,  # Niches on x-axis\n",
    "    yticklabels=cluster_order  # Clusters on y-axis\n",
    ")\n",
    "plt.title('Cell Type Distribution in Niches - Periodontitis')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save figure\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Citeseq_perio_clusters_niches.svg'\n",
    "fig = plt.gcf()\n",
    "plt.savefig(output_path, dpi=300, format='svg', bbox_inches='tight', facecolor=fig.get_facecolor())\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19dbddc-64e8-45a4-afaa-e314f9109d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Filter Xenium assay cells from the merged object\n",
    "xen_obj = merged[merged.obs['assay'] == 'Xenium'].copy()  # Ensure a copy to avoid warnings\n",
    "\n",
    "# Define the Lvl4 column as string type\n",
    "xen_obj.obs['Lvl4'] = xen_obj.obs['Lvl4'].astype('str')\n",
    "\n",
    "# Compute cell counts for all categories without filtering\n",
    "celltype_counts = pd.DataFrame(xen_obj.obs.groupby(['Lvl4', 'citeSeq_to_Xenium_label']).size()).unstack()\n",
    "celltype_counts.columns = celltype_counts.columns.droplevel()\n",
    "celltype_counts.index.name = 'Xenium cell type'\n",
    "celltype_counts.columns.name = 'Predicted citeSeq cell type'\n",
    "celltype_counts = celltype_counts.T\n",
    "\n",
    "# Normalize row-wise and handle NaNs\n",
    "celltype_counts = celltype_counts.div(celltype_counts.sum(axis=1), axis=0).fillna(0)\n",
    "\n",
    "# Specify clusters to plot (leave analysis intact)\n",
    "# Define filters\n",
    "lvl4_clusters_to_plot = ['B', 'PB', 'Pl.1', 'Pl.2', 'Th', 'Th_proximity_to_B', 'mregDC', 'Fib.1',  'Fib.2', 'Fib.5']  # Replace with your desired Lvl4 categories\n",
    "citeseq_clusters_to_plot = ['B Atypical', 'B Memory 1', 'B Memory 2', 'B Naive', 'Plasmablast', 'Plasma 1', 'Plasma 2', \n",
    "                   'Plasma 3', 'CD4 TFH','CD4 TFreg', 'mregDC', 'Fibroblast 2']  # Replace with your desired citeSeq labels\n",
    "\n",
    "\n",
    "# Filter for plotting only\n",
    "filtered_counts = celltype_counts.loc[\n",
    "    citeseq_clusters_to_plot, lvl4_clusters_to_plot\n",
    "]\n",
    "\n",
    "# Plot heatmap for selected clusters\n",
    "plt.figure(figsize=(7, 4))\n",
    "sns.heatmap(filtered_counts, cmap='plasma', fmt=\".2f\", cbar=True, linewidths=0.04, linecolor='black',)\n",
    "plt.title(\"Xenium vs Predicted scRNAseq Heatmap (Selected Clusters)\")\n",
    "# Save the plot as TIFF file\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure7/Xenium_Citeseq_clusters.pdf', format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c437249-0778-4ff6-9b63-915cc652ad1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.rcParams['pdf.fonttype'] = 42  # Embed TrueType fonts (best for Illustrator)\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = [ 'B', 'Th', 'Th_proximity_to_B', 'mregDC', 'Pl.1', 'Pl.2', 'PB', 'Fib.2', 'Mac', 'MyoF']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "   'BANK1', 'HLA-DMB', \n",
    "     'TRAC', 'CD4',\n",
    "    'LAMP3', 'CCR7',\n",
    "     'MZB1',\n",
    "    'MDM2', \n",
    "     'STAT3',\n",
    "    'UBE2C', 'MKI67',   'VCAN', 'CXCL12', 'CCL19', 'C3',\n",
    "    'CD14', 'CD163', 'STEAP4', 'ACTA2', \n",
    "   \n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8,\n",
    "      figsize=(5, 5)\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_max=0.8,\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure5/Xenium_Lvl4_TLS_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e58c5e-8d1d-4758-81fc-3aaa366c0c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = [ 'NK', 'Th', 'Th_proximity_to_B', 'T.IL7Rhi', 'Tc', 'Treg',  'T.IE', 'T_proximity_to_Fib']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'NKG7', 'GNLY', 'DUSP2', \n",
    "    'KLRD1', 'PRF1', 'IL7R', \n",
    "    'CD2', 'TRAC', 'CD4',\n",
    "    'PTPRC', 'NLRC5', 'CD3E',\n",
    "    'CD69', 'SLAMF1', 'CD28', 'CD83', 'LAMP3', 'CCR7', 'IL7R', \n",
    "    'MZB1', 'TENT5C', 'CCL5', \n",
    "    'GZMK', 'GZMA', 'CD8A',\n",
    "    'CTLA4', 'FOXP3', 'IL2RA',\n",
    "    'KRT5', 'KRT19', 'C1R', 'VCAN'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata_copy[adata_copy.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Tcell_Lvl4_dotplot.pdf'\n",
    "#dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccdfd62d-9f4c-4941-bae5-ff9c70c14784",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = [ 'NK', 'Th', 'Th_proximity_to_B', 'T.IL7Rhi', 'Tc', 'Treg',  'T.IE', 'T_proximity_to_Fib']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'NKG7', 'GNLY', 'DUSP2', \n",
    "    'KLRD1', 'PRF1', 'IL7R', \n",
    "    'CD2', 'TRAC', 'CD4',\n",
    "    'PTPRC', 'NLRC5', 'CD3E',\n",
    "    'CD69', 'SLAMF1', 'CD28', \n",
    "    'MZB1', 'TENT5C', 'CCL5', \n",
    "    'GZMK', 'GZMA', 'CD8A',\n",
    "    'CTLA4', 'FOXP3', 'IL2RA',\n",
    "    'KRT5', 'KRT19', 'C1R', 'VCAN'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Tcell_Lvl4_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dad0bcf-2a3b-4ca0-b253-c40fbee8d3b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = [ 'B', 'Pl.2', 'Pl.1', 'PB']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'MS4A1', 'BANK1', 'HLA-DMB', \n",
    "    'PTPRC', 'IRF8', 'CD19', \n",
    "    'SELL', 'CD83', 'CXCR4', 'CD79A', 'MZB1',\n",
    "    'IGHA1', 'TENT5C', 'SEC11C', 'DERL3', 'SLAMF7', 'MDM2', \n",
    "    'TNFRSF17', 'ICAM1',\n",
    "    'MEF2C', 'TCF4', 'CD27',\n",
    "    'CYTIP', 'BCL2L11', 'STAT3',\n",
    "    'UBE2C', 'MKI67']\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Bcell_Lvl4_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0042a88f-74dc-4e49-b1a9-fba1b050d16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Lang', 'Mac', 'cDC1', 'Mono', 'cDC2', 'Mast', 'mregDC', 'pDC']\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [\n",
    "    'CD207', 'HLA-DQB2', 'FCER1A', 'CD1A', 'HLA-DMB', 'HLA-DQA2', 'CD14', \n",
    "    'CD163', 'MS4A4A', 'MS4A6A', 'MPEG1', 'FGL2', 'CLEC7A', 'ANPEP', 'CSF2RA', 'FCN1', 'VCAN', 'CCR2',\n",
    "    'ITGB2', 'CLEC10A', 'CD1C', 'CPA3',\n",
    "    'MS4A2', 'KIT', 'IL1RL1', 'CD83', 'LAMP3', 'CCR7', 'IL7R', 'IRF8', 'IL3RA', 'GZMB', 'MZB1']\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Myeloid_Lvl4_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866dc947-9352-47a7-a72e-a905e7e3129f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', \n",
    "         'MyoF', 'SMC',  'VEC.1', 'VEC.2', 'VEC.3', 'VEC.4', 'VEC.5', \n",
    "         'LEC' ]\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [ 'COL16A1', 'VCAN', 'CXCL12', 'C1R', 'PRRX1', 'C3', 'CFB', 'MZB1',\n",
    "    'IGHA1', 'PDGFRA', 'APCDD1', 'LTBP2', 'TBX3', 'ITGB5', 'BMP4', 'PDGFRB', 'STEAP4', 'ACTA2', 'MYH11', 'MYLK', 'VWF', 'CD34', 'PECAM1', 'SELE', 'ACKR1', 'ICAM1', \n",
    "    'DNASE1L3', 'EDNRB', 'APCDD1', 'TCIM', 'EDN1', 'SOX17', 'BTNL9',  \n",
    "     'MMRN1', 'PROX1', 'PDPN']\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Stromal_Lvl4_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7357c50b-6901-4bc8-ad52-8d03a5a7dec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Neur',  'Fib.1', 'Fib.2', 'Fib.3', 'Fib.4', 'Fib.5', \n",
    "         'MyoF', 'SMC',  'VEC.1', 'VEC.2', 'VEC.3', 'VEC.4', 'VEC.5', \n",
    "         'LEC' ]\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = ['PCSK2', 'ITGB4', 'PMP22', 'SEMA3C', 'DST', \n",
    "    'COL16A1', 'VCAN', 'CXCL12', 'C1R', 'PRRX1', 'C3', 'CFB', 'MZB1',\n",
    "    'IGHA1', 'PDGFRA', 'APCDD1', 'LTBP2', 'TBX3', 'ITGB5', 'BMP4', 'PDGFRB', 'STEAP4', 'ACTA2', 'TFPI', \n",
    "    'SH2D3C', 'MYH11', 'MYLK', 'VWF', 'CD34', 'PECAM1', 'CLEC14A',  'MCF2L', 'SELE', 'ACKR1', 'ICAM1', \n",
    "    'DNASE1L3', 'EDNRB', 'KRT5', 'TCIM', 'GNG11', 'EDN1', 'SOX17', 'BTNL9',  \n",
    "     'MMRN1', 'PROX1', 'PDPN', 'ANGPT2']\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Stromal_Lvl4_dotplot.pdf'\n",
    "#dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f55c8982-9ae5-419e-b667-820a5e2adc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata.raw)  # Check if raw counts are stored\n",
    "print(adata.layers.keys())  # See if there's a 'normalized' or 'log' layer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc2a8bed-f0cd-4c47-94e8-12bb78e3b29a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Define the subset of clusters you want to include and their desired order\n",
    "clusters_of_interest = ['Ep.B.1',  'Ep.B.2', 'Ep.prol', 'Ep.B-PB', 'Ep.PB.1', 'Ep.PB.2', 'Ep.PB.3', 'Ep.Sp', 'Ep.K', \n",
    "                        'Ep.Cr.Sp.1', 'Ep.Cr.Sp.2', 'Ep.Cr.B', 'Ep.Cr.PB' ]\n",
    "\n",
    "# Define the subset of genes you want to plot in the desired order\n",
    "correct_order = [ \n",
    "    'KRT5',  'EHF', 'CLEC7A', 'COL17A1', 'DST', 'ITGB4', 'CXCL14', 'C1R', 'IL1R2',\n",
    "    'CAVIN1', 'PDPN', 'BASP1', 'SLC26A2', 'MYC', 'CFH', 'PTN',\n",
    "    'MKI67', 'CDK1', 'AQP3', 'KRT19', 'FGFBP1', 'S100A16', 'LY6D', 'SDC1',\n",
    "    'CYP4B1', 'MAMDC2', 'SERPINB3', 'NOD2', 'SLPI', 'IL1RN', 'ANXA1', 'ERBB2',\n",
    "    'CNFN', 'IL36G',  'SERPINB2', 'C15orf48', 'IL36A',\n",
    "    'ODAM', 'ODAPH', 'LAMC2', 'CXCL1', 'CXCL2', 'CXCL6', 'IL1A',\n",
    "    'DNASE1L3', 'SAA1', 'SAA2', 'CFHR3'\n",
    "]\n",
    "\n",
    "# Subset the AnnData object to include only the clusters of interest\n",
    "adata_subset = adata[adata.obs['Lvl4'].isin(clusters_of_interest)].copy()\n",
    "\n",
    "# Ensure clusters are ordered as specified\n",
    "adata_subset.obs['Lvl4'] = adata_subset.obs['Lvl4'].astype(\n",
    "    pd.CategoricalDtype(categories=clusters_of_interest, ordered=True)\n",
    ")\n",
    "\n",
    "# Create the DotPlot\n",
    "dotplot = sc.pl.DotPlot(\n",
    "    adata_subset,\n",
    "    var_names=correct_order,       # Genes on the x-axis\n",
    "    groupby='Lvl4',                # Clusters on the y-axis\n",
    "    standard_scale='var',          # Apply standard scaling across variables\n",
    "    vmin=0.2,\n",
    "    vmax=0.8                       # Set a max value for the scale\n",
    ")\n",
    "\n",
    "# Transpose the axes using the swap_axes method\n",
    "dotplot = dotplot.swap_axes(swap_axes=True)\n",
    "\n",
    "# Modify the style for grayscale and remove outlines\n",
    "dotplot = dotplot.style(\n",
    "    cmap=\"Greys\",               # Use grayscale colormap\n",
    "    dot_edge_color=None,        # Remove dot outlines\n",
    "    dot_edge_lw=0,              # No line width for edges\n",
    "    grid=False,                 # Optional: Disable grid lines\n",
    "    dot_min=0.1,                # Minimum dot size\n",
    "    dot_max=0.8                 # Maximum dot size\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "dotplot.show()\n",
    "\n",
    "# Save the plot as SVG with a transparent background\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_Epi_Lvl4_dotplot.pdf'\n",
    "dotplot.savefig(output_path, dpi=300, bbox_inches='tight', transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d47830ac-0cbe-41ef-9b60-9fcf4d20e62a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "df = pd.read_csv(\"/data/vasileiosionat2/Xenium/Xenium_area_calculation/expandedArea_cellLevel_metadata_HP_VT.csv\")\n",
    "\n",
    "# Calculate total area per sample\n",
    "total_area = df.groupby(\"sample\", as_index=False)[\"Region_Area\"].sum()\n",
    "total_area[\"Region\"] = \"Total\"\n",
    "\n",
    "# Append the new rows to the original DataFrame\n",
    "df = pd.concat([df, total_area], ignore_index=True)\n",
    "\n",
    "# Define the region order\n",
    "region_order = {\"Epi\": 1, \"CT\": 2, \"Total\": 3}\n",
    "\n",
    "# Sort the dataframe by Sample and Region order\n",
    "df[\"Region_order\"] = df[\"Region\"].map(region_order)\n",
    "df = df.sort_values([\"sample\", \"Region_order\"]).drop(columns=\"Region_order\")\n",
    "\n",
    "# Save the updated CSV\n",
    "df.to_csv(\"/data/vasileiosionat2/Xenium/Xenium_area_calculation/expandedArea_cellLevel_metadata_HP_VT_new.csv\", index=False)\n",
    "\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea473bf1-eb38-4243-abb4-6dfce67df90a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "df = pd.read_csv(\"/data/vasileiosionat2/Xenium/Xenium_area_calculation/expandedArea_cellLevel_metadata_HP_VT_new.csv\")\n",
    "# Filter rows where Region column contains only 'CT' and 'Epi'\n",
    "Total_df = df[df['Region'].isin(['Total'])]\n",
    "Epi_df = df[df['Region'].isin(['Epi'])]\n",
    "CT_df = df[df['Region'].isin(['CT'])]\n",
    "\n",
    "# Write the filtered dataframe to a new CSV file\n",
    "Total_df.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Total_area_merged_calculations.csv', index=False)\n",
    "Epi_df.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Epi_area_merged_calculations.csv', index=False)\n",
    "CT_df.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_CT_area_merged_calculations.csv', index=False)\n",
    "\n",
    "print(Total_df)\n",
    "print(Epi_df)\n",
    "print(CT_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a5cca2-a162-4836-ae51-dd742c36fcbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['PatientID_Atlas']=adata.obs['pt.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f261fbdd-ca43-4dd2-b0e4-719ca0b37ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure 'PatientID_Atlas' is a string type\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].astype(str)\n",
    "\n",
    "# Define the replacement dictionary\n",
    "replacement_dict = {\n",
    "'HV171' : 'HC-1',\n",
    "'HV180' : 'HC-2',\n",
    "'HV189'\t: 'HC-3',\n",
    "'HV191' : 'HC-4',\n",
    "'HV203' : 'HC-5',\n",
    "'HV207' : 'HC-6',\n",
    "'HV220' : 'HC-7',\n",
    "'HV187' : 'HC-8',\n",
    "'HV205' : 'HC-9',\n",
    "'HV137' : 'CP-1',\n",
    "'HV140'\t: 'CP-2',\n",
    "'HV154' : 'CP-3',\n",
    "'HV160' : 'CP-4',\n",
    "'HV181' : 'CP-5',\n",
    "'HV196' : 'CP-6',\n",
    "'HV204' : 'CP-7',\n",
    "'HV214' : 'CP-8',\n",
    "'HV219' : 'CP-9',\n",
    "'HV184' : 'CP-10',\n",
    "'HV185' : 'CP-11',\n",
    "'HV188' : 'CP-12',\n",
    "'HV192' : 'CP-13',\n",
    "'HV198' : 'CP-14',\n",
    "'HV211' : 'CP-15',\n",
    "    \n",
    "}\n",
    "\n",
    "# Replace only the specified categories in 'T_leiden2'\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].astype(str)  # Ensure it's string\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].replace(replacement_dict)\n",
    "\n",
    "# Check the unique values in 'T_leiden2' to confirm\n",
    "print(adata.obs['PatientID_Atlas'].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9abc34fd-125b-40cf-a9e1-67a2e96029a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure 'PatientID_Atlas' is a string type\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].astype(str)\n",
    "\n",
    "# Define the replacement dictionary\n",
    "replacement_dict = {\n",
    "'HC-1': 'H-1',\n",
    "'HC-2': 'H-2',\n",
    "'HC-3': 'H-3',\n",
    "'HC-4': 'H-4',\n",
    "'HC-5': 'H-5',\n",
    "'HC-6': 'H-6',\n",
    "'HC-7': 'H-7',\n",
    "'HC-8': 'H-8',\n",
    "'HC-9': 'H-9',\n",
    "'CP-1': 'P-1',\n",
    "'CP-2': 'P-2',\n",
    "'CP-3': 'P-3',\n",
    "'CP-4': 'P-4',\n",
    "'CP-5': 'P-5',\n",
    "'CP-6': 'P-6',\n",
    "'CP-7': 'P-7',\n",
    "'CP-8': 'P-8',\n",
    "'CP-9': 'P-9',\n",
    "'CP-10': 'P-10',\n",
    "'CP-11': 'P-11',\n",
    "'CP-12': 'P-12',\n",
    "'CP-13': 'P-13',\n",
    "'CP-14': 'P-14',\n",
    "'CP-15': 'P-15',\n",
    "    \n",
    "}\n",
    "\n",
    "# Replace only the specified categories in 'T_leiden2'\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].astype(str)  # Ensure it's string\n",
    "adata.obs['PatientID_Atlas'] = adata.obs['PatientID_Atlas'].replace(replacement_dict)\n",
    "\n",
    "# Check the unique values in 'T_leiden2' to confirm\n",
    "print(adata.obs['PatientID_Atlas'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1925f82-a9a7-42f9-b2a5-178efcd9a010",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsunique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b9a4bc-8742-4a31-9b2d-4c07dbf8de85",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65820259-853c-4136-9381-b9bf769d6803",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generation of stats table excluding samples HV160D, HV171B (small detached tissues) and HV140A2 (poor orientation with a bias for epithelium)\n",
    "annotation = \"Lvl1\"\n",
    "loc = \"region\"\n",
    "status = \"status.3\"\n",
    "sample_id = \"sample\"\n",
    "patient_id= \"PatientID_Atlas\"\n",
    "meta = adata.obs\n",
    "\n",
    "# Get unique combinations of sample_id, annotation, status, loc\n",
    "unique_combinations = meta[[patient_id, sample_id, annotation, status, loc]].drop_duplicates().astype(str)\n",
    "\n",
    "# Initialize result table\n",
    "result_table = pd.DataFrame(columns=['sample', 'patient', 'status', 'loc', 'annotation', 'total.cells'])\n",
    "result_table = result_table.sort_values(by=['annotation', 'status']).reset_index(drop=True)\n",
    "\n",
    "# Iterate through unique combinations\n",
    "for _, row in unique_combinations.iterrows():\n",
    "    current_sample_id = row[sample_id]\n",
    "    current_patient_id= row[patient_id]\n",
    "    current_annotation = row[annotation]\n",
    "    current_status = row[status]\n",
    "    current_loc = row[loc]\n",
    "\n",
    "    # Filter data based on annotation and sample_id (no need for section_id here)\n",
    "    current_meta = meta[(meta[annotation] == current_annotation) & (meta[sample_id] == current_sample_id)]\n",
    "\n",
    "    # Ensure 'current_meta' has data\n",
    "    if current_meta.empty:\n",
    "        print(f\"No data for {current_annotation} in location {current_loc}. Skipping.\")\n",
    "        continue\n",
    "\n",
    "    # Calculate total cells\n",
    "    total_cells = len(current_meta)\n",
    "\n",
    "    # Add new row to the DataFrame with total cells\n",
    "    new_row = pd.DataFrame({'sample': [current_sample_id],\n",
    "                            'patient': [current_patient_id],  # Assuming 'sample_id' is the patient id\n",
    "                            'status': [current_status],\n",
    "                            'loc': [current_loc],\n",
    "                            'annotation': [current_annotation],\n",
    "                            'total.cells': [total_cells]})\n",
    "    \n",
    "    result_table = pd.concat([result_table, new_row], ignore_index=True)\n",
    "\n",
    "# Sort result table\n",
    "result_table = result_table.sort_values(by=['annotation', 'status']).reset_index(drop=True)\n",
    "\n",
    "# Ensure numeric values for total cells\n",
    "result_table['total.cells'] = pd.to_numeric(result_table['total.cells'], errors='coerce')\n",
    "\n",
    "# Group by patient, annotation, and status\n",
    "grouped = result_table.groupby(['patient', 'sample', 'annotation', 'status']).agg(\n",
    "    total_cells_pt=('total.cells', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "\n",
    "# Filter out unwanted samples from the grouped table\n",
    "exclude_samples = ['s2r9_a', 's2r9_b', 'HV140A_C', 'HV140A_D', 's2r4_a', 's2r4_b', 's2r5', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', \n",
    "                       'HV192A_A', 'HV192A_B','HV192A_C']\n",
    "grouped_filtered = grouped[~grouped['sample'].isin(exclude_samples)]\n",
    "\n",
    "# Save the filtered result to CSV\n",
    "grouped_filtered.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_filtered.csv')\n",
    "\n",
    "# Show the filtered result\n",
    "print(grouped_filtered)\n",
    "\n",
    "\n",
    "# Show the grouped result\n",
    "print(grouped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "110fd7dc-b4fa-4f52-a9bc-c1f8902926c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV files into DataFrames\n",
    "df1 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_filtered.csv')\n",
    "df2 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Total_area_merged_calculations.csv')\n",
    "\n",
    "# Merge the tables on the 'sample' column\n",
    "df_merged = pd.merge(df1, df2[['sample', 'Region_Area']], on='sample', how='left')\n",
    "\n",
    "# Convert Region_Area to mm²\n",
    "df_merged['Region_Area_mm2'] = df_merged['Region_Area'] / 1000  # divide by 1,000,000 to convert to mm²\n",
    "df_merged['cells_per_mm2'] = df_merged['total_cells_pt'] / df_merged['Region_Area_mm2']\n",
    "df_merged.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_per_area(Section).csv')\n",
    "\n",
    "# Check the first few rows to confirm\n",
    "print(df_merged.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ef8517-cbd5-4eca-b460-90f34cb85a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV files into DataFrames\n",
    "df1 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_filtered.csv')\n",
    "df2 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Total_area_merged_calculations.csv')\n",
    "\n",
    "# Merge the tables on the 'sample' column\n",
    "df_merged2 = pd.merge(df1, df2[['sample', 'Region_Area']], on='sample', how='left')\n",
    "\n",
    "# Convert Region_Area to mm²\n",
    "df_merged2['Region_Area_mm2'] = df_merged2['Region_Area'] / 1000\n",
    "\n",
    "# Group by patient, annotation, and sample and sum the total_cells_pt\n",
    "df_merged2 = df_merged2.groupby(['patient', 'annotation', 'status']).agg({\n",
    "    'total_cells_pt': 'sum',\n",
    "    'Region_Area': 'sum',\n",
    "    'Region_Area_mm2': 'sum'\n",
    "}).reset_index()\n",
    "\n",
    "# Cells per area calculation\n",
    "df_merged2['cells_per_mm2'] = df_merged2['total_cells_pt'] / df_merged2['Region_Area_mm2']\n",
    "\n",
    "# Save the result to a CSV file\n",
    "df_merged2.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_per_area(Patient).csv')\n",
    "\n",
    "# Display the merged DataFrame\n",
    "df_merged2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ae957b-e5d6-46cf-a85d-1dc2e6985e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "# Load the CSV file into a pandas DataFrame\n",
    "df = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl1_total_cells_per_area(Patient).csv')\n",
    "\n",
    "# Define the specific order of major_cluster categories and corresponding colors\n",
    "anticipated_order = ['Other', 'Epithelial', 'Immune', 'Fibroblast', 'Vascular']\n",
    "colors = {\n",
    "    'Other': '#e5e5e5',\n",
    "    'Epithelial': '#FFEA00',\n",
    "    'Immune': '#008000',\n",
    "    'Fibroblast': '#0000FF',\n",
    "    'Vascular': '#990F0FFF',\n",
    "}\n",
    "\n",
    "# Pivot the data so that each 'patientID' has columns for each 'lvl1_spatial_cluster'\n",
    "df_pivot = df.pivot_table(index='patient', columns='annotation', values='cells_per_mm2', aggfunc='sum')\n",
    "\n",
    "# Reorder the columns (clusters) based on the anticipated_order\n",
    "df_pivot = df_pivot[anticipated_order]\n",
    "\n",
    "# Manually specify the order of patients\n",
    "manual_order = ['H-1','H-2','H-4','H-5','H-6','H-7','H-8', 'H-9',\n",
    "                'P-1', 'P-2', 'P-3', 'P-4', 'P-6', 'P-7', 'P-8', 'P-9', 'P-10', 'P-11', 'P-14']\n",
    "\n",
    "# Reorder the DataFrame rows based on the manual order\n",
    "df_pivot = df_pivot.loc[manual_order]\n",
    "\n",
    "# Plot the stacked bar plot with adjusted bar width\n",
    "plt.figure(figsize=(20, 8))\n",
    "df_pivot.plot(kind='bar', stacked=True, figsize=(20, 10), color=[colors[cluster] for cluster in anticipated_order], width=0.9)  # Adjust the width\n",
    "\n",
    "# Format the plot\n",
    "plt.gca().set_facecolor('white')\n",
    "plt.gca().yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{int(x)}'))  # Format y-axis to show integer counts\n",
    "plt.gca().spines['top'].set_visible(False)  # Remove top border\n",
    "plt.gca().spines['right'].set_visible(False)  # Remove right border\n",
    "plt.gca().spines['bottom'].set_color('black')  # Set bottom border to black\n",
    "plt.gca().spines['left'].set_color('black')  # Set left border to black\n",
    "\n",
    "plt.xlabel('Patient ID')\n",
    "plt.ylabel('Total Cells')\n",
    "plt.xticks(fontsize=20, rotation=45, ha='right')  # Adjust font size and rotation\n",
    "plt.yticks(fontsize=20)\n",
    "\n",
    "# Custom legend handles and labels\n",
    "sorted_legend_handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=colors[label], markersize=10) \n",
    "                         for label in anticipated_order]\n",
    "plt.legend(sorted_legend_handles, anticipated_order, loc='upper left', bbox_to_anchor=(1, 1), frameon=False)\n",
    "\n",
    "# Adjust layout and save the plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure1/Xenium_Lvl1_barplot_per_patient(cells_per_area).pdf', format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cafd4292-5baf-42c8-a68f-5927924ed870",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generation of stats table excluding samples HV160D, HV171B (small detached tissues) and HV140A2 (poor orientation with a bias for epithelium)\n",
    "annotation = \"Lvl2.5\"\n",
    "loc = \"region\"\n",
    "status = \"status.3\"\n",
    "sample_id = \"sample\"\n",
    "patient_id= \"PatientID_Atlas\"\n",
    "meta = adata.obs\n",
    "\n",
    "# Get unique combinations of sample_id, annotation, status, loc\n",
    "unique_combinations = meta[[patient_id, sample_id, annotation, status, loc]].drop_duplicates().astype(str)\n",
    "\n",
    "# Initialize result table\n",
    "result_table = pd.DataFrame(columns=['sample', 'patient', 'status', 'loc', 'annotation', 'total.cells'])\n",
    "result_table = result_table.sort_values(by=['annotation', 'status']).reset_index(drop=True)\n",
    "\n",
    "# Iterate through unique combinations\n",
    "for _, row in unique_combinations.iterrows():\n",
    "    current_sample_id = row[sample_id]\n",
    "    current_patient_id= row[patient_id]\n",
    "    current_annotation = row[annotation]\n",
    "    current_status = row[status]\n",
    "    current_loc = row[loc]\n",
    "\n",
    "    # Filter data based on annotation and sample_id (no need for section_id here)\n",
    "    current_meta = meta[(meta[annotation] == current_annotation) & (meta[sample_id] == current_sample_id)]\n",
    "\n",
    "    # Ensure 'current_meta' has data\n",
    "    if current_meta.empty:\n",
    "        print(f\"No data for {current_annotation} in location {current_loc}. Skipping.\")\n",
    "        continue\n",
    "\n",
    "    # Calculate total cells\n",
    "    total_cells = len(current_meta)\n",
    "\n",
    "    # Add new row to the DataFrame with total cells\n",
    "    new_row = pd.DataFrame({'sample': [current_sample_id],\n",
    "                            'patient': [current_patient_id],  # Assuming 'sample_id' is the patient id\n",
    "                            'status': [current_status],\n",
    "                            'loc': [current_loc],\n",
    "                            'annotation': [current_annotation],\n",
    "                            'total.cells': [total_cells]})\n",
    "    \n",
    "    result_table = pd.concat([result_table, new_row], ignore_index=True)\n",
    "\n",
    "# Sort result table\n",
    "result_table = result_table.sort_values(by=['annotation', 'status']).reset_index(drop=True)\n",
    "\n",
    "# Ensure numeric values for total cells\n",
    "result_table['total.cells'] = pd.to_numeric(result_table['total.cells'], errors='coerce')\n",
    "\n",
    "# Group by patient, annotation, and status\n",
    "grouped = result_table.groupby(['patient', 'sample', 'annotation', 'status']).agg(\n",
    "    total_cells_pt=('total.cells', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "\n",
    "# Filter out unwanted samples from the grouped table\n",
    "exclude_samples = ['s2r9_a', 's2r9_b', 'HV140A_C', 'HV140A_D', 's2r4_a', 's2r4_b', 's2r5', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', \n",
    "                       'HV192A_A', 'HV192A_B','HV192A_C']\n",
    "grouped_filtered = grouped[~grouped['sample'].isin(exclude_samples)]\n",
    "\n",
    "# Save the filtered result to CSV\n",
    "grouped_filtered.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_filtered.csv')\n",
    "\n",
    "# Show the filtered result\n",
    "print(grouped_filtered)\n",
    "\n",
    "\n",
    "# Show the grouped result\n",
    "print(grouped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61ca182-07e4-4ac0-ab50-f51eb38fa1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV files into DataFrames\n",
    "df1 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_filtered.csv')\n",
    "df2 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Total_area_merged_calculations.csv')\n",
    "\n",
    "# Merge the tables on the 'sample' column\n",
    "df_merged = pd.merge(df1, df2[['sample', 'Region_Area']], on='sample', how='left')\n",
    "\n",
    "# Convert Region_Area to mm²\n",
    "df_merged['Region_Area_mm2'] = df_merged['Region_Area'] / 1000  # divide by 1,000,000 to convert to mm²\n",
    "df_merged['cells_per_mm2'] = df_merged['total_cells_pt'] / df_merged['Region_Area_mm2']\n",
    "df_merged.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_per_area(Section).csv')\n",
    "\n",
    "# Check the first few rows to confirm\n",
    "print(df_merged.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a5705a-ab00-48dd-a1cb-cfef21af3242",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV files into DataFrames\n",
    "df1 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_filtered.csv')\n",
    "df2 = pd.read_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Total_area_merged_calculations.csv')\n",
    "\n",
    "# Merge the tables on the 'sample' column\n",
    "df_merged2 = pd.merge(df1, df2[['sample', 'Region_Area']], on='sample', how='left')\n",
    "\n",
    "# Convert Region_Area to mm²\n",
    "df_merged2['Region_Area_mm2'] = df_merged2['Region_Area'] / 1000\n",
    "\n",
    "# Group by patient, annotation, and sample and sum the total_cells_pt\n",
    "df_merged2 = df_merged2.groupby(['patient', 'annotation', 'status']).agg({\n",
    "    'total_cells_pt': 'sum',\n",
    "    'Region_Area': 'sum',\n",
    "    'Region_Area_mm2': 'sum'\n",
    "}).reset_index()\n",
    "\n",
    "# Cells per area calculation\n",
    "df_merged2['cells_per_mm2'] = df_merged2['total_cells_pt'] / df_merged2['Region_Area_mm2']\n",
    "\n",
    "# Save the result to a CSV file\n",
    "df_merged2.to_csv('/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_per_area(Patient).csv')\n",
    "\n",
    "# Display the merged DataFrame\n",
    "df_merged2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7ca90d-e46e-4676-baad-7172bc682146",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Load your data\n",
    "file_path = '/data/vasileiosionat2/Xenium/Xenium_area_calculation/Xenium_Lvl2.5_total_cells_per_area(Patient).csv'  # Update with your actual file path\n",
    "df = pd.read_csv(file_path)  # Adjust delimiter if necessary\n",
    "\n",
    "# Filter relevant columns\n",
    "df = df[['annotation', 'status', 'cells_per_mm2', 'patient']]  # Include 'patient' for uniqueness\n",
    "\n",
    "# Get unique annotations\n",
    "annotations = df['annotation'].unique()\n",
    "\n",
    "# Define number of rows and columns for the subplot grid\n",
    "n_rows = 4\n",
    "n_cols = (len(annotations) + n_rows - 1) // n_rows  # Calculate number of columns based on number of annotations\n",
    "\n",
    "# Set up the figure with subplots\n",
    "fig, axes = plt.subplots(n_rows, n_cols, figsize=(n_cols * 2.5, n_rows * 2))\n",
    "axes = axes.flatten()  # Flatten the axes array to easily index\n",
    "\n",
    "# Store p-values for multiple testing correction\n",
    "p_values = []\n",
    "\n",
    "# Iterate over each annotation and plot\n",
    "for idx, annotation in enumerate(annotations):\n",
    "    # Filter data for the current annotation\n",
    "    df_annotation = df[df['annotation'] == annotation]\n",
    "    \n",
    "    # Perform t-test for 'status' groups (Healthy vs Perio) in the current annotation\n",
    "    healthy_data = df_annotation[df_annotation['status'] == 'healthy']['cells_per_mm2']\n",
    "    perio_data = df_annotation[df_annotation['status'] == 'perio']['cells_per_mm2']\n",
    "    \n",
    "    # Perform t-test\n",
    "    t_stat, p_val = stats.ttest_ind(healthy_data, perio_data, equal_var=False)  # Welch's t-test\n",
    "    p_values.append(p_val)\n",
    "    # Define the desired order\n",
    "    order = [\"healthy\", \"perio\"]\n",
    "\n",
    "    # Plotting on the corresponding subplot axis\n",
    "    sns.boxplot(data=df_annotation, x=\"status\", y=\"cells_per_mm2\", \n",
    "            palette={\"healthy\": \"#006A8E\", \"perio\": \"#B1283A\"}, width=0.7, \n",
    "            fliersize=0, dodge=True, order=order, ax=axes[idx])  \n",
    "\n",
    "    # Add individual data points (jittered for visibility)\n",
    "    sns.stripplot(data=df_annotation, x=\"status\", y=\"cells_per_mm2\", \n",
    "              palette={\"healthy\": \"black\", \"perio\": \"black\"}, dodge=True,\n",
    "              jitter=True, alpha=0.7, marker=\"o\", edgecolor=\"black\", \n",
    "              linewidth=0.1, order=order, ax=axes[idx])\n",
    "\n",
    "    # Set aesthetics for the subplot\n",
    "    axes[idx].set_xlabel(\"\")\n",
    "    axes[idx].set_ylabel(\"Cells per mm²\")\n",
    "    axes[idx].set_title(f\"{annotation}\")\n",
    "    axes[idx].legend(title=\"\", loc=\"upper right\")\n",
    "    axes[idx].grid(axis=\"y\", linestyle=\"--\", alpha=0.7)\n",
    "\n",
    "# Correct p-values for multiple comparisons using Bonferroni\n",
    "_, corrected_p_values, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "# Display corrected p-values on the plots\n",
    "for idx, annotation in enumerate(annotations):\n",
    "    # Correct p-values\n",
    "    p_val = corrected_p_values[idx]\n",
    "    axes[idx].text(0.5, 0.95, f'P = {p_val:.3f}', transform=axes[idx].transAxes, ha='center', va='top', fontsize=10, color='black')\n",
    "\n",
    "# Adjust layout to ensure labels and titles fit\n",
    "plt.tight_layout()\n",
    "plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Xenium_Lvl2.5_total_clusters_per_area).pdf', format='pdf')\n",
    "# Show the figure\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a2ddfe-361a-4a93-af6c-493c21878eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'plot_niches' is a DataFrame or similar data structure\n",
    "df = pd.DataFrame(adata.obs)\n",
    "df['spatial_expression_cluster'] = df['KNN50_K10'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_knn50k10', 'Lvl4']).size().unstack(fill_value=0)\n",
    "\n",
    "# Identify clusters to remove\n",
    "clusters_to_remove = df['Lvl4'].str.contains('Mix') | df['Lvl4'].str.startswith('Oth.') | df['Lvl4'].str.startswith('Neuro') | df['Lvl4'].str.startswith('Unclea')\n",
    "\n",
    "# Get the unique clusters to be removed\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'Lvl4'].unique()\n",
    "\n",
    "# Filter out these clusters from the plot\n",
    "niche_groups_filtered = niche_groups.drop(columns=clusters_to_remove, errors='ignore')\n",
    "\n",
    "# Plot heatmap with enhanced colormap and custom vmax\n",
    "plt.figure(figsize=(25, 10))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(niche_groups_filtered, cmap='inferno', cbar_kws={'label': 'Count', 'ticks': [0, 1000, 2000, 4000], 'format': '%.0f'}, linewidths=0.05, linecolor='black', vmax=4000)\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche Cluster')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=45, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Customize the color bar label\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "cbar.set_ticks([0, 1000, 2000, 4000])\n",
    "cbar.ax.set_yticklabels(['0', '1,000', '2,000', '>4,000'])  # Set custom tick labels\n",
    "\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07092f55-a893-4209-ba31-3454b9148741",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['niche_knn50k10'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6de2fe8-e9d3-47da-a888-e382d77464a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import scimap.plotting as sm\n",
    "import matplotlib.ticker as mtick\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "# List of SampleIDs to exclude\n",
    "excluded_sample_ids = ['s2r9_a', 's2r9_b', 'HV140A_C', 'HV140A_D', 's2r4_a', 's2r4_b', 's2r5', 's2r8_HV188_a', 's2r8_HV188_b', 's2r8_HV211', \n",
    "                       'HV192A_A', 'HV192A_B','HV192A_C']\n",
    "\n",
    "# Create a subset of the AnnData object excluding the specific SampleIDs\n",
    "adata_subset = adata[~adata.obs['sample'].isin(excluded_sample_ids)].copy()\n",
    "\n",
    "# Create a new figure with custom dimensions\n",
    "plt.figure(figsize=(50, 6))  # Adjust width and height as needed\n",
    "\n",
    "# Define the specific order of major_cluster categories\n",
    "anticipated_order = [ 'Spinous', 'Keratin', 'Crevicular',  'Fib CT 1', 'Fib CT 2', 'Epi-CT 1', 'Epi-CT 2', 'Lymphoid', 'Plasma-Fib CT',  'Plasma']\n",
    "\n",
    "# Reorder the major_cluster column in the copied AnnData object\n",
    "adata_subset.obs['niche_knn50k10'] = adata_subset.obs['niche_knn50k10'].astype('category')\n",
    "adata_subset.obs['niche_knn50k10'] = adata_subset.obs['niche_knn50k10'].cat.reorder_categories(anticipated_order)\n",
    "adata_subset.obs = adata_subset.obs.sort_values('niche_knn50k10')\n",
    "\n",
    "# Define custom colors for each major_cluster category\n",
    "colors = {\n",
    "    'Fib CT 1' : 'red',\n",
    "    'Fib CT 2' : 'blue',\n",
    "     'Epi-CT 1': '#00FFFF',\n",
    "    'Epi-CT 2': 'orange',\n",
    "     'Plasma-Fib CT' : '#CF9FFF',\n",
    "     'Plasma' : 'magenta' ,\n",
    "     'Spinous' : '#5D3FD3',\n",
    "     'Keratin' : '#0096FF' ,\n",
    "     'Crevicular' : '#800080' ,\n",
    "     'Lymphoid' : 'yellow'\n",
    "    \n",
    "}\n",
    "\n",
    "# Plot the stacked bar plot with specified colors\n",
    "plot = sm.stacked_barplot(adata_subset, x_axis='status.3', y_axis='niche_knn50k10',\n",
    "                          order_yaxis=anticipated_order, method='percent', plot_tool='matplotlib', color=colors, figsize=(2, 6))\n",
    "\n",
    "# Get the current axes and set its background color to white\n",
    "plt.gca().set_facecolor('white')\n",
    "\n",
    "# Format y-axis ticks as percentages\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(1.0))\n",
    "\n",
    "# Get the x-axis limits\n",
    "x_min, x_max = plt.xlim()\n",
    "\n",
    "# Plot horizontal dotted lines at y = 0.25, 0.5, and 0.75\n",
    "plt.hlines(y=[0, 0.25, 0.5, 0.75, 1], xmin=x_min, xmax=x_max, linestyle='--', colors='grey')\n",
    "\n",
    "# Set y-axis tick labels to match the percentages\n",
    "plt.yticks([0, 0.25, 0.5, 0.75, 1], ['0%', '25%', '50%', '75%', '100%'])\n",
    "\n",
    "# Modify the borders\n",
    "plt.gca().spines['top'].set_visible(False)       # Make top border invisible\n",
    "plt.gca().spines['right'].set_visible(False)     # Make right border invisible\n",
    "plt.gca().spines['bottom'].set_color('black')    # Set bottom border color to black\n",
    "plt.gca().spines['left'].set_color('black')      # Set left border color to black\n",
    "\n",
    "# Adjust y-axis limits to start a little above the lower edge of the outline\n",
    "y_min, y_max = plt.ylim()\n",
    "plt.ylim(y_min - 0.03, y_max)\n",
    "\n",
    "# Remove 'clinical_status' from x-axis title\n",
    "plt.xlabel('')\n",
    "\n",
    "# Increase font size of x-axis tick labels (clinical_status categories)\n",
    "plt.xticks(fontsize=20)\n",
    "plt.yticks(fontsize=20)  # Adjust font size as needed\n",
    "\n",
    "# Rotate x-axis labels diagonally\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "# Custom legend handles as circles, sorted by anticipated_order in reverse\n",
    "sorted_legend_handles = [Line2D([0], [0], marker='o', color='w', markerfacecolor=colors[category], markersize=10) \n",
    "                         for category in reversed(anticipated_order)]\n",
    "sorted_legend_labels = reversed(anticipated_order)\n",
    "\n",
    "# Add legend with sorted custom handles and labels outside the plot\n",
    "legend = plt.legend(sorted_legend_handles, sorted_legend_labels, loc='upper left', bbox_to_anchor=(1, 1), markerscale=1, frameon=False, borderpad=2)\n",
    "\n",
    "# Save the plot as TIFF file\n",
    "#plt.savefig('/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure4/Xenium_niche_stacked_barplot.pdf', format='pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "934154e0-c5f9-427a-9bd5-56bcc5996025",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV205B']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['niche_knn50k10', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['niche_knn50k10'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['niche_knn50k10'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['niche_knn50k10'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2db23acd-0613-4e95-b54c-e4683955af6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV192A_C']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['niche_knn50k10', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['niche_knn50k10'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['niche_knn50k10'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['niche_knn50k10'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4b55ff8-460b-43e6-85f2-b7c130a279d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV205B']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['Lvl3', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['Lvl3'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['Lvl3'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['Lvl3'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13a968b-d2a8-46e2-8ced-0a558358da03",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6291221e-83a1-446c-b5bc-97332c89b318",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b198470-fba2-44b1-bacc-9b696381d710",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV205B']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['KNN10_K10', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['KNN10_K10'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['KNN10_K10'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['KNN10_K10'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c5ca30c-3a45-4da3-810a-9bad95d9df4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 's2r1_HV207']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['KNN10_K10', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['KNN10_K10'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['KNN10_K10'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['KNN10_K10'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f830bc13-5191-4b41-a86a-7bda2235bcb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 's2r2_HV184']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['KNN10_K10', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['KNN10_K10'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['KNN10_K10'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['KNN10_K10'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "813d7a52-8bb5-4358-8100-fba88566bed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 's1r2']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['niche_cc14', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['niche_cc14'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['niche_cc14'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['niche_cc14'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fabf9cb9-cb1e-4dfd-995f-ce2722270a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 'HV205B']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['Lvl4', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['Lvl4'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 12\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['Lvl4'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['Lvl4'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "644addc1-cb53-4194-a306-92f234177bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Filter the data for the specific sample\n",
    "adata_sample = adata[adata.obs['sample'] == 's2r3']\n",
    "\n",
    "# Ensure required columns exist\n",
    "required_columns = ['niche_cc14', 'x_centroid', 'y_centroid']\n",
    "missing_cols = [col for col in required_columns if col not in adata_sample.obs]\n",
    "if missing_cols:\n",
    "    raise KeyError(f\"Missing columns in adata_sample.obs: {missing_cols}\")\n",
    "\n",
    "# Get unique clusters\n",
    "unique_clusters = adata_sample.obs['niche_cc14'].unique()\n",
    "\n",
    "# Reorder clusters (modify this if you have a custom order)\n",
    "ordered_clusters = sorted(unique_clusters)\n",
    "\n",
    "# Define number of rows and columns for subplots\n",
    "num_rows = 6\n",
    "num_cols = int(np.ceil(len(ordered_clusters) / num_rows))  # Calculate columns dynamically\n",
    "\n",
    "# Create the figure\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(46, 45))\n",
    "fig.patch.set_facecolor('white')\n",
    "\n",
    "# Flatten the axes array for easier indexing\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Get original coordinates\n",
    "x_coords = adata_sample.obs['x_centroid']\n",
    "y_coords = adata_sample.obs['y_centroid']\n",
    "\n",
    "# Define rotation angle in degrees\n",
    "rotation_angle = 45  # Modify if needed\n",
    "\n",
    "# Apply rotation transformation\n",
    "angle = np.deg2rad(rotation_angle)\n",
    "new_x_coords = x_coords * np.cos(angle) - y_coords * np.sin(angle)\n",
    "new_y_coords = x_coords * np.sin(angle) + y_coords * np.cos(angle)\n",
    "\n",
    "# Determine aspect ratio\n",
    "x_range = new_x_coords.max() - new_x_coords.min()\n",
    "y_range = new_y_coords.max() - new_y_coords.min()\n",
    "aspect_ratio = x_range / y_range\n",
    "\n",
    "# Loop through each unique cluster\n",
    "for idx, cluster in enumerate(ordered_clusters):\n",
    "    ax = axes[idx]\n",
    "\n",
    "    # Set white background\n",
    "    ax.set_facecolor('white')\n",
    "\n",
    "    # Remove subplot outline\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "\n",
    "    # Scatter plot for all other clusters (gray)\n",
    "    mask_other = adata_sample.obs['niche_cc14'] != cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_other.to_numpy()],\n",
    "        new_y_coords[mask_other.to_numpy()],\n",
    "        c='#C0C0C0', \n",
    "        s=3\n",
    "    )\n",
    "\n",
    "    # Scatter plot for the current cluster (red)\n",
    "    mask_cluster = adata_sample.obs['niche_cc14'] == cluster\n",
    "    ax.scatter(\n",
    "        new_x_coords[mask_cluster.to_numpy()],\n",
    "        new_y_coords[mask_cluster.to_numpy()],\n",
    "        c='red', \n",
    "        s=9\n",
    "    )\n",
    "\n",
    "    # Set aspect ratio\n",
    "    ax.set_aspect(aspect_ratio)\n",
    "\n",
    "    # Add cluster name below each plot\n",
    "    ax.text(\n",
    "        0.5, 0.02, f'{cluster}', \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=ax.transAxes, \n",
    "        color='black', fontsize=20, weight='bold'\n",
    "    )\n",
    "\n",
    "    # Remove grid, ticks, and labels\n",
    "    ax.grid(False)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Turn off unused subplots\n",
    "for ax in axes[len(ordered_clusters):]:\n",
    "    ax.set_visible(False)\n",
    "\n",
    "# Adjust subplot spacing\n",
    "plt.subplots_adjust(\n",
    "    left=0.05, right=0.95, top=0.95, bottom=0.05, \n",
    "    wspace=0.1, hspace=0.05\n",
    ")\n",
    "\n",
    "# Add title\n",
    "plt.suptitle(f'All Clusters in s2r2_HV184', color='black', fontsize=20, weight='bold', y=1.02)\n",
    "\n",
    "# Show the figure\n",
    "plt.show()\n",
    "plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b6c3b20-0b70-4f23-b815-75f7fa2cbe27",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'plot_niches' is a DataFrame or similar data structure\n",
    "df = pd.DataFrame(adata_perio.obs)\n",
    "df['spatial_expression_cluster'] = df['niche_cc14'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['niche_cc14', 'Lvl5']).size().unstack(fill_value=0)\n",
    "\n",
    "# Identify clusters to remove\n",
    "clusters_to_remove = df['Lvl5'].str.contains('Mix') | df['Lvl5'].str.startswith('Oth.') | df['Lvl5'].str.startswith('Neuro') | df['Lvl5'].str.startswith('Unclea')\n",
    "\n",
    "# Get the unique clusters to be removed\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'Lvl5'].unique()\n",
    "\n",
    "# Filter out these clusters from the plot\n",
    "niche_groups_filtered = niche_groups.drop(columns=clusters_to_remove, errors='ignore')\n",
    "\n",
    "# Plot heatmap with enhanced colormap and custom vmax\n",
    "plt.figure(figsize=(25, 10))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(niche_groups_filtered, cmap='inferno', cbar_kws={'label': 'Count', 'ticks': [0, 1000, 2000, 4000], 'format': '%.0f'}, linewidths=0.05, linecolor='black', vmax=5000)\n",
    "plt.title('Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche Cluster')\n",
    "plt.ylabel('Cell Type')\n",
    "plt.xticks(rotation=45, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Customize the color bar label\n",
    "cbar = plt.gca().collections[0].colorbar\n",
    "cbar.set_ticks([0, 1000, 2000, 4000])\n",
    "cbar.ax.set_yticklabels(['0', '1,000', '2,000', '>4,000'])  # Set custom tick labels\n",
    "\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "041e6202-cb6a-4a57-86c1-397e34d62a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata.obs[['KNN10_K10', 'Lvl4']].head())  # Check for non-null entries\n",
    "print(adata.obs['KNN10_K10'].unique())  # Check unique values in KNN10_K10\n",
    "print(adata.obs['Lvl4'].unique())  # Check unique values in Lvl4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003e6d34-26bd-4c02-b0a5-2553eb2dbfb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['status.3'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0fa9f7-53bb-4a45-9b7b-c9b0fe51eeda",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=30, n_pcs=20) # Computing the neighborhood graph\n",
    "sc.tl.umap(adata) # Build a UMAP to visualize the neighbourhood graph\n",
    "sc.pl.umap(adata, color= 'Lvl4', cmap= 'vlag', use_raw=False) # View the clusteringimport numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, cluster_key, output_path, crop_coords=None):\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "        cluster_labels = subset_adata.obs[cluster_key].astype(str)\n",
    "\n",
    "        # Crop if coordinates are provided\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "            cluster_labels = subset_adata.obs[cluster_key].astype(str)\n",
    "\n",
    "        # Identify unique clusters\n",
    "        unique_clusters = sorted(cluster_labels.unique())\n",
    "        \n",
    "        # Assign colors\n",
    "        epi_clusters = [c for c in unique_clusters if c.startswith(\"Ep\")]\n",
    "        other_clusters = [c for c in unique_clusters if c not in epi_clusters]\n",
    "\n",
    "        num_epi_clusters = len(epi_clusters)\n",
    "        epi_colors = sns.color_palette(\"husl\", num_epi_clusters)  # Unique colors for \"Ep\" clusters\n",
    "        cluster_colors = {c: epi_colors[i] for i, c in enumerate(epi_clusters)}\n",
    "        cluster_colors.update({c: \"#E5E5E5\" for c in other_clusters})  # Grey for all others\n",
    "\n",
    "        # Aspect ratio for figure size\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set white background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Draw points on top\n",
    "        colors = [cluster_colors.get(c, \"#E5E5E5\") for c in cluster_labels]\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=4.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path, dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "# Example usage:\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/HV205B_Epi.pdf'\n",
    "\n",
    "# Generate plot\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl4', output_path, crop_coords)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01c2eead-c35d-4d93-9072-ac0d8fa6911b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl2'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9482d1a2-a42e-4aaf-8694-a39aa2c9a827",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin([\"Oral Epi\", \"Crevicular Epi\"]), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin([\"Oral Epi\", \"Crevicular Epi\"]), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Epi_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2', 'Lvl4', output_path, crop_coords)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9641ac94-382e-4500-bf83-279e0b104764",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin([\"Vascular\", \"Fibroblast\"]), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin([\"Vascular\", \"Fibroblast\"]), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        #plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Str_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2', 'Lvl4', output_path, crop_coords)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f684120-21c7-43bc-a5df-aa07237f3434",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin([\"T/NK\", \"B/Plasma\"]), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin([\"T/NK\", \"B/Plasma\"]), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # Specifically check if 'B' is in the clusters and assign it a color\n",
    "    if 'B' in cluster_colors:\n",
    "        cluster_colors['B'] = '#00008B'  # Or choose any color you like for 'B' (e.g., tomato red)\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lymphoid_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2', 'Lvl4', output_path, crop_coords)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a1cee7c-3f36-4b57-82c9-ee8fb7314ad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin([\"Myeloid\"]), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin([\"Myeloid\"]), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Myeloid_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2', 'Lvl4', output_path, crop_coords)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a692546-7083-4923-8891-df0a20e5b175",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin([\"Myeloid\"]), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin([\"Myeloid\"]), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Myeloid_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl2', 'Lvl4', output_path, crop_coords)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ee16d5-0dba-40d3-a5be-057adb9545a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Lvl1'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b2fd23c-83c6-45d6-ad83-e2fd168ece7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin(['Immune', 'Vascular', 'Fibroblast', 'Epithelial']), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin(['Immune', 'Vascular', 'Fibroblast', 'Epithelial']), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # Specifically check if 'B' is in the clusters and assign it a color\n",
    "    if 'B' in cluster_colors:\n",
    "        cluster_colors['B'] = '#00008B'  # Or choose any color you like for 'B' (e.g., tomato red)\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lvl4_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl1', 'Lvl4', output_path, crop_coords)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d0aee6c-38b1-4780-84b1-0db2352c636a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Define the preferred order for Lvl1 categories\n",
    "    category_order = ['Epithelial', 'Fibroblast', 'Vascular', 'Immune']\n",
    "\n",
    "    # Identify unique clusters in Lvl4 grouped by ordered Lvl1 categories\n",
    "    sorted_clusters = []\n",
    "    for category in category_order:\n",
    "        clusters = sorted(adata.obs.loc[adata.obs[lvl2_key] == category, lvl4_key].unique())\n",
    "        sorted_clusters.extend(clusters)\n",
    "\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin(category_order), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors following the specified order\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(sorted_clusters))\n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(sorted_clusters)}\n",
    "\n",
    "    # Specifically check if 'B' is in the clusters and assign it a fixed color\n",
    "    if 'B' in cluster_colors:\n",
    "        cluster_colors['B'] = '#400000'  # Dark Blue for 'B'\n",
    "\n",
    "    # Assign all non-selected clusters a grey color\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio adjustment\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl4\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only selected clusters and grouping others as 'Other'.\"\"\"\n",
    "    selected_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(selected_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(selected_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add clusters in preferred order\n",
    "    for i, (cluster, color) in enumerate(selected_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(selected_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(selected_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['HV205B']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lvl4_HV205B_{}.pdf'\n",
    "crop_coords = (2200, 3618, 180, 2074)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl1', 'Lvl4', output_path, crop_coords)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b238dac5-2c67-483a-8290-9b2bb77866d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Identify unique clusters in Lvl4 where Lvl2 == \"Myeloid\"\n",
    "    Epi_clusters = sorted(adata.obs.loc[adata.obs[lvl2_key].isin(['Immune', 'Vascular', 'Fibroblast', 'Epithelial']), lvl4_key].unique())\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin(['Immune', 'Vascular', 'Fibroblast', 'Epithelial']), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors to Myeloid Lvl4 clusters\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(Epi_clusters))  \n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(Epi_clusters)}\n",
    "\n",
    "    # Specifically check if 'B' is in the clusters and assign it a color\n",
    "    if 'B' in cluster_colors:\n",
    "        cluster_colors['B'] = '#00008B'  # Or choose any color you like for 'B' (e.g., tomato red)\n",
    "\n",
    "    # All non-Myeloid clusters in grey\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl2\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only Myeloid Lvl4 clusters and grouping others as 'Other'.\"\"\"\n",
    "    Epi_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(Epi_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(Epi_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add Myeloid Lvl4 clusters\n",
    "    for i, (cluster, color) in enumerate(Epi_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(Epi_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(Epi_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lvl4_HV214s_{}.pdf'\n",
    "crop_coords = (2832, 4250, 480, 2374)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl1', 'Lvl4', output_path, crop_coords)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5863e90b-1d2e-496a-9ab8-6913fa0288b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Filter for B/Plasma cells in the lvl2_key column\n",
    "b_plasma_cells = adata.obs[adata.obs['Lvl2'] == \"T/NK\"]\n",
    "\n",
    "# Step 2: Get the unique values in the lvl4_key column for B/Plasma cells\n",
    "b_plasma_clusters = b_plasma_cells['Lvl4'].unique()\n",
    "\n",
    "# Print the unique Lvl4 clusters for B/Plasma\n",
    "print(\"Unique Lvl4 clusters for B/Plasma:\", b_plasma_clusters)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f5ef730-4347-4044-bbc7-236de802fb1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "def generate_spatial_plot(adata, sample_names, lvl2_key, lvl4_key, output_path, crop_coords=None):\n",
    "    # Define the preferred order for Lvl1 categories\n",
    "    category_order = ['Epithelial', 'Fibroblast', 'Vascular', 'Immune']\n",
    "\n",
    "    # Identify unique clusters in Lvl4 grouped by ordered Lvl1 categories\n",
    "    sorted_clusters = []\n",
    "    for category in category_order:\n",
    "        clusters = sorted(adata.obs.loc[adata.obs[lvl2_key] == category, lvl4_key].unique())\n",
    "        sorted_clusters.extend(clusters)\n",
    "\n",
    "    other_clusters = sorted(adata.obs.loc[~adata.obs[lvl2_key].isin(category_order), lvl4_key].unique())\n",
    "\n",
    "    # Assign unique colors following the specified order\n",
    "    cmap = plt.cm.get_cmap(\"turbo\", len(sorted_clusters))\n",
    "    cluster_colors = {cluster: cmap(i) for i, cluster in enumerate(sorted_clusters)}\n",
    "\n",
    "    # Specifically check if 'B' is in the clusters and assign it a fixed color\n",
    "    if 'B' in cluster_colors:\n",
    "        cluster_colors['B'] = '#400000'  # Dark Blue for 'B'\n",
    "\n",
    "    # Assign all non-selected clusters a grey color\n",
    "    other_color = \"#A0A0A0\"\n",
    "    for cluster in other_clusters:\n",
    "        cluster_colors[cluster] = other_color\n",
    "\n",
    "    for sample_name in sample_names:\n",
    "        subset_adata = adata[adata.obs['sample'] == sample_name]\n",
    "\n",
    "        x_coords = subset_adata.obs['x_centroid'].values\n",
    "        y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        if crop_coords:\n",
    "            x_min, x_max, y_min, y_max = crop_coords\n",
    "            crop_mask = (\n",
    "                (x_coords >= x_min) & (x_coords <= x_max) &\n",
    "                (y_coords >= y_min) & (y_coords <= y_max)\n",
    "            )\n",
    "            subset_adata = subset_adata[crop_mask]\n",
    "            x_coords = subset_adata.obs['x_centroid'].values\n",
    "            y_coords = subset_adata.obs['y_centroid'].values\n",
    "\n",
    "        # Aspect ratio adjustment\n",
    "        x_range = x_coords.max() - x_coords.min()\n",
    "        y_range = y_coords.max() - y_coords.min()\n",
    "        aspect_ratio = x_range / y_range\n",
    "        figsize = (6 * aspect_ratio, 6) if aspect_ratio > 1 else (6, 6 / aspect_ratio)\n",
    "\n",
    "        # Generate plot\n",
    "        plt.figure(figsize=figsize)\n",
    "        ax = plt.gca()\n",
    "\n",
    "        # Set background\n",
    "        plt.gcf().patch.set_facecolor('white')\n",
    "        ax.set_facecolor('white')\n",
    "\n",
    "        # Assign colors based on Lvl4\n",
    "        colors = subset_adata.obs[lvl4_key].map(cluster_colors).fillna(other_color).values\n",
    "        plt.scatter(x_coords, y_coords, c=colors, s=8.5, zorder=2)\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlim(x_coords.min(), x_coords.max())\n",
    "        ax.set_ylim(y_coords.min(), y_coords.max())\n",
    "\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "        plt.gcf().set_dpi(300)\n",
    "        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "        plt.savefig(output_path.format(\"Main\"), dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # Generate a separate legend plot\n",
    "    generate_legend_plot(cluster_colors, other_color, output_path.format(\"Legend\"))\n",
    "\n",
    "def generate_legend_plot(cluster_colors, other_color, legend_output_path):\n",
    "    \"\"\"Creates and saves a separate legend plot, showing only selected clusters and grouping others as 'Other'.\"\"\"\n",
    "    selected_clusters = {k: v for k, v in cluster_colors.items() if v != other_color}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1, len(selected_clusters) * 0.3 + 0.6))  # Adjust size dynamically\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, len(selected_clusters) + 1)  # Extra space for 'Other'\n",
    "\n",
    "    # Add clusters in preferred order\n",
    "    for i, (cluster, color) in enumerate(selected_clusters.items()):\n",
    "        ax.add_patch(mpatches.Rectangle((0, i), 1, 0.8, color=color, ec=\"black\"))\n",
    "        ax.text(1.1, i + 0.4, cluster, va=\"center\", fontsize=8)\n",
    "\n",
    "    # Add \"Other\" category in grey\n",
    "    ax.add_patch(mpatches.Rectangle((0, len(selected_clusters)), 1, 0.8, color=other_color, ec=\"black\"))\n",
    "    ax.text(1.1, len(selected_clusters) + 0.4, \"Other\", va=\"center\", fontsize=8, fontweight=\"bold\")\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    plt.savefig(legend_output_path, dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "# Example usage\n",
    "sample_names = ['s1r3']\n",
    "output_path = '/data/vasileiosionat2/IBEX_FINAL/Scimap/Outputs/Figures_paper/Figure2/Lvl4_HV214s_{}.pdf'\n",
    "crop_coords = (2832, 4250, 480, 2374)\n",
    "\n",
    "generate_spatial_plot(adata, sample_names, 'Lvl1', 'Lvl4', output_path, crop_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "155dc9bc-e727-419f-b688-df0d8614e4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.obs['Lvl5'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df2bd585-bccd-43bf-90e4-62543e822c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'plot_niches' is a DataFrame or similar data structure\n",
    "df = pd.DataFrame(merged.obs)\n",
    "df['Lvl5'] = df['Lvl5'].astype('category')\n",
    "\n",
    "# Aggregate counts\n",
    "niche_groups = df.groupby(['final_label_citeSeq', 'niche_cc14']).size().unstack(fill_value=0)\n",
    "\n",
    "# Identify clusters to remove\n",
    "clusters_to_remove = df['final_label_citeSeq'].str.contains('Mix') | df['final_label_citeSeq'].str.startswith('Neuro')\n",
    "\n",
    "# Get the unique clusters to be removed\n",
    "clusters_to_remove = df.loc[clusters_to_remove, 'final_label_citeSeq'].unique()\n",
    "\n",
    "# Filter out these clusters from the plot\n",
    "niche_groups_filtered = niche_groups[~niche_groups.index.isin(clusters_to_remove)]\n",
    "\n",
    "# Calculate relative frequencies\n",
    "niche_groups_relative = niche_groups_filtered.div(niche_groups_filtered.sum(axis=1), axis=0)\n",
    "\n",
    "# Plot heatmap with enhanced colormap\n",
    "plt.figure(figsize=(16, 20))  # Adjusted figsize to accommodate the x-axis labels\n",
    "sns.heatmap(niche_groups_relative, cmap='plasma', cbar_kws={'label': 'Relative Frequency'}, linewidths=0.05, linecolor='black')\n",
    "plt.title('Relative Cell Type Distribution in Niches')\n",
    "plt.xlabel('Niche')\n",
    "plt.ylabel('Cell Cluster')\n",
    "plt.xticks(rotation=45, ha='right')  # Adjust rotation and alignment of xticks\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()  # Ensures all elements fit within the figure area\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scimap",
   "language": "python",
   "name": "scimap"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
